<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>fbaModelServices</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:root@localhost" />
</head>

<body style="background-color: white">



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#fbaModelServices">fbaModelServices</a>
    <ul>
      <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
      <li><a href="#EXAMPLE-OF-API-USE-IN-PERL">EXAMPLE OF API USE IN PERL</a></li>
      <li><a href="#AUTHENTICATION">AUTHENTICATION</a></li>
      <li><a href="#WORKSPACE">WORKSPACE</a></li>
      <li><a href="#get_models">get_models</a></li>
      <li><a href="#get_fbas">get_fbas</a></li>
      <li><a href="#get_gapfills">get_gapfills</a></li>
      <li><a href="#get_gapgens">get_gapgens</a></li>
      <li><a href="#get_reactions">get_reactions</a></li>
      <li><a href="#get_compounds">get_compounds</a></li>
      <li><a href="#get_alias">get_alias</a></li>
      <li><a href="#get_aliassets">get_aliassets</a></li>
      <li><a href="#get_media">get_media</a></li>
      <li><a href="#get_biochemistry">get_biochemistry</a></li>
      <li><a href="#import_probanno">import_probanno</a></li>
      <li><a href="#genome_object_to_workspace">genome_object_to_workspace</a></li>
      <li><a href="#genome_to_workspace">genome_to_workspace</a></li>
      <li><a href="#domains_to_workspace">domains_to_workspace</a></li>
      <li><a href="#compute_domains">compute_domains</a></li>
      <li><a href="#add_feature_translation">add_feature_translation</a></li>
      <li><a href="#genome_to_fbamodel">genome_to_fbamodel</a></li>
      <li><a href="#translate_fbamodel">translate_fbamodel</a></li>
      <li><a href="#build_pangenome">build_pangenome</a></li>
      <li><a href="#genome_heatmap_from_pangenome">genome_heatmap_from_pangenome</a></li>
      <li><a href="#ortholog_family_from_pangenome">ortholog_family_from_pangenome</a></li>
      <li><a href="#pangenome_to_proteome_comparison">pangenome_to_proteome_comparison</a></li>
      <li><a href="#import_fbamodel">import_fbamodel</a></li>
      <li><a href="#export_fbamodel">export_fbamodel</a></li>
      <li><a href="#export_object">export_object</a></li>
      <li><a href="#export_genome">export_genome</a></li>
      <li><a href="#adjust_model_reaction">adjust_model_reaction</a></li>
      <li><a href="#adjust_biomass_reaction">adjust_biomass_reaction</a></li>
      <li><a href="#addmedia">addmedia</a></li>
      <li><a href="#export_media">export_media</a></li>
      <li><a href="#runfba">runfba</a></li>
      <li><a href="#quantitative_optimization">quantitative_optimization</a></li>
      <li><a href="#generate_model_stats">generate_model_stats</a></li>
      <li><a href="#minimize_reactions">minimize_reactions</a></li>
      <li><a href="#export_fba">export_fba</a></li>
      <li><a href="#import_phenotypes">import_phenotypes</a></li>
      <li><a href="#simulate_phenotypes">simulate_phenotypes</a></li>
      <li><a href="#add_media_transporters">add_media_transporters</a></li>
      <li><a href="#export_phenotypeSimulationSet">export_phenotypeSimulationSet</a></li>
      <li><a href="#integrate_reconciliation_solutions">integrate_reconciliation_solutions</a></li>
      <li><a href="#queue_runfba">queue_runfba</a></li>
      <li><a href="#queue_gapfill_model">queue_gapfill_model</a></li>
      <li><a href="#gapfill_model">gapfill_model</a></li>
      <li><a href="#queue_gapgen_model">queue_gapgen_model</a></li>
      <li><a href="#gapgen_model">gapgen_model</a></li>
      <li><a href="#queue_wildtype_phenotype_reconciliation">queue_wildtype_phenotype_reconciliation</a></li>
      <li><a href="#queue_reconciliation_sensitivity_analysis">queue_reconciliation_sensitivity_analysis</a></li>
      <li><a href="#queue_combine_wildtype_phenotype_reconciliation">queue_combine_wildtype_phenotype_reconciliation</a></li>
      <li><a href="#run_job">run_job</a></li>
      <li><a href="#queue_job">queue_job</a></li>
      <li><a href="#set_cofactors">set_cofactors</a></li>
      <li><a href="#find_reaction_synonyms">find_reaction_synonyms</a></li>
      <li><a href="#role_to_reactions">role_to_reactions</a></li>
      <li><a href="#reaction_sensitivity_analysis">reaction_sensitivity_analysis</a></li>
      <li><a href="#filter_iterative_solutions">filter_iterative_solutions</a></li>
      <li><a href="#delete_noncontributing_reactions">delete_noncontributing_reactions</a></li>
      <li><a href="#annotate_workspace_Genome">annotate_workspace_Genome</a></li>
      <li><a href="#gtf_to_genome">gtf_to_genome</a></li>
      <li><a href="#fasta_to_ProteinSet">fasta_to_ProteinSet</a></li>
      <li><a href="#ProteinSet_to_Genome">ProteinSet_to_Genome</a></li>
      <li><a href="#fasta_to_ContigSet">fasta_to_ContigSet</a></li>
      <li><a href="#ContigSet_to_Genome">ContigSet_to_Genome</a></li>
      <li><a href="#probanno_to_genome">probanno_to_genome</a></li>
      <li><a href="#get_mapping">get_mapping</a></li>
      <li><a href="#subsystem_of_roles">subsystem_of_roles</a></li>
      <li><a href="#adjust_mapping_role">adjust_mapping_role</a></li>
      <li><a href="#adjust_mapping_complex">adjust_mapping_complex</a></li>
      <li><a href="#adjust_mapping_subsystem">adjust_mapping_subsystem</a></li>
      <li><a href="#get_template_model">get_template_model</a></li>
      <li><a href="#import_template_fbamodel">import_template_fbamodel</a></li>
      <li><a href="#adjust_template_reaction">adjust_template_reaction</a></li>
      <li><a href="#adjust_template_biomass">adjust_template_biomass</a></li>
      <li><a href="#add_stimuli">add_stimuli</a></li>
      <li><a href="#import_regulatory_model">import_regulatory_model</a></li>
      <li><a href="#compare_models">compare_models</a></li>
      <li><a href="#compare_fbas">compare_fbas</a></li>
      <li><a href="#compare_genomes">compare_genomes</a></li>
      <li><a href="#import_metagenome_annotation">import_metagenome_annotation</a></li>
      <li><a href="#models_to_community_model">models_to_community_model</a></li>
      <li><a href="#metagenome_to_fbamodels">metagenome_to_fbamodels</a></li>
      <li><a href="#import_expression">import_expression</a></li>
      <li><a href="#import_regulome">import_regulome</a></li>
      <li><a href="#create_promconstraint">create_promconstraint</a></li>
      <li><a href="#add_biochemistry_compounds">add_biochemistry_compounds</a></li>
      <li><a href="#update_object_references">update_object_references</a></li>
      <li><a href="#add_reactions">add_reactions</a></li>
      <li><a href="#remove_reactions">remove_reactions</a></li>
      <li><a href="#modify_reactions">modify_reactions</a></li>
      <li><a href="#add_features">add_features</a></li>
      <li><a href="#remove_features">remove_features</a></li>
      <li><a href="#modify_features">modify_features</a></li>
      <li><a href="#import_trainingset">import_trainingset</a></li>
      <li><a href="#preload_trainingset">preload_trainingset</a></li>
      <li><a href="#build_classifier">build_classifier</a></li>
      <li><a href="#classify_genomes">classify_genomes</a></li>
      <li><a href="#build_tissue_model">build_tissue_model</a></li>
    </ul>
  </li>
  <li><a href="#TYPES">TYPES</a>
    <ul>
      <li><a href="#bool">bool</a></li>
      <li><a href="#mdlrxn_kbid">mdlrxn_kbid</a></li>
      <li><a href="#ws_ref">ws_ref</a></li>
      <li><a href="#kbase_id">kbase_id</a></li>
      <li><a href="#workspace_id">workspace_id</a></li>
      <li><a href="#complex_id">complex_id</a></li>
      <li><a href="#template_id">template_id</a></li>
      <li><a href="#role_id">role_id</a></li>
      <li><a href="#object_type">object_type</a></li>
      <li><a href="#object_id">object_id</a></li>
      <li><a href="#username">username</a></li>
      <li><a href="#timestamp">timestamp</a></li>
      <li><a href="#compound_id">compound_id</a></li>
      <li><a href="#biochemistry_id">biochemistry_id</a></li>
      <li><a href="#genome_id">genome_id</a></li>
      <li><a href="#promconstraint_id">promconstraint_id</a></li>
      <li><a href="#sample_id">sample_id</a></li>
      <li><a href="#series_id">series_id</a></li>
      <li><a href="#contig_id">contig_id</a></li>
      <li><a href="#feature_type">feature_type</a></li>
      <li><a href="#modelcompartment_id">modelcompartment_id</a></li>
      <li><a href="#modelcompound_id">modelcompound_id</a></li>
      <li><a href="#feature_id">feature_id</a></li>
      <li><a href="#reaction_id">reaction_id</a></li>
      <li><a href="#modelreaction_id">modelreaction_id</a></li>
      <li><a href="#biomass_id">biomass_id</a></li>
      <li><a href="#media_id">media_id</a></li>
      <li><a href="#fba_id">fba_id</a></li>
      <li><a href="#gapgen_id">gapgen_id</a></li>
      <li><a href="#gapfill_id">gapfill_id</a></li>
      <li><a href="#gapgensolution_id">gapgensolution_id</a></li>
      <li><a href="#gapfillsolution_id">gapfillsolution_id</a></li>
      <li><a href="#fbamodel_id">fbamodel_id</a></li>
      <li><a href="#mapping_id">mapping_id</a></li>
      <li><a href="#regulome_id">regulome_id</a></li>
      <li><a href="#regmodel_id">regmodel_id</a></li>
      <li><a href="#compartment_id">compartment_id</a></li>
      <li><a href="#expression_id">expression_id</a></li>
      <li><a href="#phenotype_set_id">phenotype_set_id</a></li>
      <li><a href="#workspace_ref">workspace_ref</a></li>
      <li><a href="#probanno_id">probanno_id</a></li>
      <li><a href="#reaction_synonyms_id">reaction_synonyms_id</a></li>
      <li><a href="#Genome_uid">Genome_uid</a></li>
      <li><a href="#object_metadata">object_metadata</a></li>
      <li><a href="#md5">md5</a></li>
      <li><a href="#md5s">md5s</a></li>
      <li><a href="#region_of_dna">region_of_dna</a></li>
      <li><a href="#location">location</a></li>
      <li><a href="#annotation">annotation</a></li>
      <li><a href="#gene_hit">gene_hit</a></li>
      <li><a href="#alt_func">alt_func</a></li>
      <li><a href="#feature">feature</a></li>
      <li><a href="#contig">contig</a></li>
      <li><a href="#GenomeObject">GenomeObject</a></li>
      <li><a href="#annotationProbability">annotationProbability</a></li>
      <li><a href="#probanno_id1">probanno_id</a></li>
      <li><a href="#alt_func1">alt_func</a></li>
      <li><a href="#ProbAnnoFeature">ProbAnnoFeature</a></li>
      <li><a href="#ProbabilisticAnnotation">ProbabilisticAnnotation</a></li>
      <li><a href="#ReactionProbability">ReactionProbability</a></li>
      <li><a href="#Biochemistry">Biochemistry</a></li>
      <li><a href="#MediaCompound">MediaCompound</a></li>
      <li><a href="#Media">Media</a></li>
      <li><a href="#Compound">Compound</a></li>
      <li><a href="#Reaction">Reaction</a></li>
      <li><a href="#ModelCompartment">ModelCompartment</a></li>
      <li><a href="#ModelCompound">ModelCompound</a></li>
      <li><a href="#ModelReaction">ModelReaction</a></li>
      <li><a href="#BiomassCompound">BiomassCompound</a></li>
      <li><a href="#ModelBiomass">ModelBiomass</a></li>
      <li><a href="#FBAMeta">FBAMeta</a></li>
      <li><a href="#GapGenMeta">GapGenMeta</a></li>
      <li><a href="#GapFillMeta">GapFillMeta</a></li>
      <li><a href="#FBAModel">FBAModel</a></li>
      <li><a href="#GeneAssertion">GeneAssertion</a></li>
      <li><a href="#CompoundFlux">CompoundFlux</a></li>
      <li><a href="#ReactionFlux">ReactionFlux</a></li>
      <li><a href="#MetaboliteProduction">MetaboliteProduction</a></li>
      <li><a href="#MinimalMediaPrediction">MinimalMediaPrediction</a></li>
      <li><a href="#bound">bound</a></li>
      <li><a href="#term">term</a></li>
      <li><a href="#constraint">constraint</a></li>
      <li><a href="#FBAFormulation">FBAFormulation</a></li>
      <li><a href="#FBA">FBA</a></li>
      <li><a href="#GapfillingFormulation">GapfillingFormulation</a></li>
      <li><a href="#reactionAddition">reactionAddition</a></li>
      <li><a href="#biomassRemoval">biomassRemoval</a></li>
      <li><a href="#mediaAddition">mediaAddition</a></li>
      <li><a href="#GapFillSolution">GapFillSolution</a></li>
      <li><a href="#GapFill">GapFill</a></li>
      <li><a href="#GapgenFormulation">GapgenFormulation</a></li>
      <li><a href="#reactionRemoval">reactionRemoval</a></li>
      <li><a href="#biomassAddition">biomassAddition</a></li>
      <li><a href="#mediaRemoval">mediaRemoval</a></li>
      <li><a href="#GapgenSolution">GapgenSolution</a></li>
      <li><a href="#GapGen">GapGen</a></li>
      <li><a href="#Phenotype">Phenotype</a></li>
      <li><a href="#PhenotypeSet">PhenotypeSet</a></li>
      <li><a href="#phenotypeSimulationSet_id">phenotypeSimulationSet_id</a></li>
      <li><a href="#PhenotypeSimulation">PhenotypeSimulation</a></li>
      <li><a href="#PhenotypeSimulationSet">PhenotypeSimulationSet</a></li>
      <li><a href="#reactionSpecification">reactionSpecification</a></li>
      <li><a href="#PhenotypeSensitivityAnalysis">PhenotypeSensitivityAnalysis</a></li>
      <li><a href="#job_id">job_id</a></li>
      <li><a href="#CommandArguments">CommandArguments</a></li>
      <li><a href="#clusterjob">clusterjob</a></li>
      <li><a href="#JobObject">JobObject</a></li>
      <li><a href="#ETCNodes">ETCNodes</a></li>
      <li><a href="#ETCDiagramSpecs">ETCDiagramSpecs</a></li>
      <li><a href="#TemplateReactions">TemplateReactions</a></li>
      <li><a href="#ComplexReactions">ComplexReactions</a></li>
      <li><a href="#RoleComplexReactions">RoleComplexReactions</a></li>
      <li><a href="#ReactionDefinition">ReactionDefinition</a></li>
      <li><a href="#ReactionSynonyms">ReactionSynonyms</a></li>
      <li><a href="#ReactionSynonymsObject">ReactionSynonymsObject</a></li>
      <li><a href="#get_models_params">get_models_params</a></li>
      <li><a href="#get_fbas_params">get_fbas_params</a></li>
      <li><a href="#get_gapfills_params">get_gapfills_params</a></li>
      <li><a href="#get_gapgens_params">get_gapgens_params</a></li>
      <li><a href="#get_reactions_params">get_reactions_params</a></li>
      <li><a href="#get_compounds_params">get_compounds_params</a></li>
      <li><a href="#get_alias_params">get_alias_params</a></li>
      <li><a href="#get_alias_outputs">get_alias_outputs</a></li>
      <li><a href="#get_aliassets_params">get_aliassets_params</a></li>
      <li><a href="#get_media_params">get_media_params</a></li>
      <li><a href="#get_biochemistry_params">get_biochemistry_params</a></li>
      <li><a href="#import_probanno_params">import_probanno_params</a></li>
      <li><a href="#Genome_uid1">Genome_uid</a></li>
      <li><a href="#genome_object_to_workspace_params">genome_object_to_workspace_params</a></li>
      <li><a href="#genome_to_workspace_params">genome_to_workspace_params</a></li>
      <li><a href="#domains_to_workspace_params">domains_to_workspace_params</a></li>
      <li><a href="#compute_domains_params">compute_domains_params</a></li>
      <li><a href="#translation">translation</a></li>
      <li><a href="#add_feature_translation_params">add_feature_translation_params</a></li>
      <li><a href="#genome_to_fbamodel_params">genome_to_fbamodel_params</a></li>
      <li><a href="#translate_fbamodel_params">translate_fbamodel_params</a></li>
      <li><a href="#build_pangenome_params">build_pangenome_params</a></li>
      <li><a href="#heat_map_matrix">heat_map_matrix</a></li>
      <li><a href="#genome_heatmap_from_pangenome_params">genome_heatmap_from_pangenome_params</a></li>
      <li><a href="#ortholog_data">ortholog_data</a></li>
      <li><a href="#ortholog_family_from_pangenome_params">ortholog_family_from_pangenome_params</a></li>
      <li><a href="#pangenome_to_proteome_comparison_params">pangenome_to_proteome_comparison_params</a></li>
      <li><a href="#import_fbamodel_params">import_fbamodel_params</a></li>
      <li><a href="#export_fbamodel_params">export_fbamodel_params</a></li>
      <li><a href="#export_object_params">export_object_params</a></li>
      <li><a href="#export_genome_params">export_genome_params</a></li>
      <li><a href="#adjust_model_reaction_params">adjust_model_reaction_params</a></li>
      <li><a href="#adjust_biomass_reaction_params">adjust_biomass_reaction_params</a></li>
      <li><a href="#addmedia_params">addmedia_params</a></li>
      <li><a href="#export_media_params">export_media_params</a></li>
      <li><a href="#runfba_params">runfba_params</a></li>
      <li><a href="#quantitative_optimization_params">quantitative_optimization_params</a></li>
      <li><a href="#generate_model_stats_params">generate_model_stats_params</a></li>
      <li><a href="#subsystem_statistics">subsystem_statistics</a></li>
      <li><a href="#model_statistics">model_statistics</a></li>
      <li><a href="#minimize_reactions_params">minimize_reactions_params</a></li>
      <li><a href="#export_fba_params">export_fba_params</a></li>
      <li><a href="#import_phenotypes_params">import_phenotypes_params</a></li>
      <li><a href="#simulate_phenotypes_params">simulate_phenotypes_params</a></li>
      <li><a href="#add_media_transporters_params">add_media_transporters_params</a></li>
      <li><a href="#export_phenotypeSimulationSet_params">export_phenotypeSimulationSet_params</a></li>
      <li><a href="#integrate_reconciliation_solutions_params">integrate_reconciliation_solutions_params</a></li>
      <li><a href="#queue_runfba_params">queue_runfba_params</a></li>
      <li><a href="#gapfill_model_params">gapfill_model_params</a></li>
      <li><a href="#gapgen_model_params">gapgen_model_params</a></li>
      <li><a href="#wildtype_phenotype_reconciliation_params">wildtype_phenotype_reconciliation_params</a></li>
      <li><a href="#queue_reconciliation_sensitivity_analysis_params">queue_reconciliation_sensitivity_analysis_params</a></li>
      <li><a href="#combine_wildtype_phenotype_reconciliation_params">combine_wildtype_phenotype_reconciliation_params</a></li>
      <li><a href="#run_job_params">run_job_params</a></li>
      <li><a href="#queue_job_params">queue_job_params</a></li>
      <li><a href="#set_cofactors_params">set_cofactors_params</a></li>
      <li><a href="#find_reaction_synonyms_params">find_reaction_synonyms_params</a></li>
      <li><a href="#role_to_reactions_params">role_to_reactions_params</a></li>
      <li><a href="#kb_sub_id">kb_sub_id</a></li>
      <li><a href="#kb_id">kb_id</a></li>
      <li><a href="#ws_id">ws_id</a></li>
      <li><a href="#ws_sub_id">ws_sub_id</a></li>
      <li><a href="#ReactionSensitivityAnalysisCorrectedReaction">ReactionSensitivityAnalysisCorrectedReaction</a></li>
      <li><a href="#ReactionSensitivityAnalysisReaction">ReactionSensitivityAnalysisReaction</a></li>
      <li><a href="#ReactionSensitivityAnalysis">ReactionSensitivityAnalysis</a></li>
      <li><a href="#rxnprob_id">rxnprob_id</a></li>
      <li><a href="#reaction_sensitivity_analysis_params">reaction_sensitivity_analysis_params</a></li>
      <li><a href="#filter_iterative_solutions_params">filter_iterative_solutions_params</a></li>
      <li><a href="#delete_noncontributing_reactions_params">delete_noncontributing_reactions_params</a></li>
      <li><a href="#AnnotationParameters">AnnotationParameters</a></li>
      <li><a href="#annotate_workspace_Genome_params">annotate_workspace_Genome_params</a></li>
      <li><a href="#gtf_to_genome_params">gtf_to_genome_params</a></li>
      <li><a href="#fasta_to_ProteinSet_params">fasta_to_ProteinSet_params</a></li>
      <li><a href="#ProteinSet_to_Genome_params">ProteinSet_to_Genome_params</a></li>
      <li><a href="#fasta_to_ContigSet_params">fasta_to_ContigSet_params</a></li>
      <li><a href="#ContigSet_to_Genome_params">ContigSet_to_Genome_params</a></li>
      <li><a href="#probanno_to_genome_params">probanno_to_genome_params</a></li>
      <li><a href="#FunctionalRole">FunctionalRole</a></li>
      <li><a href="#ComplexRole">ComplexRole</a></li>
      <li><a href="#Complex">Complex</a></li>
      <li><a href="#subsystem_id">subsystem_id</a></li>
      <li><a href="#Subsystem">Subsystem</a></li>
      <li><a href="#Mapping">Mapping</a></li>
      <li><a href="#get_mapping_params">get_mapping_params</a></li>
      <li><a href="#subsysclass">subsysclass</a></li>
      <li><a href="#subsysclasses">subsysclasses</a></li>
      <li><a href="#subsystem_of_roles_params">subsystem_of_roles_params</a></li>
      <li><a href="#adjust_mapping_role_params">adjust_mapping_role_params</a></li>
      <li><a href="#adjust_mapping_complex_params">adjust_mapping_complex_params</a></li>
      <li><a href="#adjust_mapping_subsystem_params">adjust_mapping_subsystem_params</a></li>
      <li><a href="#temprxn_id">temprxn_id</a></li>
      <li><a href="#TemplateReaction">TemplateReaction</a></li>
      <li><a href="#TemplateBiomassCompounds">TemplateBiomassCompounds</a></li>
      <li><a href="#tempbiomass_id">tempbiomass_id</a></li>
      <li><a href="#TemplateBiomass">TemplateBiomass</a></li>
      <li><a href="#TemplateModel">TemplateModel</a></li>
      <li><a href="#get_template_model_params">get_template_model_params</a></li>
      <li><a href="#import_template_fbamodel_params">import_template_fbamodel_params</a></li>
      <li><a href="#adjust_template_reaction_params">adjust_template_reaction_params</a></li>
      <li><a href="#adjust_template_biomass_params">adjust_template_biomass_params</a></li>
      <li><a href="#add_stimuli_params">add_stimuli_params</a></li>
      <li><a href="#Stimuli">Stimuli</a></li>
      <li><a href="#RegulatoryModelRegulonStimuli">RegulatoryModelRegulonStimuli</a></li>
      <li><a href="#RegulatoryModelRegulon">RegulatoryModelRegulon</a></li>
      <li><a href="#RegulatoryModel">RegulatoryModel</a></li>
      <li><a href="#import_regulatory_model_params">import_regulatory_model_params</a></li>
      <li><a href="#compare_models_params">compare_models_params</a></li>
      <li><a href="#ModelComparisonModel">ModelComparisonModel</a></li>
      <li><a href="#ModelCompareReaction">ModelCompareReaction</a></li>
      <li><a href="#ModelComparisonData">ModelComparisonData</a></li>
      <li><a href="#compare_fbas_params">compare_fbas_params</a></li>
      <li><a href="#compare_genomes_params">compare_genomes_params</a></li>
      <li><a href="#MetagenomeAnnotationOTUFunction">MetagenomeAnnotationOTUFunction</a></li>
      <li><a href="#MetagenomeAnnotationOTU">MetagenomeAnnotationOTU</a></li>
      <li><a href="#MetagenomeAnnotation">MetagenomeAnnotation</a></li>
      <li><a href="#import_metagenome_annotation_params">import_metagenome_annotation_params</a></li>
      <li><a href="#models_to_community_model_params">models_to_community_model_params</a></li>
      <li><a href="#metagenome_to_fbamodels_params">metagenome_to_fbamodels_params</a></li>
      <li><a href="#sample_id1">sample_id</a></li>
      <li><a href="#series_id1">series_id</a></li>
      <li><a href="#measurement">measurement</a></li>
      <li><a href="#ExpressionDataSample">ExpressionDataSample</a></li>
      <li><a href="#import_expression_params">import_expression_params</a></li>
      <li><a href="#effector">effector</a></li>
      <li><a href="#locus">locus</a></li>
      <li><a href="#operon">operon</a></li>
      <li><a href="#regulon">regulon</a></li>
      <li><a href="#import_regulome_params">import_regulome_params</a></li>
      <li><a href="#CreatePromConstraintParameters">CreatePromConstraintParameters</a></li>
      <li><a href="#add_biochemistry_compounds_params">add_biochemistry_compounds_params</a></li>
      <li><a href="#update_object_references_params">update_object_references_params</a></li>
      <li><a href="#add_reactions_params">add_reactions_params</a></li>
      <li><a href="#remove_reactions_params">remove_reactions_params</a></li>
      <li><a href="#modify_reactions_params">modify_reactions_params</a></li>
      <li><a href="#add_features_params">add_features_params</a></li>
      <li><a href="#remove_features_params">remove_features_params</a></li>
      <li><a href="#modify_features_params">modify_features_params</a></li>
      <li><a href="#import_trainingset_params">import_trainingset_params</a></li>
      <li><a href="#preload_trainingset_params">preload_trainingset_params</a></li>
      <li><a href="#build_classifier_params">build_classifier_params</a></li>
      <li><a href="#classify_genomes_params">classify_genomes_params</a></li>
      <li><a href="#build_tissue_model_params">build_tissue_model_params</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::KBase::fbaModelServices::Client</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<h1 id="fbaModelServices">fbaModelServices</h1>

<h2 id="SYNOPSIS">SYNOPSIS</h2>

<p>The FBA Model Services include support related to the reconstruction, curation, reconciliation, and analysis of metabolic models. This includes commands to:</p>

<p>1.) Load genome typed objects into a workspace</p>

<p>2.) Build a model from a genome typed object and curate the model</p>

<p>3.) Analyze a model with flux balance analysis</p>

<p>4.) Simulate and reconcile a model to an imported set of growth phenotype data</p>

<h2 id="EXAMPLE-OF-API-USE-IN-PERL">EXAMPLE OF API USE IN PERL</h2>

<p>To use the API, first you need to instantiate a fbaModelServices client object:</p>

<p>my $client = Bio::KBase::fbaModelServices::Client-&gt;new;</p>

<p>Next, you can run API commands on the client object:</p>

<p>my $objmeta = $client-&gt;genome_to_workspace({ genome =&gt; &quot;kb|g.0&quot;, workspace =&gt; &quot;myWorkspace&quot; }); my $objmeta = $client-&gt;genome_to_fbamodel({ model =&gt; &quot;myModel&quot; workspace =&gt; &quot;myWorkspace&quot; });</p>

<h2 id="AUTHENTICATION">AUTHENTICATION</h2>

<p>Each and every function in this service takes a hash reference as its single argument. This hash reference may contain a key <code>auth</code> whose value is a bearer token for the user making the request. If this is not provided a default user &quot;public&quot; is assumed.</p>

<h2 id="WORKSPACE">WORKSPACE</h2>

<p>A workspace is a named collection of objects owned by a specific user, that may be viewable or editable by other users.Functions that operate on workspaces take a <code>workspace_id</code>, which is an alphanumeric string that uniquely identifies a workspace among all workspaces.</p>

<h2 id="get_models">get_models</h2>

<pre><code>  $out_models = $obj-&gt;get_models($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types">Parameter and return types</dt>
<dd>



<pre>
$input is a get_models_params
$out_models is a reference to a list where each element is an FBAModel
get_models_params is a reference to a hash where the following keys are defined:
	models has a value which is a reference to a list where each element is a fbamodel_id
	workspaces has a value which is a reference to a list where each element is a workspace_id
	auth has a value which is a string
	id_type has a value which is a string
fbamodel_id is a string
workspace_id is a string
FBAModel is a reference to a hash where the following keys are defined:
	id has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	genome has a value which is a genome_id
	genome_workspace has a value which is a workspace_id
	map has a value which is a mapping_id
	map_workspace has a value which is a workspace_id
	biochemistry has a value which is a biochemistry_id
	biochemistry_workspace has a value which is a workspace_id
	name has a value which is a string
	type has a value which is a string
	status has a value which is a string
	biomasses has a value which is a reference to a list where each element is a ModelBiomass
	compartments has a value which is a reference to a list where each element is a ModelCompartment
	reactions has a value which is a reference to a list where each element is a ModelReaction
	compounds has a value which is a reference to a list where each element is a ModelCompound
	fbas has a value which is a reference to a list where each element is an FBAMeta
	integrated_gapfillings has a value which is a reference to a list where each element is a GapFillMeta
	unintegrated_gapfillings has a value which is a reference to a list where each element is a GapFillMeta
	integrated_gapgenerations has a value which is a reference to a list where each element is a GapGenMeta
	unintegrated_gapgenerations has a value which is a reference to a list where each element is a GapGenMeta
genome_id is a string
mapping_id is a string
biochemistry_id is a string
ModelBiomass is a reference to a hash where the following keys are defined:
	id has a value which is a biomass_id
	name has a value which is a string
	definition has a value which is a string
	biomass_compounds has a value which is a reference to a list where each element is a BiomassCompound
biomass_id is a string
BiomassCompound is a reference to a list containing 3 items:
	0: (modelcompound) a modelcompound_id
	1: (coefficient) a float
	2: (name) a string
modelcompound_id is a string
ModelCompartment is a reference to a hash where the following keys are defined:
	id has a value which is a modelcompartment_id
	name has a value which is a string
	pH has a value which is a float
	potential has a value which is a float
	index has a value which is an int
modelcompartment_id is a string
ModelReaction is a reference to a hash where the following keys are defined:
	id has a value which is a modelreaction_id
	reaction has a value which is a reaction_id
	name has a value which is a string
	direction has a value which is a string
	equation has a value which is a string
	definition has a value which is a string
	gapfilled has a value which is a bool
	features has a value which is a reference to a list where each element is a feature_id
	compartment has a value which is a modelcompartment_id
modelreaction_id is a string
reaction_id is a string
bool is an int
feature_id is a string
ModelCompound is a reference to a hash where the following keys are defined:
	id has a value which is a modelcompound_id
	compound has a value which is a compound_id
	name has a value which is a string
	compartment has a value which is a modelcompartment_id
compound_id is a string
FBAMeta is a reference to a list containing 6 items:
	0: (id) a fba_id
	1: (workspace) a workspace_id
	2: (media) a media_id
	3: (media_workspace) a workspace_id
	4: (objective) a float
	5: (ko) a reference to a list where each element is a feature_id
fba_id is a string
media_id is a string
GapFillMeta is a reference to a list containing 6 items:
	0: (id) a gapfill_id
	1: (workspace) a workspace_id
	2: (media) a media_id
	3: (media_workspace) a workspace_id
	4: (done) a bool
	5: (ko) a reference to a list where each element is a feature_id
gapfill_id is a string
GapGenMeta is a reference to a list containing 6 items:
	0: (id) a gapgen_id
	1: (workspace) a workspace_id
	2: (media) a media_id
	3: (media_workspace) a workspace_id
	4: (done) a bool
	5: (ko) a reference to a list where each element is a feature_id
gapgen_id is a string</pre>

</dd>
<dt id="Description">Description</dt>
<dd>

<p>Returns model data for input ids</p>

</dd>
</dl>

<h2 id="get_fbas">get_fbas</h2>

<pre><code>  $out_fbas = $obj-&gt;get_fbas($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types1">Parameter and return types</dt>
<dd>



<pre>
$input is a get_fbas_params
$out_fbas is a reference to a list where each element is an FBA
get_fbas_params is a reference to a hash where the following keys are defined:
	fbas has a value which is a reference to a list where each element is a fba_id
	workspaces has a value which is a reference to a list where each element is a workspace_id
	auth has a value which is a string
	id_type has a value which is a string
fba_id is a string
workspace_id is a string
FBA is a reference to a hash where the following keys are defined:
	id has a value which is a fba_id
	workspace has a value which is a workspace_id
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	objective has a value which is a float
	isComplete has a value which is a bool
	formulation has a value which is an FBAFormulation
	minimalMediaPredictions has a value which is a reference to a list where each element is a MinimalMediaPrediction
	metaboliteProductions has a value which is a reference to a list where each element is a MetaboliteProduction
	reactionFluxes has a value which is a reference to a list where each element is a ReactionFlux
	compoundFluxes has a value which is a reference to a list where each element is a CompoundFlux
	geneAssertions has a value which is a reference to a list where each element is a GeneAssertion
fbamodel_id is a string
bool is an int
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
MinimalMediaPrediction is a reference to a hash where the following keys are defined:
	optionalNutrients has a value which is a reference to a list where each element is a compound_id
	essentialNutrients has a value which is a reference to a list where each element is a compound_id
MetaboliteProduction is a reference to a list containing 3 items:
	0: (maximumProduction) a float
	1: (modelcompound) a modelcompound_id
	2: (name) a string
modelcompound_id is a string
ReactionFlux is a reference to a list containing 8 items:
	0: (reaction) a modelreaction_id
	1: (value) a float
	2: (upperBound) a float
	3: (lowerBound) a float
	4: (max) a float
	5: (min) a float
	6: (type) a string
	7: (definition) a string
modelreaction_id is a string
CompoundFlux is a reference to a list containing 8 items:
	0: (compound) a modelcompound_id
	1: (value) a float
	2: (upperBound) a float
	3: (lowerBound) a float
	4: (max) a float
	5: (min) a float
	6: (type) a string
	7: (name) a string
GeneAssertion is a reference to a list containing 4 items:
	0: (feature) a feature_id
	1: (growthFraction) a float
	2: (growth) a float
	3: (isEssential) a bool</pre>

</dd>
<dt id="Description1">Description</dt>
<dd>

<p>Returns data for the requested flux balance analysis formulations</p>

</dd>
</dl>

<h2 id="get_gapfills">get_gapfills</h2>

<pre><code>  $out_gapfills = $obj-&gt;get_gapfills($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types2">Parameter and return types</dt>
<dd>



<pre>
$input is a get_gapfills_params
$out_gapfills is a reference to a list where each element is a GapFill
get_gapfills_params is a reference to a hash where the following keys are defined:
	gapfills has a value which is a reference to a list where each element is a gapfill_id
	workspaces has a value which is a reference to a list where each element is a workspace_id
	auth has a value which is a string
	id_type has a value which is a string
gapfill_id is a string
workspace_id is a string
GapFill is a reference to a hash where the following keys are defined:
	id has a value which is a gapfill_id
	workspace has a value which is a workspace_id
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	isComplete has a value which is a bool
	formulation has a value which is a GapfillingFormulation
	solutions has a value which is a reference to a list where each element is a GapFillSolution
fbamodel_id is a string
bool is an int
GapfillingFormulation is a reference to a hash where the following keys are defined:
	formulation has a value which is an FBAFormulation
	num_solutions has a value which is an int
	nomediahyp has a value which is a bool
	nobiomasshyp has a value which is a bool
	nogprhyp has a value which is a bool
	nopathwayhyp has a value which is a bool
	allowunbalanced has a value which is a bool
	activitybonus has a value which is a float
	drainpen has a value which is a float
	directionpen has a value which is a float
	nostructpen has a value which is a float
	unfavorablepen has a value which is a float
	nodeltagpen has a value which is a float
	biomasstranspen has a value which is a float
	singletranspen has a value which is a float
	transpen has a value which is a float
	blacklistedrxns has a value which is a reference to a list where each element is a reaction_id
	gauranteedrxns has a value which is a reference to a list where each element is a reaction_id
	allowedcmps has a value which is a reference to a list where each element is a compartment_id
	probabilisticAnnotation has a value which is a probanno_id
	probabilisticAnnotation_workspace has a value which is a workspace_id
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
compartment_id is a string
probanno_id is a string
GapFillSolution is a reference to a hash where the following keys are defined:
	id has a value which is a gapfillsolution_id
	objective has a value which is a float
	integrated has a value which is a bool
	biomassRemovals has a value which is a reference to a list where each element is a biomassRemoval
	mediaAdditions has a value which is a reference to a list where each element is a mediaAddition
	reactionAdditions has a value which is a reference to a list where each element is a reactionAddition
gapfillsolution_id is a string
biomassRemoval is a reference to a list containing 2 items:
	0: (compound) a compound_id
	1: (name) a string
mediaAddition is a reference to a list containing 2 items:
	0: (compound) a compound_id
	1: (name) a string
reactionAddition is a reference to a list containing 5 items:
	0: (reaction) a reaction_id
	1: (direction) a string
	2: (compartment_id) a string
	3: (equation) a string
	4: (definition) a string</pre>

</dd>
<dt id="Description2">Description</dt>
<dd>

<p>Returns data for the requested gap filling simulations</p>

</dd>
</dl>

<h2 id="get_gapgens">get_gapgens</h2>

<pre><code>  $out_gapgens = $obj-&gt;get_gapgens($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types3">Parameter and return types</dt>
<dd>



<pre>
$input is a get_gapgens_params
$out_gapgens is a reference to a list where each element is a GapGen
get_gapgens_params is a reference to a hash where the following keys are defined:
	gapgens has a value which is a reference to a list where each element is a gapgen_id
	workspaces has a value which is a reference to a list where each element is a workspace_id
	auth has a value which is a string
	id_type has a value which is a string
gapgen_id is a string
workspace_id is a string
GapGen is a reference to a hash where the following keys are defined:
	id has a value which is a gapgen_id
	workspace has a value which is a workspace_id
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	isComplete has a value which is a bool
	formulation has a value which is a GapgenFormulation
	solutions has a value which is a reference to a list where each element is a GapgenSolution
fbamodel_id is a string
bool is an int
GapgenFormulation is a reference to a hash where the following keys are defined:
	formulation has a value which is an FBAFormulation
	refmedia has a value which is a media_id
	refmedia_workspace has a value which is a workspace_id
	num_solutions has a value which is an int
	nomediahyp has a value which is a bool
	nobiomasshyp has a value which is a bool
	nogprhyp has a value which is a bool
	nopathwayhyp has a value which is a bool
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
GapgenSolution is a reference to a hash where the following keys are defined:
	id has a value which is a gapgensolution_id
	objective has a value which is a float
	biomassAdditions has a value which is a reference to a list where each element is a biomassAddition
	mediaRemovals has a value which is a reference to a list where each element is a mediaRemoval
	reactionRemovals has a value which is a reference to a list where each element is a reactionRemoval
gapgensolution_id is a string
biomassAddition is a reference to a list containing 2 items:
	0: (compound) a compound_id
	1: (name) a string
mediaRemoval is a reference to a list containing 2 items:
	0: (compound) a compound_id
	1: (name) a string
reactionRemoval is a reference to a list containing 4 items:
	0: (reaction) a modelreaction_id
	1: (direction) a string
	2: (equation) a string
	3: (definition) a string
modelreaction_id is a string</pre>

</dd>
<dt id="Description3">Description</dt>
<dd>

<p>Returns data for the requested gap generation simulations</p>

</dd>
</dl>

<h2 id="get_reactions">get_reactions</h2>

<pre><code>  $out_reactions = $obj-&gt;get_reactions($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types4">Parameter and return types</dt>
<dd>



<pre>
$input is a get_reactions_params
$out_reactions is a reference to a list where each element is a Reaction
get_reactions_params is a reference to a hash where the following keys are defined:
	reactions has a value which is a reference to a list where each element is a reaction_id
	auth has a value which is a string
	id_type has a value which is a string
reaction_id is a string
Reaction is a reference to a hash where the following keys are defined:
	id has a value which is a reaction_id
	name has a value which is a string
	abbrev has a value which is a string
	enzymes has a value which is a reference to a list where each element is a string
	direction has a value which is a string
	reversibility has a value which is a string
	deltaG has a value which is a float
	deltaGErr has a value which is a float
	equation has a value which is a string
	definition has a value which is a string</pre>

</dd>
<dt id="Description4">Description</dt>
<dd>

<p>Returns data for the requested reactions</p>

</dd>
</dl>

<h2 id="get_compounds">get_compounds</h2>

<pre><code>  $out_compounds = $obj-&gt;get_compounds($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types5">Parameter and return types</dt>
<dd>



<pre>
$input is a get_compounds_params
$out_compounds is a reference to a list where each element is a Compound
get_compounds_params is a reference to a hash where the following keys are defined:
	compounds has a value which is a reference to a list where each element is a compound_id
	auth has a value which is a string
	id_type has a value which is a string
compound_id is a string
Compound is a reference to a hash where the following keys are defined:
	id has a value which is a compound_id
	abbrev has a value which is a string
	name has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	charge has a value which is a float
	deltaG has a value which is a float
	deltaGErr has a value which is a float
	formula has a value which is a string</pre>

</dd>
<dt id="Description5">Description</dt>
<dd>

<p>Returns data for the requested compounds</p>

</dd>
</dl>

<h2 id="get_alias">get_alias</h2>

<pre><code>  $output = $obj-&gt;get_alias($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types6">Parameter and return types</dt>
<dd>



<pre>
$input is a get_alias_params
$output is a reference to a list where each element is a get_alias_outputs
get_alias_params is a reference to a hash where the following keys are defined:
	object_type has a value which is a string
	input_id_type has a value which is a string
	output_id_type has a value which is a string
	input_ids has a value which is a reference to a list where each element is a string
	auth has a value which is a string
get_alias_outputs is a reference to a hash where the following keys are defined:
	original_id has a value which is a string
	aliases has a value which is a reference to a list where each element is a string</pre>

</dd>
<dt id="Description6">Description</dt>
<dd>

<p>Turns one compound I into another of a different type</p>

</dd>
</dl>

<h2 id="get_aliassets">get_aliassets</h2>

<pre><code>  $aliassets = $obj-&gt;get_aliassets($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types7">Parameter and return types</dt>
<dd>



<pre>
$input is a get_aliassets_params
$aliassets is a reference to a list where each element is a string
get_aliassets_params is a reference to a hash where the following keys are defined:
	object_type has a value which is a string
	auth has a value which is a string</pre>

</dd>
<dt id="Description7">Description</dt>
<dd>

<p>Get possible types of aliases (alias sets)</p>

</dd>
</dl>

<h2 id="get_media">get_media</h2>

<pre><code>  $out_media = $obj-&gt;get_media($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types8">Parameter and return types</dt>
<dd>



<pre>
$input is a get_media_params
$out_media is a reference to a list where each element is a Media
get_media_params is a reference to a hash where the following keys are defined:
	medias has a value which is a reference to a list where each element is a media_id
	workspaces has a value which is a reference to a list where each element is a workspace_id
	auth has a value which is a string
media_id is a string
workspace_id is a string
Media is a reference to a hash where the following keys are defined:
	id has a value which is a media_id
	name has a value which is a string
	media_compounds has a value which is a reference to a list where each element is a MediaCompound
	pH has a value which is a float
	temperature has a value which is a float
MediaCompound is a reference to a hash where the following keys are defined:
	compound has a value which is a compound_id
	name has a value which is a string
	concentration has a value which is a float
	max_flux has a value which is a float
	min_flux has a value which is a float
compound_id is a string</pre>

</dd>
<dt id="Description8">Description</dt>
<dd>

<p>Returns data for the requested media formulations</p>

</dd>
</dl>

<h2 id="get_biochemistry">get_biochemistry</h2>

<pre><code>  $out_biochemistry = $obj-&gt;get_biochemistry($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types9">Parameter and return types</dt>
<dd>



<pre>
$input is a get_biochemistry_params
$out_biochemistry is a Biochemistry
get_biochemistry_params is a reference to a hash where the following keys are defined:
	biochemistry has a value which is a biochemistry_id
	biochemistry_workspace has a value which is a workspace_id
	id_type has a value which is a string
	auth has a value which is a string
biochemistry_id is a string
workspace_id is a string
Biochemistry is a reference to a hash where the following keys are defined:
	id has a value which is a biochemistry_id
	name has a value which is a string
	compounds has a value which is a reference to a list where each element is a compound_id
	reactions has a value which is a reference to a list where each element is a reaction_id
	media has a value which is a reference to a list where each element is a media_id
compound_id is a string
reaction_id is a string
media_id is a string</pre>

</dd>
<dt id="Description9">Description</dt>
<dd>

<p>Returns biochemistry object</p>

</dd>
</dl>

<h2 id="import_probanno">import_probanno</h2>

<pre><code>  $probannoMeta = $obj-&gt;import_probanno($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types10">Parameter and return types</dt>
<dd>



<pre>
$input is an import_probanno_params
$probannoMeta is an object_metadata
import_probanno_params is a reference to a hash where the following keys are defined:
	probanno has a value which is a probanno_id
	workspace has a value which is a workspace_id
	genome has a value which is a genome_id
	genome_workspace has a value which is a workspace_id
	annotationProbabilities has a value which is a reference to a list where each element is an annotationProbability
	ignore_errors has a value which is a bool
	auth has a value which is a string
	overwrite has a value which is a bool
probanno_id is a string
workspace_id is a string
genome_id is a string
annotationProbability is a reference to a list containing 3 items:
	0: (feature) a feature_id
	1: (function) a string
	2: (probability) a float
feature_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description10">Description</dt>
<dd>

<p>Loads an input genome object into the workspace.</p>

</dd>
</dl>

<h2 id="genome_object_to_workspace">genome_object_to_workspace</h2>

<pre><code>  $genomeMeta = $obj-&gt;genome_object_to_workspace($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types11">Parameter and return types</dt>
<dd>



<pre>
$input is a genome_object_to_workspace_params
$genomeMeta is an object_metadata
genome_object_to_workspace_params is a reference to a hash where the following keys are defined:
	uid has a value which is a Genome_uid
	genomeobj has a value which is a GenomeObject
	workspace has a value which is a workspace_id
	auth has a value which is a string
	overwrite has a value which is a bool
Genome_uid is a string
GenomeObject is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	contigs has a value which is a reference to a list where each element is a contig
	features has a value which is a reference to a list where each element is a feature
genome_id is a string
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
contig_id is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	alternative_functions has a value which is a reference to a list where each element is an alt_func
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	annotations has a value which is a reference to a list where each element is an annotation
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
alt_func is a reference to a list containing 2 items:
	0: (function) a string
	1: (probability) a float
gene_hit is a reference to a list containing 2 items:
	0: (gene) a feature_id
	1: (blast_score) a float
annotation is a reference to a list containing 3 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) an int
workspace_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description11">Description</dt>
<dd>

<p>Loads an input genome object into the workspace.</p>

</dd>
</dl>

<h2 id="genome_to_workspace">genome_to_workspace</h2>

<pre><code>  $genomeMeta = $obj-&gt;genome_to_workspace($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types12">Parameter and return types</dt>
<dd>



<pre>
$input is a genome_to_workspace_params
$genomeMeta is an object_metadata
genome_to_workspace_params is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	workspace has a value which is a workspace_id
	sourceLogin has a value which is a string
	sourcePassword has a value which is a string
	source has a value which is a string
	auth has a value which is a string
	overwrite has a value which is a bool
	uid has a value which is a Genome_uid
genome_id is a string
workspace_id is a string
bool is an int
Genome_uid is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description12">Description</dt>
<dd>

<p>Retrieves a genome from the CDM and saves it as a genome object in the workspace.</p>

</dd>
</dl>

<h2 id="domains_to_workspace">domains_to_workspace</h2>

<pre><code>  $GenomeDomainMeta = $obj-&gt;domains_to_workspace($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types13">Parameter and return types</dt>
<dd>



<pre>
$input is a domains_to_workspace_params
$GenomeDomainMeta is an object_metadata
domains_to_workspace_params is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	output_id has a value which is a string
	workspace has a value which is a workspace_id
	auth has a value which is a string
genome_id is a string
workspace_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description13">Description</dt>
<dd>

<p>Computes or fetches domains for a genome</p>

</dd>
</dl>

<h2 id="compute_domains">compute_domains</h2>

<pre><code>  $output = $obj-&gt;compute_domains($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types14">Parameter and return types</dt>
<dd>



<pre>
$params is a compute_domains_params
$output is an object_metadata
compute_domains_params is a reference to a hash where the following keys are defined:
	genome has a value which is a string
	genome_workspace has a value which is a string
	proteins has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: a string
	1: a string	workspace has a value which is a workspace_id
workspace_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string
</pre>

</dd>
<dt id="Description14">Description</dt>
<dd>

<p>Computes domains for either a genome or a list of proteins</p>

</dd>
</dl>

<h2 id="add_feature_translation">add_feature_translation</h2>

<pre><code>  $genomeMeta = $obj-&gt;add_feature_translation($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types15">Parameter and return types</dt>
<dd>



<pre>
$input is an add_feature_translation_params
$genomeMeta is an object_metadata
add_feature_translation_params is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	workspace has a value which is a workspace_id
	translations has a value which is a reference to a list where each element is a translation
	id_type has a value which is a string
	auth has a value which is a string
	overwrite has a value which is a bool
genome_id is a string
workspace_id is a string
translation is a reference to a list containing 2 items:
	0: (foreign_id) a string
	1: (feature) a feature_id
feature_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description15">Description</dt>
<dd>

<p>Adds a new set of alternative feature IDs to the specified genome typed object</p>

</dd>
</dl>

<h2 id="genome_to_fbamodel">genome_to_fbamodel</h2>

<pre><code>  $modelMeta = $obj-&gt;genome_to_fbamodel($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types16">Parameter and return types</dt>
<dd>



<pre>
$input is a genome_to_fbamodel_params
$modelMeta is an object_metadata
genome_to_fbamodel_params is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_workspace has a value which is a workspace_id
	templatemodel has a value which is a template_id
	templatemodel_workspace has a value which is a workspace_id
	model has a value which is a fbamodel_id
	coremodel has a value which is a bool
	workspace has a value which is a workspace_id
	auth has a value which is a string
	fulldb has a value which is a bool
genome_id is a string
workspace_id is a string
template_id is a string
fbamodel_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description16">Description</dt>
<dd>

<p>Build a genome-scale metabolic model based on annotations in an input genome typed object</p>

</dd>
</dl>

<h2 id="translate_fbamodel">translate_fbamodel</h2>

<pre><code>  $modelMeta = $obj-&gt;translate_fbamodel($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types17">Parameter and return types</dt>
<dd>



<pre>
$input is a translate_fbamodel_params
$modelMeta is an object_metadata
translate_fbamodel_params is a reference to a hash where the following keys are defined:
	protcomp has a value which is a string
	protcomp_workspace has a value which is a string
	model has a value which is a string
	model_workspace has a value which is a string
	workspace has a value which is a workspace_id
workspace_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description17">Description</dt>
<dd>

<p>Translate an existing model to a new genome based on the genome comparison object</p>

</dd>
</dl>

<h2 id="build_pangenome">build_pangenome</h2>

<pre><code>  $output = $obj-&gt;build_pangenome($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types18">Parameter and return types</dt>
<dd>



<pre>
$input is a build_pangenome_params
$output is an object_metadata
build_pangenome_params is a reference to a hash where the following keys are defined:
	genomes has a value which is a reference to a list where each element is a string
	genome_workspace has a value which is a reference to a list where each element is a string
	workspace has a value which is a workspace_id
workspace_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description18">Description</dt>
<dd>

<p>Translate an existing model to a new genome based on the genome comparison object</p>

</dd>
</dl>

<h2 id="genome_heatmap_from_pangenome">genome_heatmap_from_pangenome</h2>

<pre><code>  $output = $obj-&gt;genome_heatmap_from_pangenome($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types19">Parameter and return types</dt>
<dd>



<pre>
$input is a genome_heatmap_from_pangenome_params
$output is a heat_map_matrix
genome_heatmap_from_pangenome_params is a reference to a hash where the following keys are defined:
	pangenome has a value which is a string
	pangenome_workspace has a value which is a string
	workspace has a value which is a string
heat_map_matrix is a reference to a hash where the following keys are defined:
	is_refs has a value which is a bool
	labels has a value which is a reference to a list where each element is a string
	matrix has a value which is a reference to a list where each element is a reference to a list where each element is a float
bool is an int</pre>

</dd>
<dt id="Description19">Description</dt>
<dd>

<p>Builds a comparason matrix for genomes included in a pangenome object</p>

</dd>
</dl>

<h2 id="ortholog_family_from_pangenome">ortholog_family_from_pangenome</h2>

<pre><code>  $output = $obj-&gt;ortholog_family_from_pangenome($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types20">Parameter and return types</dt>
<dd>



<pre>
$input is an ortholog_family_from_pangenome_params
$output is an ortholog_data
ortholog_family_from_pangenome_params is a reference to a hash where the following keys are defined:
	pangenome has a value which is a string
	pangenome_workspace has a value which is a string
	orthologid has a value which is a string
	workspace has a value which is a string
ortholog_data is a reference to a hash where the following keys are defined:
	gene_data has a value which is a reference to a list where each element is a reference to a list containing 5 items:
	0: a string
	1: a string
	2: a string
	3: a string
	4: a float	protein_heatmap has a value which is a heat_map_matrix
heat_map_matrix is a reference to a hash where the following keys are defined:
	is_refs has a value which is a bool
	labels has a value which is a reference to a list where each element is a string
	matrix has a value which is a reference to a list where each element is a reference to a list where each element is a float
bool is an int
</pre>

</dd>
<dt id="Description20">Description</dt>
<dd>

<p>Returns more detailed data from a single ortholog family from a pangenome object</p>

</dd>
</dl>

<h2 id="pangenome_to_proteome_comparison">pangenome_to_proteome_comparison</h2>

<pre><code>  $output = $obj-&gt;pangenome_to_proteome_comparison($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types21">Parameter and return types</dt>
<dd>



<pre>
$input is a pangenome_to_proteome_comparison_params
$output is an object_metadata
pangenome_to_proteome_comparison_params is a reference to a hash where the following keys are defined:
	pangenome has a value which is a string
	pangenome_workspace has a value which is a string
	outputid has a value which is a string
	workspace has a value which is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description21">Description</dt>
<dd>

<p>Builds a proteome comparison object from a pangenome object</p>

</dd>
</dl>

<h2 id="import_fbamodel">import_fbamodel</h2>

<pre><code>  $modelMeta = $obj-&gt;import_fbamodel($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types22">Parameter and return types</dt>
<dd>



<pre>
$input is an import_fbamodel_params
$modelMeta is an object_metadata
import_fbamodel_params is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_workspace has a value which is a workspace_id
	biomass has a value which is a string
	reactions has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a string
	1: (direction) a string
	2: (compartment) a string
	3: (gpr) a string	model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	ignore_errors has a value which is a bool
	auth has a value which is a string
	overwrite has a value which is a bool
genome_id is a string
workspace_id is a string
fbamodel_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string
</pre>

</dd>
<dt id="Description22">Description</dt>
<dd>

<p>Import a model from an input table of model and gene IDs</p>

</dd>
</dl>

<h2 id="export_fbamodel">export_fbamodel</h2>

<pre><code>  $output = $obj-&gt;export_fbamodel($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types23">Parameter and return types</dt>
<dd>



<pre>
$input is an export_fbamodel_params
$output is a string
export_fbamodel_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	fbas has a value which is a reference to a list where each element is a fba_id
	format has a value which is a string
	auth has a value which is a string
fbamodel_id is a string
workspace_id is a string
fba_id is a string</pre>

</dd>
<dt id="Description23">Description</dt>
<dd>

<p>This function exports the specified FBAModel to a specified format (sbml,html)</p>

</dd>
</dl>

<h2 id="export_object">export_object</h2>

<pre><code>  $output = $obj-&gt;export_object($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types24">Parameter and return types</dt>
<dd>



<pre>
$input is an export_object_params
$output is a string
export_object_params is a reference to a hash where the following keys are defined:
	reference has a value which is a workspace_ref
	type has a value which is a string
	format has a value which is a string
	auth has a value which is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description24">Description</dt>
<dd>

<p>This function prints the object pointed to by the input reference in the specified format</p>

</dd>
</dl>

<h2 id="export_genome">export_genome</h2>

<pre><code>  $output = $obj-&gt;export_genome($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types25">Parameter and return types</dt>
<dd>



<pre>
$input is an export_genome_params
$output is a string
export_genome_params is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	workspace has a value which is a workspace_id
	format has a value which is a string
	auth has a value which is a string
genome_id is a string
workspace_id is a string</pre>

</dd>
<dt id="Description25">Description</dt>
<dd>

<p>This function exports the specified FBAModel to a specified format (sbml,html)</p>

</dd>
</dl>

<h2 id="adjust_model_reaction">adjust_model_reaction</h2>

<pre><code>  $modelMeta = $obj-&gt;adjust_model_reaction($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types26">Parameter and return types</dt>
<dd>



<pre>
$input is an adjust_model_reaction_params
$modelMeta is an object_metadata
adjust_model_reaction_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	reaction has a value which is a reference to a list where each element is a reaction_id
	direction has a value which is a reference to a list where each element is a string
	compartment has a value which is a reference to a list where each element is a compartment_id
	compartmentIndex has a value which is a reference to a list where each element is an int
	gpr has a value which is a reference to a list where each element is a string
	removeReaction has a value which is a bool
	addReaction has a value which is a bool
	overwrite has a value which is a bool
	auth has a value which is a string
fbamodel_id is a string
workspace_id is a string
reaction_id is a string
compartment_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description26">Description</dt>
<dd>

<p>Enables the manual addition of a reaction to model</p>

</dd>
</dl>

<h2 id="adjust_biomass_reaction">adjust_biomass_reaction</h2>

<pre><code>  $modelMeta = $obj-&gt;adjust_biomass_reaction($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types27">Parameter and return types</dt>
<dd>



<pre>
$input is an adjust_biomass_reaction_params
$modelMeta is an object_metadata
adjust_biomass_reaction_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	biomass has a value which is a biomass_id
	coefficients has a value which is a reference to a list where each element is a float
	compounds has a value which is a reference to a list where each element is a compound_id
	compartments has a value which is a reference to a list where each element is a compartment_id
	compartmentIndecies has a value which is a reference to a list where each element is an int
	auth has a value which is a string
fbamodel_id is a string
workspace_id is a string
biomass_id is a string
compound_id is a string
compartment_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description27">Description</dt>
<dd>

<p>Enables the manual adjustment of model biomass reaction</p>

</dd>
</dl>

<h2 id="addmedia">addmedia</h2>

<pre><code>  $mediaMeta = $obj-&gt;addmedia($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types28">Parameter and return types</dt>
<dd>



<pre>
$input is an addmedia_params
$mediaMeta is an object_metadata
addmedia_params is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	workspace has a value which is a workspace_id
	name has a value which is a string
	isDefined has a value which is a bool
	isMinimal has a value which is a bool
	type has a value which is a string
	compounds has a value which is a reference to a list where each element is a string
	concentrations has a value which is a reference to a list where each element is a float
	maxflux has a value which is a reference to a list where each element is a float
	minflux has a value which is a reference to a list where each element is a float
	overwrite has a value which is a bool
	auth has a value which is a string
media_id is a string
workspace_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description28">Description</dt>
<dd>

<p>Add media condition to workspace</p>

</dd>
</dl>

<h2 id="export_media">export_media</h2>

<pre><code>  $output = $obj-&gt;export_media($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types29">Parameter and return types</dt>
<dd>



<pre>
$input is an export_media_params
$output is a string
export_media_params is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	workspace has a value which is a workspace_id
	format has a value which is a string
	auth has a value which is a string
media_id is a string
workspace_id is a string</pre>

</dd>
<dt id="Description29">Description</dt>
<dd>

<p>Exports media in specified format (html,readable)</p>

</dd>
</dl>

<h2 id="runfba">runfba</h2>

<pre><code>  $fbaMeta = $obj-&gt;runfba($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types30">Parameter and return types</dt>
<dd>



<pre>
$input is a runfba_params
$fbaMeta is an object_metadata
runfba_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	formulation has a value which is an FBAFormulation
	fva has a value which is a bool
	simulateko has a value which is a bool
	minimizeflux has a value which is a bool
	findminmedia has a value which is a bool
	notes has a value which is a string
	fba has a value which is a fba_id
	workspace has a value which is a workspace_id
	auth has a value which is a string
	overwrite has a value which is a bool
	add_to_model has a value which is a bool
fbamodel_id is a string
workspace_id is a string
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
bool is an int
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
fba_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description30">Description</dt>
<dd>

<p>Run flux balance analysis and return ID of FBA object with results</p>

</dd>
</dl>

<h2 id="quantitative_optimization">quantitative_optimization</h2>

<pre><code>  $output = $obj-&gt;quantitative_optimization($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types31">Parameter and return types</dt>
<dd>



<pre>
$input is a quantitative_optimization_params
$output is an object_metadata
quantitative_optimization_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	formulation has a value which is an FBAFormulation
	outputid has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	biomass has a value which is a string
fbamodel_id is a string
workspace_id is a string
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
bool is an int
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description31">Description</dt>
<dd>

<p>Identify ways to adjust model to quantitatively match specified uptake, growth, and excretion constraints</p>

</dd>
</dl>

<h2 id="generate_model_stats">generate_model_stats</h2>

<pre><code>  $output = $obj-&gt;generate_model_stats($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types32">Parameter and return types</dt>
<dd>



<pre>
$input is a generate_model_stats_params
$output is a model_statistics
generate_model_stats_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
fbamodel_id is a string
workspace_id is a string
model_statistics is a reference to a hash where the following keys are defined:
	total_reactions has a value which is an int
	total_genes has a value which is an int
	total_compounds has a value which is an int
	extracellular_compounds has a value which is an int
	intracellular_compounds has a value which is an int
	transport_reactions has a value which is an int
	subsystem_reactions has a value which is an int
	subsystem_genes has a value which is an int
	spontaneous_reactions has a value which is an int
	reactions_with_genes has a value which is an int
	gapfilled_reactions has a value which is an int
	model_genes has a value which is an int
	minimal_essential_genes has a value which is an int
	complete_essential_genes has a value which is an int
	minimal_essential_reactions has a value which is an int
	complete_essential_reactions has a value which is an int
	minimal_blocked_reactions has a value which is an int
	complete_blocked_reactions has a value which is an int
	minimal_variable_reactions has a value which is an int
	complete_variable_reactions has a value which is an int
	growth_complete_media has a value which is a bool
	growth_minimal_media has a value which is a bool
	subsystems has a value which is a reference to a list where each element is a subsystem_statistics
bool is an int
subsystem_statistics is a reference to a hash where the following keys are defined:
	name has a value which is a string
	class has a value which is a string
	subclass has a value which is a string
	genes has a value which is an int
	reactions has a value which is an int
	model_genes has a value which is an int
	minimal_essential_genes has a value which is an int
	complete_essential_genes has a value which is an int
	minimal_essential_reactions has a value which is an int
	complete_essential_reactions has a value which is an int
	minimal_blocked_reactions has a value which is an int
	complete_blocked_reactions has a value which is an int
	minimal_variable_reactions has a value which is an int
	complete_variable_reactions has a value which is an int</pre>

</dd>
<dt id="Description32">Description</dt>
<dd>

<p>Generate statistics with model and associated genome properties</p>

</dd>
</dl>

<h2 id="minimize_reactions">minimize_reactions</h2>

<pre><code>  $fbaMeta = $obj-&gt;minimize_reactions($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types33">Parameter and return types</dt>
<dd>



<pre>
$input is a minimize_reactions_params
$fbaMeta is an object_metadata
minimize_reactions_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	workspace has a value which is a workspace_id
	formulation has a value which is an FBAFormulation
	reactions has a value which is a reference to a list where each element is a string
	all_model_reactions has a value which is a bool
	reaction_costs has a value which is a reference to a hash where the key is a string and the value is a float
	output_id has a value which is a fba_id
fbamodel_id is a string
workspace_id is a string
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
bool is an int
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
fba_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description33">Description</dt>
<dd>

<p>Minimize the specified set of reactions while maintaining the FBA objective above a specified threshold</p>

</dd>
</dl>

<h2 id="export_fba">export_fba</h2>

<pre><code>  $output = $obj-&gt;export_fba($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types34">Parameter and return types</dt>
<dd>



<pre>
$input is an export_fba_params
$output is a string
export_fba_params is a reference to a hash where the following keys are defined:
	fba has a value which is a fba_id
	workspace has a value which is a workspace_id
	format has a value which is a string
	auth has a value which is a string
fba_id is a string
workspace_id is a string</pre>

</dd>
<dt id="Description34">Description</dt>
<dd>

<p>Export an FBA solution for viewing</p>

</dd>
</dl>

<h2 id="import_phenotypes">import_phenotypes</h2>

<pre><code>  $output = $obj-&gt;import_phenotypes($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types35">Parameter and return types</dt>
<dd>



<pre>
$input is an import_phenotypes_params
$output is an object_metadata
import_phenotypes_params is a reference to a hash where the following keys are defined:
	phenotypeSet has a value which is a phenotype_set_id
	workspace has a value which is a workspace_id
	genome has a value which is a genome_id
	genome_workspace has a value which is a workspace_id
	phenotypes has a value which is a reference to a list where each element is a Phenotype
	name has a value which is a string
	source has a value which is a string
	ignore_errors has a value which is a bool
	auth has a value which is a string
phenotype_set_id is a string
workspace_id is a string
genome_id is a string
Phenotype is a reference to a list containing 6 items:
	0: (geneKO) a reference to a list where each element is a feature_id
	1: (baseMedia) a media_id
	2: (media_workspace) a workspace_id
	3: (additionalCpd) a reference to a list where each element is a compound_id
	4: (normalizedGrowth) a float
	5: (label) a string
feature_id is a string
media_id is a string
compound_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description35">Description</dt>
<dd>

<p>Loads the specified phenotypes into the workspace</p>

</dd>
</dl>

<h2 id="simulate_phenotypes">simulate_phenotypes</h2>

<pre><code>  $output = $obj-&gt;simulate_phenotypes($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types36">Parameter and return types</dt>
<dd>



<pre>
$input is a simulate_phenotypes_params
$output is an object_metadata
simulate_phenotypes_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	phenotypeSet has a value which is a phenotype_set_id
	phenotypeSet_workspace has a value which is a workspace_id
	formulation has a value which is an FBAFormulation
	notes has a value which is a string
	phenotypeSimulationSet has a value which is a phenotypeSimulationSet_id
	workspace has a value which is a workspace_id
	overwrite has a value which is a bool
	auth has a value which is a string
	all_transporters has a value which is a bool
	positive_transporters has a value which is a bool
fbamodel_id is a string
workspace_id is a string
phenotype_set_id is a string
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
bool is an int
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
phenotypeSimulationSet_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description36">Description</dt>
<dd>

<p>Simulates the specified phenotype set</p>

</dd>
</dl>

<h2 id="add_media_transporters">add_media_transporters</h2>

<pre><code>  $output = $obj-&gt;add_media_transporters($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types37">Parameter and return types</dt>
<dd>



<pre>
$input is an add_media_transporters_params
$output is an object_metadata
add_media_transporters_params is a reference to a hash where the following keys are defined:
	phenotypeSet has a value which is a phenotype_set_id
	phenotypeSet_workspace has a value which is a workspace_id
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	outmodel has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	overwrite has a value which is a bool
	auth has a value which is a string
	all_transporters has a value which is a bool
	positive_transporters has a value which is a bool
phenotype_set_id is a string
workspace_id is a string
fbamodel_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description37">Description</dt>
<dd>

<p>Adds transporters for media in a PhenotypeSet to a model</p>

</dd>
</dl>

<h2 id="export_phenotypeSimulationSet">export_phenotypeSimulationSet</h2>

<pre><code>  $output = $obj-&gt;export_phenotypeSimulationSet($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types38">Parameter and return types</dt>
<dd>



<pre>
$input is an export_phenotypeSimulationSet_params
$output is a string
export_phenotypeSimulationSet_params is a reference to a hash where the following keys are defined:
	phenotypeSimulationSet has a value which is a phenotypeSimulationSet_id
	workspace has a value which is a workspace_id
	format has a value which is a string
	auth has a value which is a string
phenotypeSimulationSet_id is a string
workspace_id is a string</pre>

</dd>
<dt id="Description38">Description</dt>
<dd>

<p>Export a PhenotypeSimulationSet for viewing</p>

</dd>
</dl>

<h2 id="integrate_reconciliation_solutions">integrate_reconciliation_solutions</h2>

<pre><code>  $modelMeta = $obj-&gt;integrate_reconciliation_solutions($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types39">Parameter and return types</dt>
<dd>



<pre>
$input is an integrate_reconciliation_solutions_params
$modelMeta is an object_metadata
integrate_reconciliation_solutions_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	gapfillSolutions has a value which is a reference to a list where each element is a gapfillsolution_id
	gapgenSolutions has a value which is a reference to a list where each element is a gapgensolution_id
	out_model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	auth has a value which is a string
	overwrite has a value which is a bool
fbamodel_id is a string
workspace_id is a string
gapfillsolution_id is a string
gapgensolution_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description39">Description</dt>
<dd>

<p>Integrates the specified gapfill and gapgen solutions into the specified model</p>

</dd>
</dl>

<h2 id="queue_runfba">queue_runfba</h2>

<pre><code>  $job = $obj-&gt;queue_runfba($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types40">Parameter and return types</dt>
<dd>



<pre>
$input is a queue_runfba_params
$job is a JobObject
queue_runfba_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	formulation has a value which is an FBAFormulation
	fva has a value which is a bool
	simulateko has a value which is a bool
	minimizeflux has a value which is a bool
	findminmedia has a value which is a bool
	notes has a value which is a string
	fba has a value which is a fba_id
	workspace has a value which is a workspace_id
	auth has a value which is a string
	overwrite has a value which is a bool
	add_to_model has a value which is a bool
fbamodel_id is a string
workspace_id is a string
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
bool is an int
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
fba_id is a string
JobObject is a reference to a hash where the following keys are defined:
	id has a value which is a job_id
	type has a value which is a string
	auth has a value which is a string
	status has a value which is a string
	jobdata has a value which is a reference to a hash where the key is a string and the value is a string
	queuetime has a value which is a string
	starttime has a value which is a string
	completetime has a value which is a string
	owner has a value which is a string
	queuecommand has a value which is a string
job_id is a string</pre>

</dd>
<dt id="Description40">Description</dt>
<dd>

<p>Queues an FBA job in a single media condition</p>

</dd>
</dl>

<h2 id="queue_gapfill_model">queue_gapfill_model</h2>

<pre><code>  $job = $obj-&gt;queue_gapfill_model($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types41">Parameter and return types</dt>
<dd>



<pre>
$input is a gapfill_model_params
$job is a JobObject
gapfill_model_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	formulation has a value which is a GapfillingFormulation
	phenotypeSet has a value which is a phenotype_set_id
	phenotypeSet_workspace has a value which is a workspace_id
	integrate_solution has a value which is a bool
	target_reactions has a value which is a reference to a list where each element is a string
	out_model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	gapFill has a value which is a gapfill_id
	timePerSolution has a value which is an int
	totalTimeLimit has a value which is an int
	auth has a value which is a string
	overwrite has a value which is a bool
	completeGapfill has a value which is a bool
fbamodel_id is a string
workspace_id is a string
GapfillingFormulation is a reference to a hash where the following keys are defined:
	formulation has a value which is an FBAFormulation
	num_solutions has a value which is an int
	nomediahyp has a value which is a bool
	nobiomasshyp has a value which is a bool
	nogprhyp has a value which is a bool
	nopathwayhyp has a value which is a bool
	allowunbalanced has a value which is a bool
	activitybonus has a value which is a float
	drainpen has a value which is a float
	directionpen has a value which is a float
	nostructpen has a value which is a float
	unfavorablepen has a value which is a float
	nodeltagpen has a value which is a float
	biomasstranspen has a value which is a float
	singletranspen has a value which is a float
	transpen has a value which is a float
	blacklistedrxns has a value which is a reference to a list where each element is a reaction_id
	gauranteedrxns has a value which is a reference to a list where each element is a reaction_id
	allowedcmps has a value which is a reference to a list where each element is a compartment_id
	probabilisticAnnotation has a value which is a probanno_id
	probabilisticAnnotation_workspace has a value which is a workspace_id
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
bool is an int
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
compartment_id is a string
probanno_id is a string
phenotype_set_id is a string
gapfill_id is a string
JobObject is a reference to a hash where the following keys are defined:
	id has a value which is a job_id
	type has a value which is a string
	auth has a value which is a string
	status has a value which is a string
	jobdata has a value which is a reference to a hash where the key is a string and the value is a string
	queuetime has a value which is a string
	starttime has a value which is a string
	completetime has a value which is a string
	owner has a value which is a string
	queuecommand has a value which is a string
job_id is a string</pre>

</dd>
<dt id="Description41">Description</dt>
<dd>

<p>Queues an FBAModel gapfilling job in single media condition</p>

</dd>
</dl>

<h2 id="gapfill_model">gapfill_model</h2>

<pre><code>  $modelMeta = $obj-&gt;gapfill_model($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types42">Parameter and return types</dt>
<dd>



<pre>
$input is a gapfill_model_params
$modelMeta is an object_metadata
gapfill_model_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	formulation has a value which is a GapfillingFormulation
	phenotypeSet has a value which is a phenotype_set_id
	phenotypeSet_workspace has a value which is a workspace_id
	integrate_solution has a value which is a bool
	target_reactions has a value which is a reference to a list where each element is a string
	out_model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	gapFill has a value which is a gapfill_id
	timePerSolution has a value which is an int
	totalTimeLimit has a value which is an int
	auth has a value which is a string
	overwrite has a value which is a bool
	completeGapfill has a value which is a bool
fbamodel_id is a string
workspace_id is a string
GapfillingFormulation is a reference to a hash where the following keys are defined:
	formulation has a value which is an FBAFormulation
	num_solutions has a value which is an int
	nomediahyp has a value which is a bool
	nobiomasshyp has a value which is a bool
	nogprhyp has a value which is a bool
	nopathwayhyp has a value which is a bool
	allowunbalanced has a value which is a bool
	activitybonus has a value which is a float
	drainpen has a value which is a float
	directionpen has a value which is a float
	nostructpen has a value which is a float
	unfavorablepen has a value which is a float
	nodeltagpen has a value which is a float
	biomasstranspen has a value which is a float
	singletranspen has a value which is a float
	transpen has a value which is a float
	blacklistedrxns has a value which is a reference to a list where each element is a reaction_id
	gauranteedrxns has a value which is a reference to a list where each element is a reaction_id
	allowedcmps has a value which is a reference to a list where each element is a compartment_id
	probabilisticAnnotation has a value which is a probanno_id
	probabilisticAnnotation_workspace has a value which is a workspace_id
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
bool is an int
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
compartment_id is a string
probanno_id is a string
phenotype_set_id is a string
gapfill_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description42">Description</dt>
<dd>

</dd>
</dl>

<h2 id="queue_gapgen_model">queue_gapgen_model</h2>

<pre><code>  $job = $obj-&gt;queue_gapgen_model($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types43">Parameter and return types</dt>
<dd>



<pre>
$input is a gapgen_model_params
$job is a JobObject
gapgen_model_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	formulation has a value which is a GapgenFormulation
	phenotypeSet has a value which is a phenotype_set_id
	phenotypeSet_workspace has a value which is a workspace_id
	integrate_solution has a value which is a bool
	out_model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	gapGen has a value which is a gapgen_id
	auth has a value which is a string
	timePerSolution has a value which is an int
	totalTimeLimit has a value which is an int
	overwrite has a value which is a bool
fbamodel_id is a string
workspace_id is a string
GapgenFormulation is a reference to a hash where the following keys are defined:
	formulation has a value which is an FBAFormulation
	refmedia has a value which is a media_id
	refmedia_workspace has a value which is a workspace_id
	num_solutions has a value which is an int
	nomediahyp has a value which is a bool
	nobiomasshyp has a value which is a bool
	nogprhyp has a value which is a bool
	nopathwayhyp has a value which is a bool
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
bool is an int
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
phenotype_set_id is a string
gapgen_id is a string
JobObject is a reference to a hash where the following keys are defined:
	id has a value which is a job_id
	type has a value which is a string
	auth has a value which is a string
	status has a value which is a string
	jobdata has a value which is a reference to a hash where the key is a string and the value is a string
	queuetime has a value which is a string
	starttime has a value which is a string
	completetime has a value which is a string
	owner has a value which is a string
	queuecommand has a value which is a string
job_id is a string</pre>

</dd>
<dt id="Description43">Description</dt>
<dd>

<p>Queues an FBAModel gapfilling job in single media condition</p>

</dd>
</dl>

<h2 id="gapgen_model">gapgen_model</h2>

<pre><code>  $modelMeta = $obj-&gt;gapgen_model($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types44">Parameter and return types</dt>
<dd>



<pre>
$input is a gapgen_model_params
$modelMeta is an object_metadata
gapgen_model_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	formulation has a value which is a GapgenFormulation
	phenotypeSet has a value which is a phenotype_set_id
	phenotypeSet_workspace has a value which is a workspace_id
	integrate_solution has a value which is a bool
	out_model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	gapGen has a value which is a gapgen_id
	auth has a value which is a string
	timePerSolution has a value which is an int
	totalTimeLimit has a value which is an int
	overwrite has a value which is a bool
fbamodel_id is a string
workspace_id is a string
GapgenFormulation is a reference to a hash where the following keys are defined:
	formulation has a value which is an FBAFormulation
	refmedia has a value which is a media_id
	refmedia_workspace has a value which is a workspace_id
	num_solutions has a value which is an int
	nomediahyp has a value which is a bool
	nobiomasshyp has a value which is a bool
	nogprhyp has a value which is a bool
	nopathwayhyp has a value which is a bool
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
bool is an int
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
phenotype_set_id is a string
gapgen_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description44">Description</dt>
<dd>

</dd>
</dl>

<h2 id="queue_wildtype_phenotype_reconciliation">queue_wildtype_phenotype_reconciliation</h2>

<pre><code>  $job = $obj-&gt;queue_wildtype_phenotype_reconciliation($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types45">Parameter and return types</dt>
<dd>



<pre>
$input is a wildtype_phenotype_reconciliation_params
$job is a JobObject
wildtype_phenotype_reconciliation_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	fba_formulation has a value which is an FBAFormulation
	gapfill_formulation has a value which is a GapfillingFormulation
	gapgen_formulation has a value which is a GapgenFormulation
	phenotypeSet has a value which is a phenotype_set_id
	phenotypeSet_workspace has a value which is a workspace_id
	out_model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	gapFills has a value which is a reference to a list where each element is a gapfill_id
	gapGens has a value which is a reference to a list where each element is a gapgen_id
	queueSensitivityAnalysis has a value which is a bool
	queueReconciliationCombination has a value which is a bool
	auth has a value which is a string
	overwrite has a value which is a bool
fbamodel_id is a string
workspace_id is a string
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
bool is an int
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
GapfillingFormulation is a reference to a hash where the following keys are defined:
	formulation has a value which is an FBAFormulation
	num_solutions has a value which is an int
	nomediahyp has a value which is a bool
	nobiomasshyp has a value which is a bool
	nogprhyp has a value which is a bool
	nopathwayhyp has a value which is a bool
	allowunbalanced has a value which is a bool
	activitybonus has a value which is a float
	drainpen has a value which is a float
	directionpen has a value which is a float
	nostructpen has a value which is a float
	unfavorablepen has a value which is a float
	nodeltagpen has a value which is a float
	biomasstranspen has a value which is a float
	singletranspen has a value which is a float
	transpen has a value which is a float
	blacklistedrxns has a value which is a reference to a list where each element is a reaction_id
	gauranteedrxns has a value which is a reference to a list where each element is a reaction_id
	allowedcmps has a value which is a reference to a list where each element is a compartment_id
	probabilisticAnnotation has a value which is a probanno_id
	probabilisticAnnotation_workspace has a value which is a workspace_id
compartment_id is a string
probanno_id is a string
GapgenFormulation is a reference to a hash where the following keys are defined:
	formulation has a value which is an FBAFormulation
	refmedia has a value which is a media_id
	refmedia_workspace has a value which is a workspace_id
	num_solutions has a value which is an int
	nomediahyp has a value which is a bool
	nobiomasshyp has a value which is a bool
	nogprhyp has a value which is a bool
	nopathwayhyp has a value which is a bool
phenotype_set_id is a string
gapfill_id is a string
gapgen_id is a string
JobObject is a reference to a hash where the following keys are defined:
	id has a value which is a job_id
	type has a value which is a string
	auth has a value which is a string
	status has a value which is a string
	jobdata has a value which is a reference to a hash where the key is a string and the value is a string
	queuetime has a value which is a string
	starttime has a value which is a string
	completetime has a value which is a string
	owner has a value which is a string
	queuecommand has a value which is a string
job_id is a string</pre>

</dd>
<dt id="Description45">Description</dt>
<dd>

<p>Queues an FBAModel reconciliation job</p>

</dd>
</dl>

<h2 id="queue_reconciliation_sensitivity_analysis">queue_reconciliation_sensitivity_analysis</h2>

<pre><code>  $job = $obj-&gt;queue_reconciliation_sensitivity_analysis($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types46">Parameter and return types</dt>
<dd>



<pre>
$input is a wildtype_phenotype_reconciliation_params
$job is a JobObject
wildtype_phenotype_reconciliation_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	fba_formulation has a value which is an FBAFormulation
	gapfill_formulation has a value which is a GapfillingFormulation
	gapgen_formulation has a value which is a GapgenFormulation
	phenotypeSet has a value which is a phenotype_set_id
	phenotypeSet_workspace has a value which is a workspace_id
	out_model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	gapFills has a value which is a reference to a list where each element is a gapfill_id
	gapGens has a value which is a reference to a list where each element is a gapgen_id
	queueSensitivityAnalysis has a value which is a bool
	queueReconciliationCombination has a value which is a bool
	auth has a value which is a string
	overwrite has a value which is a bool
fbamodel_id is a string
workspace_id is a string
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
bool is an int
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
GapfillingFormulation is a reference to a hash where the following keys are defined:
	formulation has a value which is an FBAFormulation
	num_solutions has a value which is an int
	nomediahyp has a value which is a bool
	nobiomasshyp has a value which is a bool
	nogprhyp has a value which is a bool
	nopathwayhyp has a value which is a bool
	allowunbalanced has a value which is a bool
	activitybonus has a value which is a float
	drainpen has a value which is a float
	directionpen has a value which is a float
	nostructpen has a value which is a float
	unfavorablepen has a value which is a float
	nodeltagpen has a value which is a float
	biomasstranspen has a value which is a float
	singletranspen has a value which is a float
	transpen has a value which is a float
	blacklistedrxns has a value which is a reference to a list where each element is a reaction_id
	gauranteedrxns has a value which is a reference to a list where each element is a reaction_id
	allowedcmps has a value which is a reference to a list where each element is a compartment_id
	probabilisticAnnotation has a value which is a probanno_id
	probabilisticAnnotation_workspace has a value which is a workspace_id
compartment_id is a string
probanno_id is a string
GapgenFormulation is a reference to a hash where the following keys are defined:
	formulation has a value which is an FBAFormulation
	refmedia has a value which is a media_id
	refmedia_workspace has a value which is a workspace_id
	num_solutions has a value which is an int
	nomediahyp has a value which is a bool
	nobiomasshyp has a value which is a bool
	nogprhyp has a value which is a bool
	nopathwayhyp has a value which is a bool
phenotype_set_id is a string
gapfill_id is a string
gapgen_id is a string
JobObject is a reference to a hash where the following keys are defined:
	id has a value which is a job_id
	type has a value which is a string
	auth has a value which is a string
	status has a value which is a string
	jobdata has a value which is a reference to a hash where the key is a string and the value is a string
	queuetime has a value which is a string
	starttime has a value which is a string
	completetime has a value which is a string
	owner has a value which is a string
	queuecommand has a value which is a string
job_id is a string</pre>

</dd>
<dt id="Description46">Description</dt>
<dd>

<p>Queues an FBAModel reconciliation job</p>

</dd>
</dl>

<h2 id="queue_combine_wildtype_phenotype_reconciliation">queue_combine_wildtype_phenotype_reconciliation</h2>

<pre><code>  $job = $obj-&gt;queue_combine_wildtype_phenotype_reconciliation($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types47">Parameter and return types</dt>
<dd>



<pre>
$input is a combine_wildtype_phenotype_reconciliation_params
$job is a JobObject
combine_wildtype_phenotype_reconciliation_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_workspace has a value which is a workspace_id
	fba_formulation has a value which is an FBAFormulation
	gapfill_formulation has a value which is a GapfillingFormulation
	gapgen_formulation has a value which is a GapgenFormulation
	phenotypeSet has a value which is a phenotype_set_id
	phenotypeSet_workspace has a value which is a workspace_id
	out_model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	gapFills has a value which is a reference to a list where each element is a gapfill_id
	gapGens has a value which is a reference to a list where each element is a gapgen_id
	auth has a value which is a string
	overwrite has a value which is a bool
fbamodel_id is a string
workspace_id is a string
FBAFormulation is a reference to a hash where the following keys are defined:
	media has a value which is a media_id
	additionalcpds has a value which is a reference to a list where each element is a compound_id
	promconstraint has a value which is a promconstraint_id
	promconstraint_workspace has a value which is a workspace_id
	eflux_sample has a value which is a sample_id
	eflux_series has a value which is a series_id
	eflux_workspace has a value which is a workspace_id
	media_workspace has a value which is a workspace_id
	objfraction has a value which is a float
	allreversible has a value which is a bool
	maximizeObjective has a value which is a bool
	objectiveTerms has a value which is a reference to a list where each element is a term
	geneko has a value which is a reference to a list where each element is a feature_id
	rxnko has a value which is a reference to a list where each element is a reaction_id
	bounds has a value which is a reference to a list where each element is a bound
	constraints has a value which is a reference to a list where each element is a constraint
	uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
	defaultmaxflux has a value which is a float
	defaultminuptake has a value which is a float
	defaultmaxuptake has a value which is a float
	simplethermoconst has a value which is a bool
	thermoconst has a value which is a bool
	nothermoerror has a value which is a bool
	minthermoerror has a value which is a bool
media_id is a string
compound_id is a string
promconstraint_id is a string
sample_id is a string
series_id is a string
bool is an int
term is a reference to a list containing 3 items:
	0: (coefficient) a float
	1: (varType) a string
	2: (variable) a string
feature_id is a string
reaction_id is a string
bound is a reference to a list containing 4 items:
	0: (min) a float
	1: (max) a float
	2: (varType) a string
	3: (variable) a string
constraint is a reference to a list containing 4 items:
	0: (rhs) a float
	1: (sign) a string
	2: (terms) a reference to a list where each element is a term
	3: (name) a string
GapfillingFormulation is a reference to a hash where the following keys are defined:
	formulation has a value which is an FBAFormulation
	num_solutions has a value which is an int
	nomediahyp has a value which is a bool
	nobiomasshyp has a value which is a bool
	nogprhyp has a value which is a bool
	nopathwayhyp has a value which is a bool
	allowunbalanced has a value which is a bool
	activitybonus has a value which is a float
	drainpen has a value which is a float
	directionpen has a value which is a float
	nostructpen has a value which is a float
	unfavorablepen has a value which is a float
	nodeltagpen has a value which is a float
	biomasstranspen has a value which is a float
	singletranspen has a value which is a float
	transpen has a value which is a float
	blacklistedrxns has a value which is a reference to a list where each element is a reaction_id
	gauranteedrxns has a value which is a reference to a list where each element is a reaction_id
	allowedcmps has a value which is a reference to a list where each element is a compartment_id
	probabilisticAnnotation has a value which is a probanno_id
	probabilisticAnnotation_workspace has a value which is a workspace_id
compartment_id is a string
probanno_id is a string
GapgenFormulation is a reference to a hash where the following keys are defined:
	formulation has a value which is an FBAFormulation
	refmedia has a value which is a media_id
	refmedia_workspace has a value which is a workspace_id
	num_solutions has a value which is an int
	nomediahyp has a value which is a bool
	nobiomasshyp has a value which is a bool
	nogprhyp has a value which is a bool
	nopathwayhyp has a value which is a bool
phenotype_set_id is a string
gapfill_id is a string
gapgen_id is a string
JobObject is a reference to a hash where the following keys are defined:
	id has a value which is a job_id
	type has a value which is a string
	auth has a value which is a string
	status has a value which is a string
	jobdata has a value which is a reference to a hash where the key is a string and the value is a string
	queuetime has a value which is a string
	starttime has a value which is a string
	completetime has a value which is a string
	owner has a value which is a string
	queuecommand has a value which is a string
job_id is a string</pre>

</dd>
<dt id="Description47">Description</dt>
<dd>

<p>Queues an FBAModel reconciliation job</p>

</dd>
</dl>

<h2 id="run_job">run_job</h2>

<pre><code>  $job = $obj-&gt;run_job($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types48">Parameter and return types</dt>
<dd>



<pre>
$input is a run_job_params
$job is a JobObject
run_job_params is a reference to a hash where the following keys are defined:
	job has a value which is a job_id
	auth has a value which is a string
job_id is a string
JobObject is a reference to a hash where the following keys are defined:
	id has a value which is a job_id
	type has a value which is a string
	auth has a value which is a string
	status has a value which is a string
	jobdata has a value which is a reference to a hash where the key is a string and the value is a string
	queuetime has a value which is a string
	starttime has a value which is a string
	completetime has a value which is a string
	owner has a value which is a string
	queuecommand has a value which is a string</pre>

</dd>
<dt id="Description48">Description</dt>
<dd>

<p>Runs specified job</p>

</dd>
</dl>

<h2 id="queue_job">queue_job</h2>

<pre><code>  $job = $obj-&gt;queue_job($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types49">Parameter and return types</dt>
<dd>



<pre>
$input is a queue_job_params
$job is a JobObject
queue_job_params is a reference to a hash where the following keys are defined:
	method has a value which is a string
	parameters has a value which is a reference to a hash where the key is a string and the value is a string
JobObject is a reference to a hash where the following keys are defined:
	id has a value which is a job_id
	type has a value which is a string
	auth has a value which is a string
	status has a value which is a string
	jobdata has a value which is a reference to a hash where the key is a string and the value is a string
	queuetime has a value which is a string
	starttime has a value which is a string
	completetime has a value which is a string
	owner has a value which is a string
	queuecommand has a value which is a string
job_id is a string</pre>

</dd>
<dt id="Description49">Description</dt>
<dd>

<p>Queues the specified command to run as a job</p>

</dd>
</dl>

<h2 id="set_cofactors">set_cofactors</h2>

<pre><code>  $output = $obj-&gt;set_cofactors($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types50">Parameter and return types</dt>
<dd>



<pre>
$input is a set_cofactors_params
$output is an object_metadata
set_cofactors_params is a reference to a hash where the following keys are defined:
	cofactors has a value which is a reference to a list where each element is a compound_id
	biochemistry has a value which is a biochemistry_id
	biochemistry_workspace has a value which is a workspace_id
	reset has a value which is a bool
	overwrite has a value which is a bool
	auth has a value which is a string
compound_id is a string
biochemistry_id is a string
workspace_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description50">Description</dt>
<dd>

</dd>
</dl>

<h2 id="find_reaction_synonyms">find_reaction_synonyms</h2>

<pre><code>  $output = $obj-&gt;find_reaction_synonyms($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types51">Parameter and return types</dt>
<dd>



<pre>
$input is a find_reaction_synonyms_params
$output is an object_metadata
find_reaction_synonyms_params is a reference to a hash where the following keys are defined:
	reaction_synonyms has a value which is a reaction_synonyms_id
	workspace has a value which is a workspace_id
	biochemistry has a value which is a biochemistry_id
	biochemistry_workspace has a value which is a workspace_id
	overwrite has a value which is a bool
	auth has a value which is a string
reaction_synonyms_id is a string
workspace_id is a string
biochemistry_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description51">Description</dt>
<dd>

</dd>
</dl>

<h2 id="role_to_reactions">role_to_reactions</h2>

<pre><code>  $output = $obj-&gt;role_to_reactions($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types52">Parameter and return types</dt>
<dd>



<pre>
$params is a role_to_reactions_params
$output is a reference to a list where each element is a RoleComplexReactions
role_to_reactions_params is a reference to a hash where the following keys are defined:
	templateModel has a value which is a template_id
	workspace has a value which is a workspace_id
	auth has a value which is a string
template_id is a string
workspace_id is a string
RoleComplexReactions is a reference to a hash where the following keys are defined:
	role has a value which is a role_id
	name has a value which is a string
	complexes has a value which is a reference to a list where each element is a ComplexReactions
role_id is a string
ComplexReactions is a reference to a hash where the following keys are defined:
	complex has a value which is a complex_id
	name has a value which is a string
	reactions has a value which is a reference to a list where each element is a TemplateReactions
complex_id is a string
TemplateReactions is a reference to a hash where the following keys are defined:
	reaction has a value which is a reaction_id
	direction has a value which is a string
	equation has a value which is a string
	compartment has a value which is a compartment_id
reaction_id is a string
compartment_id is a string</pre>

</dd>
<dt id="Description52">Description</dt>
<dd>

<p>Retrieves a list of roles mapped to reactions based on input template model</p>

</dd>
</dl>

<h2 id="reaction_sensitivity_analysis">reaction_sensitivity_analysis</h2>

<pre><code>  $output = $obj-&gt;reaction_sensitivity_analysis($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types53">Parameter and return types</dt>
<dd>



<pre>
$input is a reaction_sensitivity_analysis_params
$output is an object_metadata
reaction_sensitivity_analysis_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	model_ws has a value which is a workspace_id
	rxnsens_uid has a value which is a string
	workspace has a value which is a workspace_id
	reactions_to_delete has a value which is a reference to a list where each element is a reaction_id
	gapfill_solution_id has a value which is a gapfillsolution_id
	delete_noncontributing_reactions has a value which is a bool
	rxnprobs_id has a value which is a rxnprob_id
	rxnprobs_ws has a value which is a workspace_id
	type has a value which is a string
	auth has a value which is a string
fbamodel_id is a string
workspace_id is a string
reaction_id is a string
gapfillsolution_id is a string
bool is an int
rxnprob_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description53">Description</dt>
<dd>

<p>Queues a sensitivity analysis on the knockout of model reactions</p>

</dd>
</dl>

<h2 id="filter_iterative_solutions">filter_iterative_solutions</h2>

<pre><code>  $output = $obj-&gt;filter_iterative_solutions($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types54">Parameter and return types</dt>
<dd>



<pre>
$input is a filter_iterative_solutions_params
$output is an object_metadata
filter_iterative_solutions_params is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	outmodel has a value which is a fbamodel_id
	cutoff has a value which is a float
	gapfillsln has a value which is a gapfillsolution_id
	workspace has a value which is a workspace_id
	input_model_ws has a value which is a workspace_id
	auth has a value which is a string
fbamodel_id is a string
gapfillsolution_id is a string
workspace_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description54">Description</dt>
<dd>

<p>Apply a cutoff to remove high-cost iterations from an iterative gapfill run.</p>

</dd>
</dl>

<h2 id="delete_noncontributing_reactions">delete_noncontributing_reactions</h2>

<pre><code>  $output = $obj-&gt;delete_noncontributing_reactions($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types55">Parameter and return types</dt>
<dd>



<pre>
$input is a delete_noncontributing_reactions_params
$output is an object_metadata
delete_noncontributing_reactions_params is a reference to a hash where the following keys are defined:
	rxn_sensitivity_ws has a value which is a workspace_id
	rxn_sensitivity has a value which is a string
	workspace has a value which is a workspace_id
	new_model_uid has a value which is a fbamodel_id
	new_rxn_sensitivity_uid has a value which is a string
	auth has a value which is a string
workspace_id is a string
fbamodel_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description55">Description</dt>
<dd>

<p>Deleted flagged reactions from a RxnSensitivity object</p>

</dd>
</dl>

<h2 id="annotate_workspace_Genome">annotate_workspace_Genome</h2>

<pre><code>  $output = $obj-&gt;annotate_workspace_Genome($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types56">Parameter and return types</dt>
<dd>



<pre>
$params is an annotate_workspace_Genome_params
$output is an object_metadata
annotate_workspace_Genome_params is a reference to a hash where the following keys are defined:
	Genome_uid has a value which is a string
	Genome_ws has a value which is a string
	new_uid has a value which is a string
	workspace has a value which is a workspace_id
	annotation_parameters has a value which is an AnnotationParameters
	auth has a value which is a string
workspace_id is a string
AnnotationParameters is a reference to a hash where the following keys are defined:
	call_genes has a value which is a bool
	annotate_genes has a value which is a bool
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description56">Description</dt>
<dd>

<p>Create a job that runs the genome annotation pipeline on a genome object in a workspace</p>

</dd>
</dl>

<h2 id="gtf_to_genome">gtf_to_genome</h2>

<pre><code>  $output = $obj-&gt;gtf_to_genome($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types57">Parameter and return types</dt>
<dd>



<pre>
$params is a gtf_to_genome_params
$output is an object_metadata
gtf_to_genome_params is a reference to a hash where the following keys are defined:
	contigset has a value which is a string
	contigset_ws has a value which is a workspace_id
	workspace has a value which is a workspace_id
	genome_uid has a value which is a string
	source_id has a value which is a string
	source has a value which is a string
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	taxonomy has a value which is a string
	gtf_file has a value which is a string
workspace_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description57">Description</dt>
<dd>

<p>Loads a gtf file to a genome typed object in the workspace</p>

</dd>
</dl>

<h2 id="fasta_to_ProteinSet">fasta_to_ProteinSet</h2>

<pre><code>  $output = $obj-&gt;fasta_to_ProteinSet($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types58">Parameter and return types</dt>
<dd>



<pre>
$params is a fasta_to_ProteinSet_params
$output is an object_metadata
fasta_to_ProteinSet_params is a reference to a hash where the following keys are defined:
	uid has a value which is a string
	fasta has a value which is a string
	workspace has a value which is a workspace_id
	auth has a value which is a string
	name has a value which is a string
	sourceid has a value which is a string
	source has a value which is a string
	type has a value which is a string
workspace_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description58">Description</dt>
<dd>

<p>Loads a fasta file as a ProteinSet object in the workspace</p>

</dd>
</dl>

<h2 id="ProteinSet_to_Genome">ProteinSet_to_Genome</h2>

<pre><code>  $output = $obj-&gt;ProteinSet_to_Genome($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types59">Parameter and return types</dt>
<dd>



<pre>
$params is a ProteinSet_to_Genome_params
$output is an object_metadata
ProteinSet_to_Genome_params is a reference to a hash where the following keys are defined:
	ProteinSet_uid has a value which is a string
	ProteinSet_ws has a value which is a workspace_id
	workspace has a value which is a workspace_id
	uid has a value which is a string
	auth has a value which is a string
	scientific_name has a value which is a string
	domain has a value which is a string
	annotation_parameters has a value which is an AnnotationParameters
workspace_id is a string
AnnotationParameters is a reference to a hash where the following keys are defined:
	call_genes has a value which is a bool
	annotate_genes has a value which is a bool
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description59">Description</dt>
<dd>

<p>Creates a Genome associated with the ProteinSet object. You cannot recall genes on this genome.</p>

</dd>
</dl>

<h2 id="fasta_to_ContigSet">fasta_to_ContigSet</h2>

<pre><code>  $output = $obj-&gt;fasta_to_ContigSet($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types60">Parameter and return types</dt>
<dd>



<pre>
$params is a fasta_to_ContigSet_params
$output is an object_metadata
fasta_to_ContigSet_params is a reference to a hash where the following keys are defined:
	uid has a value which is a string
	fasta has a value which is a string
	workspace has a value which is a workspace_id
	auth has a value which is a string
	name has a value which is a string
	sourceid has a value which is a string
	source has a value which is a string
	type has a value which is a string
workspace_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description60">Description</dt>
<dd>

<p>Loads a fasta file as a ContigSet object in the workspace</p>

</dd>
</dl>

<h2 id="ContigSet_to_Genome">ContigSet_to_Genome</h2>

<pre><code>  $output = $obj-&gt;ContigSet_to_Genome($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types61">Parameter and return types</dt>
<dd>



<pre>
$params is a ContigSet_to_Genome_params
$output is an object_metadata
ContigSet_to_Genome_params is a reference to a hash where the following keys are defined:
	ContigSet_uid has a value which is a string
	ContigSet_ws has a value which is a workspace_id
	workspace has a value which is a workspace_id
	uid has a value which is a string
	auth has a value which is a string
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	annotation_parameters has a value which is an AnnotationParameters
workspace_id is a string
AnnotationParameters is a reference to a hash where the following keys are defined:
	call_genes has a value which is a bool
	annotate_genes has a value which is a bool
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description61">Description</dt>
<dd>

<p>Creates a genome associated with the ContigSet object</p>

</dd>
</dl>

<h2 id="probanno_to_genome">probanno_to_genome</h2>

<pre><code>  $output = $obj-&gt;probanno_to_genome($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types62">Parameter and return types</dt>
<dd>



<pre>
$params is a probanno_to_genome_params
$output is an object_metadata
probanno_to_genome_params is a reference to a hash where the following keys are defined:
	pa_id has a value which is a probanno_id
	pa_ws has a value which is a workspace_id
	workspace has a value which is a workspace_id
	g_id has a value which is a genome_id
	threshold has a value which is a float
	auth has a value which is a string
probanno_id is a string
workspace_id is a string
genome_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description62">Description</dt>
<dd>

<p>Converts a probabilistic annotation into a genome with the same annotations</p>

</dd>
</dl>

<h2 id="get_mapping">get_mapping</h2>

<pre><code>  $output = $obj-&gt;get_mapping($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types63">Parameter and return types</dt>
<dd>



<pre>
$params is a get_mapping_params
$output is a Mapping
get_mapping_params is a reference to a hash where the following keys are defined:
	map has a value which is a mapping_id
	workspace has a value which is a workspace_id
	auth has a value which is a string
mapping_id is a string
workspace_id is a string
Mapping is a reference to a hash where the following keys are defined:
	id has a value which is a mapping_id
	name has a value which is a string
	subsystems has a value which is a reference to a list where each element is a Subsystem
	roles has a value which is a reference to a list where each element is a FunctionalRole
	complexes has a value which is a reference to a list where each element is a Complex
Subsystem is a reference to a hash where the following keys are defined:
	id has a value which is a subsystem_id
	name has a value which is a string
	phenoclass has a value which is a string
	subclass has a value which is a string
	type has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	roles has a value which is a reference to a list where each element is a role_id
subsystem_id is a string
role_id is a string
FunctionalRole is a reference to a hash where the following keys are defined:
	id has a value which is a role_id
	name has a value which is a string
	feature has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	complexes has a value which is a reference to a list where each element is a complex_id
complex_id is a string
Complex is a reference to a hash where the following keys are defined:
	id has a value which is a complex_id
	name has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	roles has a value which is a reference to a list where each element is a ComplexRole
ComplexRole is a reference to a list containing 4 items:
	0: (id) a role_id
	1: (roleType) a string
	2: (optional_role) a bool
	3: (triggering) a bool
bool is an int</pre>

</dd>
<dt id="Description63">Description</dt>
<dd>

<p>Annotates contigs object creating a genome object</p>

</dd>
</dl>

<h2 id="subsystem_of_roles">subsystem_of_roles</h2>

<pre><code>  $output = $obj-&gt;subsystem_of_roles($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types64">Parameter and return types</dt>
<dd>



<pre>
$params is a subsystem_of_roles_params
$output is a reference to a hash where the key is a string and the value is a subsysclasses
subsystem_of_roles_params is a reference to a hash where the following keys are defined:
	roles has a value which is a reference to a list where each element is a string
	map has a value which is a string
	map_workspace has a value which is a string
subsysclasses is a reference to a hash where the key is a string and the value is a subsysclass
subsysclass is a reference to a list containing 2 items:
	0: a string
	1: a string</pre>

</dd>
<dt id="Description64">Description</dt>
<dd>

<p>Returns subsystems for list roles</p>

</dd>
</dl>

<h2 id="adjust_mapping_role">adjust_mapping_role</h2>

<pre><code>  $output = $obj-&gt;adjust_mapping_role($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types65">Parameter and return types</dt>
<dd>



<pre>
$params is an adjust_mapping_role_params
$output is a FunctionalRole
adjust_mapping_role_params is a reference to a hash where the following keys are defined:
	map has a value which is a mapping_id
	workspace has a value which is a workspace_id
	role has a value which is a string
	new has a value which is a bool
	name has a value which is a string
	feature has a value which is a string
	aliasesToAdd has a value which is a reference to a list where each element is a string
	aliasesToRemove has a value which is a reference to a list where each element is a string
	delete has a value which is a bool
	auth has a value which is a string
mapping_id is a string
workspace_id is a string
bool is an int
FunctionalRole is a reference to a hash where the following keys are defined:
	id has a value which is a role_id
	name has a value which is a string
	feature has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	complexes has a value which is a reference to a list where each element is a complex_id
role_id is a string
complex_id is a string</pre>

</dd>
<dt id="Description65">Description</dt>
<dd>

<p>An API function supporting the curation of functional roles in a mapping object</p>

</dd>
</dl>

<h2 id="adjust_mapping_complex">adjust_mapping_complex</h2>

<pre><code>  $output = $obj-&gt;adjust_mapping_complex($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types66">Parameter and return types</dt>
<dd>



<pre>
$params is an adjust_mapping_complex_params
$output is a Complex
adjust_mapping_complex_params is a reference to a hash where the following keys are defined:
	map has a value which is a mapping_id
	workspace has a value which is a workspace_id
	complex has a value which is a string
	new has a value which is a bool
	name has a value which is a string
	rolesToAdd has a value which is a reference to a list where each element is a string
	rolesToRemove has a value which is a reference to a list where each element is a string
	delete has a value which is a bool
	auth has a value which is a string
mapping_id is a string
workspace_id is a string
bool is an int
Complex is a reference to a hash where the following keys are defined:
	id has a value which is a complex_id
	name has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	roles has a value which is a reference to a list where each element is a ComplexRole
complex_id is a string
ComplexRole is a reference to a list containing 4 items:
	0: (id) a role_id
	1: (roleType) a string
	2: (optional_role) a bool
	3: (triggering) a bool
role_id is a string</pre>

</dd>
<dt id="Description66">Description</dt>
<dd>

<p>An API function supporting the curation of complexes in a mapping object</p>

</dd>
</dl>

<h2 id="adjust_mapping_subsystem">adjust_mapping_subsystem</h2>

<pre><code>  $output = $obj-&gt;adjust_mapping_subsystem($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types67">Parameter and return types</dt>
<dd>



<pre>
$params is an adjust_mapping_subsystem_params
$output is a Subsystem
adjust_mapping_subsystem_params is a reference to a hash where the following keys are defined:
	map has a value which is a mapping_id
	workspace has a value which is a workspace_id
	subsystem has a value which is a string
	new has a value which is a bool
	name has a value which is a string
	type has a value which is a string
	primclass has a value which is a string
	subclass has a value which is a string
	rolesToAdd has a value which is a reference to a list where each element is a string
	rolesToRemove has a value which is a reference to a list where each element is a string
	delete has a value which is a bool
	auth has a value which is a string
mapping_id is a string
workspace_id is a string
bool is an int
Subsystem is a reference to a hash where the following keys are defined:
	id has a value which is a subsystem_id
	name has a value which is a string
	phenoclass has a value which is a string
	subclass has a value which is a string
	type has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	roles has a value which is a reference to a list where each element is a role_id
subsystem_id is a string
role_id is a string</pre>

</dd>
<dt id="Description67">Description</dt>
<dd>

<p>An API function supporting the curation of subsystems in a mapping object</p>

</dd>
</dl>

<h2 id="get_template_model">get_template_model</h2>

<pre><code>  $output = $obj-&gt;get_template_model($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types68">Parameter and return types</dt>
<dd>



<pre>
$params is a get_template_model_params
$output is a TemplateModel
get_template_model_params is a reference to a hash where the following keys are defined:
	templateModel has a value which is a template_id
	workspace has a value which is a workspace_id
	auth has a value which is a string
template_id is a string
workspace_id is a string
TemplateModel is a reference to a hash where the following keys are defined:
	id has a value which is a template_id
	name has a value which is a string
	type has a value which is a string
	domain has a value which is a string
	map has a value which is a mapping_id
	mappingws has a value which is a workspace_id
	mapping_ref has a value which is a string
	biochemistry_ref has a value which is a string
	reactions has a value which is a reference to a list where each element is a TemplateReaction
	biomasses has a value which is a reference to a list where each element is a TemplateBiomass
mapping_id is a string
TemplateReaction is a reference to a hash where the following keys are defined:
	id has a value which is a temprxn_id
	compartment has a value which is a compartment_id
	reaction has a value which is a reaction_id
	complexes has a value which is a reference to a list where each element is a complex_id
	direction has a value which is a string
	type has a value which is a string
temprxn_id is a string
compartment_id is a string
reaction_id is a string
complex_id is a string
TemplateBiomass is a reference to a hash where the following keys are defined:
	id has a value which is a tempbiomass_id
	name has a value which is a string
	type has a value which is a string
	other has a value which is a string
	protein has a value which is a string
	dna has a value which is a string
	rna has a value which is a string
	cofactor has a value which is a string
	energy has a value which is a string
	cellwall has a value which is a string
	lipid has a value which is a string
	compounds has a value which is a reference to a list where each element is a TemplateBiomassCompounds
tempbiomass_id is a string
TemplateBiomassCompounds is a reference to a list containing 7 items:
	0: (compound) a compound_id
	1: (compartment) a compartment_id
	2: (cpdclass) a string
	3: (universal) a string
	4: (coefficientType) a string
	5: (coefficient) a string
	6: (linkedCompounds) a reference to a list where each element is a reference to a list containing 2 items:
		0: (coeffficient) a string
		1: (compound) a compound_idcompound_id is a string</pre>

</dd>
<dt id="Description68">Description</dt>
<dd>

<p>Retrieves the specified template model</p>

</dd>
</dl>

<h2 id="import_template_fbamodel">import_template_fbamodel</h2>

<pre><code>  $modelMeta = $obj-&gt;import_template_fbamodel($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types69">Parameter and return types</dt>
<dd>



<pre>
$input is an import_template_fbamodel_params
$modelMeta is an object_metadata
import_template_fbamodel_params is a reference to a hash where the following keys are defined:
	map has a value which is a mapping_id
	mapping_workspace has a value which is a workspace_id
	templateReactions has a value which is a reference to a list where each element is a reference to a list containing 5 items:
	0: (id) a string
	1: (compartment) a string
	2: (direction) a string
	3: (type) a string
	4: (complexes) a reference to a list where each element is a string	templateBiomass has a value which is a reference to a list where each element is a reference to a list containing 11 items:
	0: (name) a string
	1: (type) a string
	2: (dna) a float
	3: (rna) a float
	4: (protein) a float
	5: (lipid) a float
	6: (cellwall) a float
	7: (cofactor) a float
	8: (energy) a float
	9: (other) a float
	10: (compounds) a reference to a list where each element is a reference to a list containing 6 items:
		0: (id) a string
		1: (compartment) a string
		2: (cpdclass) a string
		3: (coefficientType) a string
		4: (coefficient) a float
		5: (conditions) a string


	name has a value which is a string
	modelType has a value which is a string
	domain has a value which is a string
	id has a value which is a template_id
	workspace has a value which is a workspace_id
	ignore_errors has a value which is a bool
	auth has a value which is a string
mapping_id is a string
workspace_id is a string
template_id is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string
</pre>

</dd>
<dt id="Description69">Description</dt>
<dd>

<p>Import a template model from an input table of template reactions and biomass components</p>

</dd>
</dl>

<h2 id="adjust_template_reaction">adjust_template_reaction</h2>

<pre><code>  $modelMeta = $obj-&gt;adjust_template_reaction($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types70">Parameter and return types</dt>
<dd>



<pre>
$params is an adjust_template_reaction_params
$modelMeta is an object_metadata
adjust_template_reaction_params is a reference to a hash where the following keys are defined:
	templateModel has a value which is a template_id
	workspace has a value which is a workspace_id
	reaction has a value which is a string
	clearComplexes has a value which is a bool
	new has a value which is a bool
	delete has a value which is a bool
	compartment has a value which is a compartment_id
	complexesToAdd has a value which is a reference to a list where each element is a complex_id
	complexesToRemove has a value which is a reference to a list where each element is a complex_id
	direction has a value which is a string
	type has a value which is a string
	auth has a value which is a string
template_id is a string
workspace_id is a string
bool is an int
compartment_id is a string
complex_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description70">Description</dt>
<dd>

<p>Modifies a reaction of a template model</p>

</dd>
</dl>

<h2 id="adjust_template_biomass">adjust_template_biomass</h2>

<pre><code>  $modelMeta = $obj-&gt;adjust_template_biomass($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types71">Parameter and return types</dt>
<dd>



<pre>
$params is an adjust_template_biomass_params
$modelMeta is an object_metadata
adjust_template_biomass_params is a reference to a hash where the following keys are defined:
	templateModel has a value which is a template_id
	workspace has a value which is a workspace_id
	biomass has a value which is a string
	new has a value which is a bool
	delete has a value which is a bool
	clearBiomassCompounds has a value which is a bool
	name has a value which is a string
	type has a value which is a string
	other has a value which is a string
	protein has a value which is a string
	dna has a value which is a string
	rna has a value which is a string
	cofactor has a value which is a string
	energy has a value which is a string
	cellwall has a value which is a string
	lipid has a value which is a string
	compoundsToRemove has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (compound) a compound_id
	1: (compartment) a compartment_id	compoundsToAdd has a value which is a reference to a list where each element is a reference to a list containing 7 items:
	0: (compound) a compound_id
	1: (compartment) a compartment_id
	2: (cpdclass) a string
	3: (universal) a string
	4: (coefficientType) a string
	5: (coefficient) a string
	6: (linkedCompounds) a reference to a list where each element is a reference to a list containing 2 items:
		0: (coeffficient) a string
		1: (compound) a compound_id


	auth has a value which is a string
template_id is a string
workspace_id is a string
bool is an int
compound_id is a string
compartment_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string
</pre>

</dd>
<dt id="Description71">Description</dt>
<dd>

<p>Modifies the biomass of a template model</p>

</dd>
</dl>

<h2 id="add_stimuli">add_stimuli</h2>

<pre><code>  $output = $obj-&gt;add_stimuli($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types72">Parameter and return types</dt>
<dd>



<pre>
$params is an add_stimuli_params
$output is an object_metadata
add_stimuli_params is a reference to a hash where the following keys are defined:
	biochemid has a value which is a string
	biochem_workspace has a value which is a string
	stimuliid has a value which is a string
	name has a value which is a string
	abbreviation has a value which is a string
	type has a value which is a string
	description has a value which is a string
	compounds has a value which is a reference to a list where each element is a string
	workspace has a value which is a string
	auth has a value which is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description72">Description</dt>
<dd>

<p>Adds a stimuli either to the central database or as an object in a workspace</p>

</dd>
</dl>

<h2 id="import_regulatory_model">import_regulatory_model</h2>

<pre><code>  $output = $obj-&gt;import_regulatory_model($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types73">Parameter and return types</dt>
<dd>



<pre>
$params is an import_regulatory_model_params
$output is an object_metadata
import_regulatory_model_params is a reference to a hash where the following keys are defined:
	regmodel_uid has a value which is a string
	workspace has a value which is a workspace_id
	genome has a value which is a string
	genome_ws has a value which is a workspace_id
	name has a value which is a string
	type has a value which is a string
	regulons has a value which is a reference to a list where each element is a reference to a list containing 3 items:
	0: (name) a string
	1: (features) a reference to a list where each element is a string
	2: (stimuli) a reference to a list where each element is a reference to a list containing 6 items:
		0: (stimuli) a string
		1: (in_inhibitor) a bool
		2: (strength) a float
		3: (min_conc) a float
		4: (max_conc) a float
		5: (regulators) a reference to a list where each element is a kbase_id	auth has a value which is a string
workspace_id is a string
bool is an int
kbase_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string
</pre>

</dd>
<dt id="Description73">Description</dt>
<dd>

<p>Imports a regulatory model into the KBase workspace</p>

</dd>
</dl>

<h2 id="compare_models">compare_models</h2>

<pre><code>  $output = $obj-&gt;compare_models($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types74">Parameter and return types</dt>
<dd>



<pre>
$params is a compare_models_params
$output is a ModelComparisonData
compare_models_params is a reference to a hash where the following keys are defined:
	models has a value which is a reference to a list where each element is a fbamodel_id
	workspaces has a value which is a reference to a list where each element is a workspace_id
	auth has a value which is a string
fbamodel_id is a string
workspace_id is a string
ModelComparisonData is a reference to a hash where the following keys are defined:
	model_comparisons has a value which is a reference to a list where each element is a ModelComparisonModel
	reaction_comparisons has a value which is a reference to a list where each element is a ModelCompareReaction
	auth has a value which is a string
ModelComparisonModel is a reference to a hash where the following keys are defined:
	model has a value which is a fbamodel_id
	workspace has a value which is a workspace_id
	model_name has a value which is a string
	genome has a value which is a genome_id
	genome_name has a value which is a string
	gapfilled_reactions has a value which is an int
	core_reactions has a value which is an int
	noncore_reactions has a value which is an int
genome_id is a string
ModelCompareReaction is a reference to a hash where the following keys are defined:
	reaction has a value which is a reaction_id
	compartment has a value which is a string
	equation has a value which is a string
	core has a value which is a bool
	model_features has a value which is a reference to a hash where the key is a fbamodel_id and the value is a reference to a list where each element is a feature_id
	role has a value which is a string
	subsystem has a value which is a string
	primclass has a value which is a string
	subclass has a value which is a string
	number_models has a value which is an int
	fraction_models has a value which is a float
reaction_id is a string
bool is an int
feature_id is a string</pre>

</dd>
<dt id="Description74">Description</dt>
<dd>

<p>Compares the specified models and computes unique reactions and core reactions</p>

</dd>
</dl>

<h2 id="compare_fbas">compare_fbas</h2>

<pre><code>  $output = $obj-&gt;compare_fbas($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types75">Parameter and return types</dt>
<dd>



<pre>
$params is a compare_fbas_params
$output is an object_metadata
compare_fbas_params is a reference to a hash where the following keys are defined:
	output_id has a value which is a string
	fbas has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: a workspace_id
	1: a fba_id	workspace has a value which is a workspace_id
workspace_id is a string
fba_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string
</pre>

</dd>
<dt id="Description75">Description</dt>
<dd>

<p>Compares the specified flux balance analyses and saves comparison results to workspace</p>

</dd>
</dl>

<h2 id="compare_genomes">compare_genomes</h2>

<pre><code>  $output = $obj-&gt;compare_genomes($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types76">Parameter and return types</dt>
<dd>



<pre>
$params is a compare_genomes_params
$output is an object_metadata
compare_genomes_params is a reference to a hash where the following keys are defined:
	pangenome_id has a value which is a string
	pangenome_ws has a value which is a string
	protcomp_id has a value which is a string
	protcomp_ws has a value which is a string
	output_id has a value which is a string
	workspace has a value which is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description76">Description</dt>
<dd>

<p>Compares the specified genomes and computes unique features and core features</p>

</dd>
</dl>

<h2 id="import_metagenome_annotation">import_metagenome_annotation</h2>

<pre><code>  $output = $obj-&gt;import_metagenome_annotation($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types77">Parameter and return types</dt>
<dd>



<pre>
$params is an import_metagenome_annotation_params
$output is an object_metadata
import_metagenome_annotation_params is a reference to a hash where the following keys are defined:
	metaanno_uid has a value which is a string
	workspace has a value which is a workspace_id
	source_id has a value which is a string
	source has a value which is a string
	type has a value which is a string
	confidence_type has a value which is a string
	name has a value which is a string
	annotations has a value which is a reference to a list where each element is a reference to a list containing 5 items:
	0: (genes) a reference to a list where each element is a string
	1: (functional_role) a string
	2: (otu) a string
	3: (abundance) an int
	4: (confidence) a float	auth has a value which is a string
workspace_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string
</pre>

</dd>
<dt id="Description77">Description</dt>
<dd>

<p>Imports metagenome annotation data into a metagenome annotation object</p>

</dd>
</dl>

<h2 id="models_to_community_model">models_to_community_model</h2>

<pre><code>  $output = $obj-&gt;models_to_community_model($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types78">Parameter and return types</dt>
<dd>



<pre>
$params is an import_metagenome_annotation_params
$output is an object_metadata
import_metagenome_annotation_params is a reference to a hash where the following keys are defined:
	metaanno_uid has a value which is a string
	workspace has a value which is a workspace_id
	source_id has a value which is a string
	source has a value which is a string
	type has a value which is a string
	confidence_type has a value which is a string
	name has a value which is a string
	annotations has a value which is a reference to a list where each element is a reference to a list containing 5 items:
	0: (genes) a reference to a list where each element is a string
	1: (functional_role) a string
	2: (otu) a string
	3: (abundance) an int
	4: (confidence) a float	auth has a value which is a string
workspace_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string
</pre>

</dd>
<dt id="Description78">Description</dt>
<dd>

<p>Combines multiple single genome models into a single community model</p>

</dd>
</dl>

<h2 id="metagenome_to_fbamodels">metagenome_to_fbamodels</h2>

<pre><code>  $outputs = $obj-&gt;metagenome_to_fbamodels($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types79">Parameter and return types</dt>
<dd>



<pre>
$params is a metagenome_to_fbamodels_params
$outputs is a reference to a list where each element is an object_metadata
metagenome_to_fbamodels_params is a reference to a hash where the following keys are defined:
	model_uids has a value which is a reference to a hash where the key is a string and the value is a string
	workspace has a value which is a workspace_id
	metaanno_uid has a value which is a string
	metaanno_ws has a value which is a workspace_id
	min_abundance has a value which is a float
	confidence_threshold has a value which is a float
	max_otu_models has a value which is an int
	min_reactions has a value which is an int
	templates has a value which is a reference to a hash where the key is a string and the value is a reference to a list containing 2 items:
	0: (template_ws) a workspace_id
	1: (template_uid) a template_id	auth has a value which is a string
workspace_id is a string
template_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string
</pre>

</dd>
<dt id="Description79">Description</dt>
<dd>

<p>Constructs models from metagenome annotation OTUs</p>

</dd>
</dl>

<h2 id="import_expression">import_expression</h2>

<pre><code>  $expression_meta = $obj-&gt;import_expression($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types80">Parameter and return types</dt>
<dd>



<pre>
$input is an import_expression_params
$expression_meta is an object_metadata
import_expression_params is a reference to a hash where the following keys are defined:
	expression_data_sample_series has a value which is a reference to a hash where the key is a sample_id and the value is an ExpressionDataSample
	series has a value which is a series_id
	source_id has a value which is a string
	source_date has a value which is a string
	description has a value which is a string
	processing_comments has a value which is a string
	workspace has a value which is a workspace_id
	genome_id has a value which is a genome_id
	numerical_interpretation has a value which is a string
sample_id is a string
ExpressionDataSample is a reference to a hash where the following keys are defined:
	sample_id has a value which is a string
	data_expression_levels_for_sample has a value which is a reference to a hash where the key is a feature_id and the value is a measurement
feature_id is a string
measurement is a float
series_id is a string
workspace_id is a string
genome_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description80">Description</dt>
<dd>

<p>Import gene expression.</p>

</dd>
</dl>

<h2 id="import_regulome">import_regulome</h2>

<pre><code>  $regulome_meta = $obj-&gt;import_regulome($input)</code></pre>

<dl>

<dt id="Parameter-and-return-types81">Parameter and return types</dt>
<dd>



<pre>
$input is an import_regulome_params
$regulome_meta is an object_metadata
import_regulome_params is a reference to a hash where the following keys are defined:
	regulons has a value which is a reference to a list where each element is a regulon
	workspace has a value which is a workspace_id
	genome_workspace has a value which is a workspace_id
	genome_id has a value which is a genome_id
regulon is a reference to a hash where the following keys are defined:
	operons has a value which is a reference to a list where each element is an operon
	transcription_factor has a value which is a locus
	effectors has a value which is a reference to a list where each element is an effector
	sign has a value which is a string
operon is a reference to a list where each element is a locus
locus is a reference to a hash where the following keys are defined:
	name has a value which is a string
	locus has a value which is a string
effector is a reference to a hash where the following keys are defined:
	name has a value which is a string
	class has a value which is a string
workspace_id is a string
genome_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description81">Description</dt>
<dd>

</dd>
</dl>

<h2 id="create_promconstraint">create_promconstraint</h2>

<pre><code>  $promconstraint_meta = $obj-&gt;create_promconstraint($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types82">Parameter and return types</dt>
<dd>



<pre>
$params is a CreatePromConstraintParameters
$promconstraint_meta is an object_metadata
CreatePromConstraintParameters is a reference to a hash where the following keys are defined:
	genome_id has a value which is a genome_id
	series_id has a value which is a series_id
	regulome_id has a value which is a regulome_id
	promconstraint_id has a value which is a promconstraint_id
genome_id is a string
series_id is a string
regulome_id is a string
promconstraint_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description82">Description</dt>
<dd>

<p>This method creates a set of Prom constraints for a given genome annotation based on a regulatory network and a collection of gene expression data stored on a workspace. Parameters are specified in the CreatePromconstraintParameters object. The ID of the new Prom constraints object is returned. The Prom constraints can then be used in conjunction with an FBA model using FBA Model Services.</p>

</dd>
</dl>

<h2 id="add_biochemistry_compounds">add_biochemistry_compounds</h2>

<pre><code>  $output = $obj-&gt;add_biochemistry_compounds($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types83">Parameter and return types</dt>
<dd>



<pre>
$params is an add_biochemistry_compounds_params
$output is an object_metadata
add_biochemistry_compounds_params is a reference to a hash where the following keys are defined:
	compounds has a value which is a reference to a list where each element is a reference to a list containing 9 items:
	0: (abbreviation) a string
	1: (name) a string
	2: (aliases) a reference to a list where each element is a string
	3: (formula) a string
	4: (charge) a float
	5: (isCofactor) a bool
	6: (structureString) a string
	7: (structureType) a string
	8: (id) a string	workspace has a value which is a string
	biochemistry has a value which is a string
	biochemistry_ws has a value which is a string
	output_id has a value which is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string
</pre>

</dd>
<dt id="Description83">Description</dt>
<dd>

</dd>
</dl>

<h2 id="update_object_references">update_object_references</h2>

<pre><code>  $output = $obj-&gt;update_object_references($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types84">Parameter and return types</dt>
<dd>



<pre>
$params is an update_object_references_params
$output is an object_metadata
update_object_references_params is a reference to a hash where the following keys are defined:
	object has a value which is a string
	object_workspace has a value which is a string
	original_object has a value which is a string
	original_workspace has a value which is a string
	original_instance has a value which is a string
	reference_field has a value which is a string
	newobject has a value which is a string
	newobject_workspace has a value which is a string
	newobject_instance has a value which is a string
	create_newobject has a value which is a bool
	update_subrefs has a value which is a bool
	output_id has a value which is a string
	workspace has a value which is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description84">Description</dt>
<dd>

</dd>
</dl>

<h2 id="add_reactions">add_reactions</h2>

<pre><code>  $output = $obj-&gt;add_reactions($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types85">Parameter and return types</dt>
<dd>



<pre>
$params is an add_reactions_params
$output is an object_metadata
add_reactions_params is a reference to a hash where the following keys are defined:
	model has a value which is a string
	model_workspace has a value which is a string
	output_id has a value which is a string
	workspace has a value which is a string
	reactions has a value which is a reference to a list where each element is a reference to a list containing 9 items:
	0: (reaction_id) a string
	1: (compartment) a string
	2: (direction) a string
	3: (gpr) a string
	4: (pathway) a string
	5: (name) a string
	6: (reference) a string
	7: (enzyme) a string
	8: (equation) a stringobject_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description85">Description</dt>
<dd>

<p>Add new reactions to the model from the biochemistry or custom reactions</p>

</dd>
</dl>

<h2 id="remove_reactions">remove_reactions</h2>

<pre><code>  $output = $obj-&gt;remove_reactions($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types86">Parameter and return types</dt>
<dd>



<pre>
$params is a remove_reactions_params
$output is an object_metadata
remove_reactions_params is a reference to a hash where the following keys are defined:
	model has a value which is a string
	model_workspace has a value which is a string
	output_id has a value which is a string
	workspace has a value which is a string
	reactions has a value which is a reference to a list where each element is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description86">Description</dt>
<dd>

<p>Remove reactions from the model</p>

</dd>
</dl>

<h2 id="modify_reactions">modify_reactions</h2>

<pre><code>  $output = $obj-&gt;modify_reactions($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types87">Parameter and return types</dt>
<dd>



<pre>
$params is a modify_reactions_params
$output is an object_metadata
modify_reactions_params is a reference to a hash where the following keys are defined:
	model has a value which is a string
	model_workspace has a value which is a string
	output_id has a value which is a string
	workspace has a value which is a string
	reactions has a value which is a reference to a list where each element is a reference to a list containing 7 items:
	0: (reaction_id) a string
	1: (direction) a string
	2: (gpr) a string
	3: (pathway) a string
	4: (name) a string
	5: (reference) a string
	6: (enzyme) a stringobject_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description87">Description</dt>
<dd>

<p>Modify reactions in the model</p>

</dd>
</dl>

<h2 id="add_features">add_features</h2>

<pre><code>  $output = $obj-&gt;add_features($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types88">Parameter and return types</dt>
<dd>



<pre>
$params is an add_features_params
$output is an object_metadata
add_features_params is a reference to a hash where the following keys are defined:
	genome has a value which is a string
	genome_workspace has a value which is a string
	output_id has a value which is a string
	workspace has a value which is a string
	genes has a value which is a reference to a list where each element is a reference to a list containing 9 items:
	0: (feature) a feature_id
	1: (function) a string
	2: (type) a string
	3: (aliases) a reference to a list where each element is a string
	4: (publications) a reference to a list where each element is a string
	5: (annotations) a reference to a list where each element is a string
	6: (protein_translation) a string
	7: (dna_sequence) a string
	8: (locations) a reference to a list where each element is a reference to a list containing 4 items:
		0: a string
		1: an int
		2: a string
		3: an intfeature_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description88">Description</dt>
<dd>

<p>Add new features to the genome</p>

</dd>
</dl>

<h2 id="remove_features">remove_features</h2>

<pre><code>  $output = $obj-&gt;remove_features($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types89">Parameter and return types</dt>
<dd>



<pre>
$params is a remove_features_params
$output is an object_metadata
remove_features_params is a reference to a hash where the following keys are defined:
	genome has a value which is a string
	genome_workspace has a value which is a string
	output_id has a value which is a string
	workspace has a value which is a string
	features has a value which is a reference to a list where each element is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description89">Description</dt>
<dd>

<p>Remove features from the genome</p>

</dd>
</dl>

<h2 id="modify_features">modify_features</h2>

<pre><code>  $output = $obj-&gt;modify_features($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types90">Parameter and return types</dt>
<dd>



<pre>
$params is a modify_features_params
$output is an object_metadata
modify_features_params is a reference to a hash where the following keys are defined:
	genome has a value which is a string
	genome_workspace has a value which is a string
	output_id has a value which is a string
	workspace has a value which is a string
	genes has a value which is a reference to a list where each element is a reference to a list containing 9 items:
	0: (feature) a feature_id
	1: (function) a string
	2: (type) a string
	3: (aliases) a reference to a list where each element is a string
	4: (publications) a reference to a list where each element is a string
	5: (annotations) a reference to a list where each element is a string
	6: (protein_translation) a string
	7: (dna_sequence) a string
	8: (locations) a reference to a list where each element is a reference to a list containing 4 items:
		0: a string
		1: an int
		2: a string
		3: an intfeature_id is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description90">Description</dt>
<dd>

<p>Modify features in the genome</p>

</dd>
</dl>

<h2 id="import_trainingset">import_trainingset</h2>

<pre><code>  $output = $obj-&gt;import_trainingset($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types91">Parameter and return types</dt>
<dd>



<pre>
$params is an import_trainingset_params
$output is an object_metadata
import_trainingset_params is a reference to a hash where the following keys are defined:
	workspace_training_set has a value which is a reference to a list where each element is a reference to a list containing 3 items:
	0: (workspace_id) a string
	1: (genome_id) a string
	2: (class) a string	external_training_set has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (database) a string
	1: (genome_id) a string
	2: (class) a string
	3: (attributes) a reference to a list where each element is a string

	description has a value which is a string
	class_data has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (class) a string
	1: (description) a string

	attribute_type has a value which is a string
	preload_attributes has a value which is a bool
	workspace has a value which is a string
	output_id has a value which is a string
bool is an int
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string
</pre>

</dd>
<dt id="Description91">Description</dt>
<dd>

<p>Import a training set of genomes and classifications</p>

</dd>
</dl>

<h2 id="preload_trainingset">preload_trainingset</h2>

<pre><code>  $output = $obj-&gt;preload_trainingset($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types92">Parameter and return types</dt>
<dd>



<pre>
$params is a preload_trainingset_params
$output is an object_metadata
preload_trainingset_params is a reference to a hash where the following keys are defined:
	trainingset has a value which is a string
	trainingset_ws has a value which is a string
	attribute_type has a value which is a string
	workspace has a value which is a string
	output_id has a value which is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description92">Description</dt>
<dd>

<p>Preloads a training set with attributes, cutting time to produce distinct classifiers</p>

</dd>
</dl>

<h2 id="build_classifier">build_classifier</h2>

<pre><code>  $output = $obj-&gt;build_classifier($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types93">Parameter and return types</dt>
<dd>



<pre>
$params is a build_classifier_params
$output is an object_metadata
build_classifier_params is a reference to a hash where the following keys are defined:
	trainingset has a value which is a string
	trainingset_ws has a value which is a string
	attribute_type has a value which is a string
	classifier_type has a value which is a string
	workspace has a value which is a string
	output_id has a value which is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description93">Description</dt>
<dd>

<p>Build a classifier for the input set of genomes</p>

</dd>
</dl>

<h2 id="classify_genomes">classify_genomes</h2>

<pre><code>  $output = $obj-&gt;classify_genomes($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types94">Parameter and return types</dt>
<dd>



<pre>
$params is a classify_genomes_params
$output is an object_metadata
classify_genomes_params is a reference to a hash where the following keys are defined:
	workspace_genomes has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (workspace_id) a string
	1: (genome_id) a string	external_genomes has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (database) a string
	1: (genome_id) a string

	workspace has a value which is a string
	output_id has a value which is a string
	classifier_ws has a value which is a string
	classifier has a value which is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string
</pre>

</dd>
<dt id="Description94">Description</dt>
<dd>

<p>Build a classifier for the input set of genomes</p>

</dd>
</dl>

<h2 id="build_tissue_model">build_tissue_model</h2>

<pre><code>  $output = $obj-&gt;build_tissue_model($params)</code></pre>

<dl>

<dt id="Parameter-and-return-types95">Parameter and return types</dt>
<dd>



<pre>
$params is a build_tissue_model_params
$output is an object_metadata
build_tissue_model_params is a reference to a hash where the following keys are defined:
	expsample_ws has a value which is a string
	expsample has a value which is a string
	model_ws has a value which is a string
	model has a value which is a string
	workspace has a value which is a string
	output_id has a value which is a string
object_metadata is a reference to a list containing 11 items:
	0: (id) an object_id
	1: (type) an object_type
	2: (moddate) a timestamp
	3: (instance) an int
	4: (command) a string
	5: (lastmodifier) a username
	6: (owner) a username
	7: (workspace) a workspace_id
	8: (ref) a workspace_ref
	9: (chsum) a string
	10: (metadata) a reference to a hash where the key is a string and the value is a string
object_id is a string
object_type is a string
timestamp is a string
username is a string
workspace_id is a string
workspace_ref is a string</pre>

</dd>
<dt id="Description95">Description</dt>
<dd>

<p>Build a tissue model based on the input expression data</p>

</dd>
</dl>

<h1 id="TYPES">TYPES</h1>

<h2 id="bool">bool</h2>

<dl>

<dt id="Description96">Description</dt>
<dd>

<p>******************************************************************************** Universal simple type definitions ********************************************************************************</p>

</dd>
<dt id="Definition">Definition</dt>
<dd>



<pre>
an int
</pre>

</dd>
</dl>

<h2 id="mdlrxn_kbid">mdlrxn_kbid</h2>

<dl>

<dt id="Description97">Description</dt>
<dd>

<p>KBase ID for a model reaction</p>

</dd>
<dt id="Definition1">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="ws_ref">ws_ref</h2>

<dl>

<dt id="Description98">Description</dt>
<dd>

<p>A string specifying a full workspace ID</p>

</dd>
<dt id="Definition2">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="kbase_id">kbase_id</h2>

<dl>

<dt id="Description99">Description</dt>
<dd>

<p>A string specifying a KBase ID</p>

</dd>
<dt id="Definition3">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="workspace_id">workspace_id</h2>

<dl>

<dt id="Description100">Description</dt>
<dd>

<p>A string used as an ID for a workspace. Any string consisting of alphanumeric characters and &quot;-&quot; is acceptable</p>

</dd>
<dt id="Definition4">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="complex_id">complex_id</h2>

<dl>

<dt id="Description101">Description</dt>
<dd>

<p>A string used as an ID for a complex.</p>

</dd>
<dt id="Definition5">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="template_id">template_id</h2>

<dl>

<dt id="Description102">Description</dt>
<dd>

<p>A string used as an ID for a complex.</p>

</dd>
<dt id="Definition6">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="role_id">role_id</h2>

<dl>

<dt id="Description103">Description</dt>
<dd>

<p>A string used as an ID for a complex.</p>

</dd>
<dt id="Definition7">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="object_type">object_type</h2>

<dl>

<dt id="Description104">Description</dt>
<dd>

<p>A string indicating the &quot;type&quot; of an object stored in a workspace. Acceptable types are returned by the &quot;get_types()&quot; command in the workspace_service</p>

</dd>
<dt id="Definition8">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="object_id">object_id</h2>

<dl>

<dt id="Description105">Description</dt>
<dd>

<p>ID of an object stored in the workspace. Any string consisting of alphanumeric characters and &quot;-&quot; is acceptable</p>

</dd>
<dt id="Definition9">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="username">username</h2>

<dl>

<dt id="Description106">Description</dt>
<dd>

<p>Login name of KBase useraccount to which permissions for workspaces are mapped</p>

</dd>
<dt id="Definition10">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="timestamp">timestamp</h2>

<dl>

<dt id="Description107">Description</dt>
<dd>

<p>Exact time for workspace operations. e.g. 2012-12-17T23:24:06</p>

</dd>
<dt id="Definition11">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="compound_id">compound_id</h2>

<dl>

<dt id="Description108">Description</dt>
<dd>

<p>An identifier for compounds in the KBase biochemistry database. e.g. cpd00001</p>

</dd>
<dt id="Definition12">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="biochemistry_id">biochemistry_id</h2>

<dl>

<dt id="Description109">Description</dt>
<dd>

<p>A string used to identify a particular biochemistry database object in KBase. e.g. &quot;default&quot; is the ID of the standard KBase biochemistry</p>

</dd>
<dt id="Definition13">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="genome_id">genome_id</h2>

<dl>

<dt id="Description110">Description</dt>
<dd>

<p>A string identifier for a genome in KBase. e.g. &quot;kb|g.0&quot; is the ID for E. coli</p>

</dd>
<dt id="Definition14">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="promconstraint_id">promconstraint_id</h2>

<dl>

<dt id="Description111">Description</dt>
<dd>

<p>A string identifier for a promconstraint in KBase.</p>

</dd>
<dt id="Definition15">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="sample_id">sample_id</h2>

<dl>

<dt id="Description112">Description</dt>
<dd>

<p>A string identifier for a gene expression sample in KBase.</p>

</dd>
<dt id="Definition16">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="series_id">series_id</h2>

<dl>

<dt id="Description113">Description</dt>
<dd>

<p>A string identifier for a gene expression sample series in KBase.</p>

</dd>
<dt id="Definition17">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="contig_id">contig_id</h2>

<dl>

<dt id="Description114">Description</dt>
<dd>

<p>A string identifier for a contiguous piece of DNA in KBase, representing a chromosome or an assembled fragment</p>

</dd>
<dt id="Definition18">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="feature_type">feature_type</h2>

<dl>

<dt id="Description115">Description</dt>
<dd>

<p>A string specifying the type of genome features in KBase</p>

</dd>
<dt id="Definition19">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="modelcompartment_id">modelcompartment_id</h2>

<dl>

<dt id="Description116">Description</dt>
<dd>

<p>A string identifier used for compartments in models in KBase. Compartments could represet organelles in a eukaryotic model, or entire cells in a community model</p>

</dd>
<dt id="Definition20">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="modelcompound_id">modelcompound_id</h2>

<dl>

<dt id="Description117">Description</dt>
<dd>

<p>A string identifier used for compounds in models in KBase.</p>

</dd>
<dt id="Definition21">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="feature_id">feature_id</h2>

<dl>

<dt id="Description118">Description</dt>
<dd>

<p>A string identifier used for a feature in a genome.</p>

</dd>
<dt id="Definition22">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="reaction_id">reaction_id</h2>

<dl>

<dt id="Description119">Description</dt>
<dd>

<p>A string identifier used for a reaction in a KBase biochemistry.</p>

</dd>
<dt id="Definition23">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="modelreaction_id">modelreaction_id</h2>

<dl>

<dt id="Description120">Description</dt>
<dd>

<p>A string identifier used for a reaction in a model in KBase.</p>

</dd>
<dt id="Definition24">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="biomass_id">biomass_id</h2>

<dl>

<dt id="Description121">Description</dt>
<dd>

<p>A string identifier used for a biomass reaction in a KBase model.</p>

</dd>
<dt id="Definition25">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="media_id">media_id</h2>

<dl>

<dt id="Description122">Description</dt>
<dd>

<p>A string identifier used for a media condition in the KBase database.</p>

</dd>
<dt id="Definition26">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="fba_id">fba_id</h2>

<dl>

<dt id="Description123">Description</dt>
<dd>

<p>A string identifier used for a flux balance analysis study in KBase.</p>

</dd>
<dt id="Definition27">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="gapgen_id">gapgen_id</h2>

<dl>

<dt id="Description124">Description</dt>
<dd>

<p>A string identifier for a gap generation study in KBase.</p>

</dd>
<dt id="Definition28">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="gapfill_id">gapfill_id</h2>

<dl>

<dt id="Description125">Description</dt>
<dd>

<p>A string identifier for a gap filling study in KBase.</p>

</dd>
<dt id="Definition29">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="gapgensolution_id">gapgensolution_id</h2>

<dl>

<dt id="Description126">Description</dt>
<dd>

<p>A string identifier for a solution from a gap generation study in KBase.</p>

</dd>
<dt id="Definition30">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="gapfillsolution_id">gapfillsolution_id</h2>

<dl>

<dt id="Description127">Description</dt>
<dd>

<p>A string identifier for a solution from a gap filling study in KBase.</p>

</dd>
<dt id="Definition31">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="fbamodel_id">fbamodel_id</h2>

<dl>

<dt id="Description128">Description</dt>
<dd>

<p>A string identifier for a metabolic model in KBase.</p>

</dd>
<dt id="Definition32">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="mapping_id">mapping_id</h2>

<dl>

<dt id="Description129">Description</dt>
<dd>

<p>A string identifier for a Mapping object in KBase.</p>

</dd>
<dt id="Definition33">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="regulome_id">regulome_id</h2>

<dl>

<dt id="Description130">Description</dt>
<dd>

<p>A string identifier for a regulome in KBase.</p>

</dd>
<dt id="Definition34">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="regmodel_id">regmodel_id</h2>

<dl>

<dt id="Description131">Description</dt>
<dd>

<p>A string identifier for a regulatory model in KBase.</p>

</dd>
<dt id="Definition35">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="compartment_id">compartment_id</h2>

<dl>

<dt id="Description132">Description</dt>
<dd>

<p>A string identifier for a compartment in KBase.</p>

</dd>
<dt id="Definition36">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="expression_id">expression_id</h2>

<dl>

<dt id="Description133">Description</dt>
<dd>

<p>A string identifier for an expression dataset in KBase.</p>

</dd>
<dt id="Definition37">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="phenotype_set_id">phenotype_set_id</h2>

<dl>

<dt id="Description134">Description</dt>
<dd>

<p>A string identifier used for a set of phenotype data loaded into KBase.</p>

</dd>
<dt id="Definition38">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="workspace_ref">workspace_ref</h2>

<dl>

<dt id="Description135">Description</dt>
<dd>

<p>A permanent reference to an object in a workspace.</p>

</dd>
<dt id="Definition39">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="probanno_id">probanno_id</h2>

<dl>

<dt id="Description136">Description</dt>
<dd>

<p>A string identifier used for a probabilistic annotation in KBase.</p>

</dd>
<dt id="Definition40">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="reaction_synonyms_id">reaction_synonyms_id</h2>

<dl>

<dt id="Description137">Description</dt>
<dd>

<p>A string identifier for a reaction synonyms in KBase.</p>

</dd>
<dt id="Definition41">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="Genome_uid">Genome_uid</h2>

<dl>

<dt id="Description138">Description</dt>
<dd>

<p>A user ID for a genome in KBase</p>

</dd>
<dt id="Definition42">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="object_metadata">object_metadata</h2>

<dl>

<dt id="Description139">Description</dt>
<dd>

<p>******************************************************************************** Object type definition ********************************************************************************</p>

</dd>
<dt id="Definition43">Definition</dt>
<dd>



<pre>
a reference to a list containing 11 items:
0: (id) an object_id
1: (type) an object_type
2: (moddate) a timestamp
3: (instance) an int
4: (command) a string
5: (lastmodifier) a username
6: (owner) a username
7: (workspace) a workspace_id
8: (ref) a workspace_ref
9: (chsum) a string
10: (metadata) a reference to a hash where the key is a string and the value is a string</pre>

</dd>
</dl>

<h2 id="md5">md5</h2>

<dl>

<dt id="Description140">Description</dt>
<dd>

<p>******************************************************************************** Probabilistic Annotation type definition ********************************************************************************</p>

</dd>
<dt id="Definition44">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="md5s">md5s</h2>

<dl>

<dt id="Definition45">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a md5
</pre>

</dd>
</dl>

<h2 id="region_of_dna">region_of_dna</h2>

<dl>

<dt id="Description141">Description</dt>
<dd>

<p>A region of DNA is maintained as a tuple of four components:</p>

<pre><code>        the contig
        the beginning position (from 1)
        the strand
        the length

        We often speak of &quot;a region&quot;.  By &quot;location&quot;, we mean a sequence
        of regions from the same genome (perhaps from distinct contigs).</code></pre>

</dd>
<dt id="Definition46">Definition</dt>
<dd>



<pre>
a reference to a list containing 4 items:
0: a contig_id
1: (begin) an int
2: (strand) a string
3: (length) an int</pre>

</dd>
</dl>

<h2 id="location">location</h2>

<dl>

<dt id="Description142">Description</dt>
<dd>

<p>a &quot;location&quot; refers to a sequence of regions</p>

</dd>
<dt id="Definition47">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a region_of_dna
</pre>

</dd>
</dl>

<h2 id="annotation">annotation</h2>

<dl>

<dt id="Definition48">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: (comment) a string
1: (annotator) a string
2: (annotation_time) an int</pre>

</dd>
</dl>

<h2 id="gene_hit">gene_hit</h2>

<dl>

<dt id="Definition49">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: (gene) a feature_id
1: (blast_score) a float</pre>

</dd>
</dl>

<h2 id="alt_func">alt_func</h2>

<dl>

<dt id="Definition50">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: (function) a string
1: (probability) a float
2: (gene_hits) a reference to a list where each element is a gene_hit</pre>

</dd>
</dl>

<h2 id="feature">feature</h2>

<dl>

<dt id="Definition51">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a feature_id
location has a value which is a location
type has a value which is a feature_type
function has a value which is a string
alternative_functions has a value which is a reference to a list where each element is an alt_func
protein_translation has a value which is a string
aliases has a value which is a reference to a list where each element is a string
annotations has a value which is a reference to a list where each element is an annotation</pre>

</dd>
</dl>

<h2 id="contig">contig</h2>

<dl>

<dt id="Definition52">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a contig_id
dna has a value which is a string</pre>

</dd>
</dl>

<h2 id="GenomeObject">GenomeObject</h2>

<dl>

<dt id="Definition53">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a genome_id
scientific_name has a value which is a string
domain has a value which is a string
genetic_code has a value which is an int
source has a value which is a string
source_id has a value which is a string
contigs has a value which is a reference to a list where each element is a contig
features has a value which is a reference to a list where each element is a feature</pre>

</dd>
</dl>

<h2 id="annotationProbability">annotationProbability</h2>

<dl>

<dt id="Description143">Description</dt>
<dd>

<p>Data structures to hold a single annotation probability for a single gene</p>

<p>feature_id feature - feature the annotation is associated with string function - the name of the functional role being annotated to the feature float probability - the probability that the functional role is associated with the feature</p>

</dd>
<dt id="Definition54">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: (feature) a feature_id
1: (function) a string
2: (probability) a float</pre>

</dd>
</dl>

<h2 id="probanno_id1">probanno_id</h2>

<dl>

<dt id="Definition55">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="alt_func1">alt_func</h2>

<dl>

<dt id="Definition56">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: (function) a string
1: (probability) a float</pre>

</dd>
</dl>

<h2 id="ProbAnnoFeature">ProbAnnoFeature</h2>

<dl>

<dt id="Description144">Description</dt>
<dd>

<p>Object to carry alternative functions for each feature</p>

<p>feature_id id ID of the feature. Required.</p>

<p>string function Primary annotated function of the feature in the genome annotation. Required.</p>

<p>list&lt;alt_func&gt; alternative_functions List of tuples containing alternative functions and probabilities. Required.</p>

</dd>
<dt id="Definition57">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a feature_id
alternative_functions has a value which is a reference to a list where each element is an alt_func</pre>

</dd>
</dl>

<h2 id="ProbabilisticAnnotation">ProbabilisticAnnotation</h2>

<dl>

<dt id="Description145">Description</dt>
<dd>

<p>Object to carry alternative functions and probabilities for genes in a genome</p>

<pre><code>    probanno_id id - ID of the probabilistic annotation object. Required.    
    genome_id genome - ID of the genome the probabilistic annotation was built for. Required.
    workspace_ref genome_uuid - Reference to retrieve genome from workspace service. Required.
    list&lt;ProbAnnoFeature&gt; featureAlternativeFunctions - List of ProbAnnoFeature objects holding alternative functions for features. Required.</code></pre>

</dd>
<dt id="Definition58">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a probanno_id
genome has a value which is a genome_id
genome_uuid has a value which is a workspace_ref
featureAlternativeFunctions has a value which is a reference to a list where each element is a ProbAnnoFeature</pre>

</dd>
</dl>

<h2 id="ReactionProbability">ReactionProbability</h2>

<dl>

<dt id="Description146">Description</dt>
<dd>

<p>Data structure to hold probability of a reaction</p>

<pre><code>        reaction_id reaction - ID of the reaction
        float probability - Probability of the reaction
        string gene_list - List of genes most likely to be attached to reaction</code></pre>

</dd>
<dt id="Definition59">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: (reaction) a reaction_id
1: (probability) a float
2: (gene_list) a string</pre>

</dd>
</dl>

<h2 id="Biochemistry">Biochemistry</h2>

<dl>

<dt id="Description147">Description</dt>
<dd>

<p>******************************************************************************** Biochemistry type definition ********************************************************************************</p>

</dd>
<dt id="Definition60">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a biochemistry_id
name has a value which is a string
compounds has a value which is a reference to a list where each element is a compound_id
reactions has a value which is a reference to a list where each element is a reaction_id
media has a value which is a reference to a list where each element is a media_id</pre>

</dd>
</dl>

<h2 id="MediaCompound">MediaCompound</h2>

<dl>

<dt id="Description148">Description</dt>
<dd>

<p>Data structures for media compound formulation</p>

<p>compound_id compound - ID of compound in media string name - name of compound in media float concentration - concentration of compound in media float maxFlux - maximum flux of compound in media float minFlux - minimum flux of compound in media</p>

</dd>
<dt id="Definition61">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
compound has a value which is a compound_id
name has a value which is a string
concentration has a value which is a float
max_flux has a value which is a float
min_flux has a value which is a float</pre>

</dd>
</dl>

<h2 id="Media">Media</h2>

<dl>

<dt id="Description149">Description</dt>
<dd>

<p>Data structures for media formulation</p>

<p>media_id id - ID of media formulation string name - name of media formulaiton list&lt;MediaCompound&gt; media_compounds - list of compounds in media formulation float pH - pH of media condition float temperature - temperature of media condition</p>

</dd>
<dt id="Definition62">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a media_id
name has a value which is a string
media_compounds has a value which is a reference to a list where each element is a MediaCompound
pH has a value which is a float
temperature has a value which is a float</pre>

</dd>
</dl>

<h2 id="Compound">Compound</h2>

<dl>

<dt id="Description150">Description</dt>
<dd>

<p>Data structures for media formulation</p>

<p>compound_id id - ID of compound string abbrev - abbreviated name of compound string name - primary name of compound list&lt;string&gt; aliases - list of aliases for compound float charge - molecular charge of compound float deltaG - estimated compound delta G float deltaGErr - uncertainty in estimated compound delta G string formula - molecular formula of compound</p>

</dd>
<dt id="Definition63">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a compound_id
abbrev has a value which is a string
name has a value which is a string
aliases has a value which is a reference to a list where each element is a string
charge has a value which is a float
deltaG has a value which is a float
deltaGErr has a value which is a float
formula has a value which is a string</pre>

</dd>
</dl>

<h2 id="Reaction">Reaction</h2>

<dl>

<dt id="Description151">Description</dt>
<dd>

<p>Data structures for media formulation</p>

<p>reaction_id id - ID of reaction string name - primary name of reaction string abbrev - abbreviated name of reaction list&lt;string&gt; enzymes - list of EC numbers for reaction string direction - directionality of reaction string reversibility - reversibility of reaction float deltaG - estimated delta G of reaction float deltaGErr - uncertainty in estimated delta G of reaction string equation - reaction equation in terms of compound IDs string definition - reaction equation in terms of compound names</p>

</dd>
<dt id="Definition64">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a reaction_id
name has a value which is a string
abbrev has a value which is a string
enzymes has a value which is a reference to a list where each element is a string
direction has a value which is a string
reversibility has a value which is a string
deltaG has a value which is a float
deltaGErr has a value which is a float
equation has a value which is a string
definition has a value which is a string</pre>

</dd>
</dl>

<h2 id="ModelCompartment">ModelCompartment</h2>

<dl>

<dt id="Description152">Description</dt>
<dd>

<p>******************************************************************************** FBAModel type definition ********************************************************************************</p>

</dd>
<dt id="Definition65">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a modelcompartment_id
name has a value which is a string
pH has a value which is a float
potential has a value which is a float
index has a value which is an int</pre>

</dd>
</dl>

<h2 id="ModelCompound">ModelCompound</h2>

<dl>

<dt id="Description153">Description</dt>
<dd>

<p>Data structures for a compound in a model</p>

<p>modelcompound_id id - ID of the specific instance of the compound in the model compound_id compound - ID of the compound associated with the model compound string name - name of the compound associated with the model compound modelcompartment_id compartment - ID of the compartment containing the compound</p>

</dd>
<dt id="Definition66">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a modelcompound_id
compound has a value which is a compound_id
name has a value which is a string
compartment has a value which is a modelcompartment_id</pre>

</dd>
</dl>

<h2 id="ModelReaction">ModelReaction</h2>

<dl>

<dt id="Description154">Description</dt>
<dd>

<p>Data structures for a reaction in a model</p>

<p>modelreaction_id id - ID of the specific instance of the reaction in the model reaction_id reaction - ID of the reaction string name - name of the reaction string direction - directionality of the reaction string equation - stoichiometric equation of the reaction in terms of compound IDs string definition - stoichiometric equation of the reaction in terms of compound names list&lt;feature_id&gt; features - list of features associated with the reaction modelcompartment_id compartment - ID of the compartment containing the reaction</p>

</dd>
<dt id="Definition67">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a modelreaction_id
reaction has a value which is a reaction_id
name has a value which is a string
direction has a value which is a string
equation has a value which is a string
definition has a value which is a string
gapfilled has a value which is a bool
features has a value which is a reference to a list where each element is a feature_id
compartment has a value which is a modelcompartment_id</pre>

</dd>
</dl>

<h2 id="BiomassCompound">BiomassCompound</h2>

<dl>

<dt id="Description155">Description</dt>
<dd>

<p>Data structures for a reaction in a model</p>

<p>modelcompound_id modelcompound - ID of model compound in biomass reaction float coefficient - coefficient of compound in biomass reaction string name - name of compound in biomass reaction</p>

</dd>
<dt id="Definition68">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: (modelcompound) a modelcompound_id
1: (coefficient) a float
2: (name) a string</pre>

</dd>
</dl>

<h2 id="ModelBiomass">ModelBiomass</h2>

<dl>

<dt id="Description156">Description</dt>
<dd>

<p>Data structures for a reaction in a model</p>

<p>biomass_id id - ID of biomass reaction string name - name of biomass reaction string definition - stoichiometric equation of biomass reaction in terms of compound names list&lt;BiomassCompound&gt; biomass_compounds - list of compounds in biomass reaction</p>

</dd>
<dt id="Definition69">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a biomass_id
name has a value which is a string
definition has a value which is a string
biomass_compounds has a value which is a reference to a list where each element is a BiomassCompound</pre>

</dd>
</dl>

<h2 id="FBAMeta">FBAMeta</h2>

<dl>

<dt id="Description157">Description</dt>
<dd>

<p>Data structures for a reaction in a model</p>

<p>fba_id id - ID of the FBA object workspace_id workspace - ID of the workspace containing the FBA object media_id media - ID of the media the FBA was performed in workspace_id media_workspace - ID of the workspace containing the media formulation float objective - optimized objective value of the FBA study list&lt;feature_id&gt; ko - list of genes knocked out in the FBA study</p>

</dd>
<dt id="Definition70">Definition</dt>
<dd>



<pre>
a reference to a list containing 6 items:
0: (id) a fba_id
1: (workspace) a workspace_id
2: (media) a media_id
3: (media_workspace) a workspace_id
4: (objective) a float
5: (ko) a reference to a list where each element is a feature_id</pre>

</dd>
</dl>

<h2 id="GapGenMeta">GapGenMeta</h2>

<dl>

<dt id="Description158">Description</dt>
<dd>

<p>Metadata object providing a summary of a gapgen simulation</p>

<p>gapgen_id id - ID of gapgen study object workspace_id workspace - workspace containing gapgen study media_id media - media formulation for gapgen study workspace_id media_workspace - ID of the workspace containing the media formulation bool done - boolean indicating if gapgen study is complete list&lt;feature_id&gt; ko - list of genes knocked out in gapgen study</p>

</dd>
<dt id="Definition71">Definition</dt>
<dd>



<pre>
a reference to a list containing 6 items:
0: (id) a gapgen_id
1: (workspace) a workspace_id
2: (media) a media_id
3: (media_workspace) a workspace_id
4: (done) a bool
5: (ko) a reference to a list where each element is a feature_id</pre>

</dd>
</dl>

<h2 id="GapFillMeta">GapFillMeta</h2>

<dl>

<dt id="Description159">Description</dt>
<dd>

<p>Metadata object providing a summary of a gapfilling simulation</p>

<p>gapfill_id id - ID of gapfill study object workspace_id workspace - workspace containing gapfill study media_id media - media formulation for gapfill study workspace_id media_workspace - ID of the workspace containing the media formulation bool done - boolean indicating if gapfill study is complete list&lt;feature_id&gt; ko - list of genes knocked out in gapfill study</p>

</dd>
<dt id="Definition72">Definition</dt>
<dd>



<pre>
a reference to a list containing 6 items:
0: (id) a gapfill_id
1: (workspace) a workspace_id
2: (media) a media_id
3: (media_workspace) a workspace_id
4: (done) a bool
5: (ko) a reference to a list where each element is a feature_id</pre>

</dd>
</dl>

<h2 id="FBAModel">FBAModel</h2>

<dl>

<dt id="Description160">Description</dt>
<dd>

<p>Data structure holding data for metabolic model</p>

<p>fbamodel_id id - ID of model workspace_id workspace - workspace containing model genome_id genome - ID of associated genome workspace_id genome_workspace - workspace with associated genome mapping_id map - ID of associated mapping database workspace_id map_workspace - workspace with associated mapping database biochemistry_id biochemistry - ID of associated biochemistry database workspace_id biochemistry_workspace - workspace with associated biochemistry database string name - name of the model string type - type of model (e.g. single genome, community) string status - status of model (e.g. under construction) list&lt;ModelBiomass&gt; biomasses - list of biomass reactions in model list&lt;ModelCompartment&gt; compartments - list of compartments in model list&lt;ModelReaction&gt; reactions - list of reactions in model list&lt;ModelCompound&gt; compounds - list of compounds in model list&lt;FBAMeta&gt; fbas - list of flux balance analysis studies for model list&lt;GapFillMeta&gt; integrated_gapfillings - list of integrated gapfilling solutions list&lt;GapFillMeta&gt; unintegrated_gapfillings - list of unintegrated gapfilling solutions list&lt;GapGenMeta&gt; integrated_gapgenerations - list of integrated gapgen solutions list&lt;GapGenMeta&gt; unintegrated_gapgenerations - list of unintegrated gapgen solutions</p>

</dd>
<dt id="Definition73">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a fbamodel_id
workspace has a value which is a workspace_id
genome has a value which is a genome_id
genome_workspace has a value which is a workspace_id
map has a value which is a mapping_id
map_workspace has a value which is a workspace_id
biochemistry has a value which is a biochemistry_id
biochemistry_workspace has a value which is a workspace_id
name has a value which is a string
type has a value which is a string
status has a value which is a string
biomasses has a value which is a reference to a list where each element is a ModelBiomass
compartments has a value which is a reference to a list where each element is a ModelCompartment
reactions has a value which is a reference to a list where each element is a ModelReaction
compounds has a value which is a reference to a list where each element is a ModelCompound
fbas has a value which is a reference to a list where each element is an FBAMeta
integrated_gapfillings has a value which is a reference to a list where each element is a GapFillMeta
unintegrated_gapfillings has a value which is a reference to a list where each element is a GapFillMeta
integrated_gapgenerations has a value which is a reference to a list where each element is a GapGenMeta
unintegrated_gapgenerations has a value which is a reference to a list where each element is a GapGenMeta</pre>

</dd>
</dl>

<h2 id="GeneAssertion">GeneAssertion</h2>

<dl>

<dt id="Description161">Description</dt>
<dd>

<p>******************************************************************************** Flux Balance Analysis type definition ********************************************************************************</p>

</dd>
<dt id="Definition74">Definition</dt>
<dd>



<pre>
a reference to a list containing 4 items:
0: (feature) a feature_id
1: (growthFraction) a float
2: (growth) a float
3: (isEssential) a bool</pre>

</dd>
</dl>

<h2 id="CompoundFlux">CompoundFlux</h2>

<dl>

<dt id="Description162">Description</dt>
<dd>

<p>Compound variable in FBA solution</p>

<p>modelcompound_id compound - ID of compound in model in FBA solution float value - flux uptake of compound in FBA solution float upperBound - maximum uptake of compoundin FBA simulation float lowerBound - minimum uptake of compoundin FBA simulation float max - maximum uptake of compoundin FBA simulation float min - minimum uptake of compoundin FBA simulation string type - type of compound variable string name - name of compound</p>

</dd>
<dt id="Definition75">Definition</dt>
<dd>



<pre>
a reference to a list containing 8 items:
0: (compound) a modelcompound_id
1: (value) a float
2: (upperBound) a float
3: (lowerBound) a float
4: (max) a float
5: (min) a float
6: (type) a string
7: (name) a string</pre>

</dd>
</dl>

<h2 id="ReactionFlux">ReactionFlux</h2>

<dl>

<dt id="Description163">Description</dt>
<dd>

<p>Reaction variable in FBA solution</p>

<p>modelreaction_id reaction - ID of reaction in model in FBA solution float value - flux through reaction in FBA solution float upperBound - maximum flux through reaction in FBA simulation float lowerBound - minimum flux through reaction in FBA simulation float max - maximum flux through reaction in FBA simulation float min - minimum flux through reaction in FBA simulation string type - type of reaction variable string definition - stoichiometry of solution reaction in terms of compound names</p>

</dd>
<dt id="Definition76">Definition</dt>
<dd>



<pre>
a reference to a list containing 8 items:
0: (reaction) a modelreaction_id
1: (value) a float
2: (upperBound) a float
3: (lowerBound) a float
4: (max) a float
5: (min) a float
6: (type) a string
7: (definition) a string</pre>

</dd>
</dl>

<h2 id="MetaboliteProduction">MetaboliteProduction</h2>

<dl>

<dt id="Description164">Description</dt>
<dd>

<p>Maximum production of compound in FBA simulation</p>

<p>float maximumProduction - maximum production of compound modelcompound_id modelcompound - ID of compound with production maximized string name - name of compound with simulated production</p>

</dd>
<dt id="Definition77">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: (maximumProduction) a float
1: (modelcompound) a modelcompound_id
2: (name) a string</pre>

</dd>
</dl>

<h2 id="MinimalMediaPrediction">MinimalMediaPrediction</h2>

<dl>

<dt id="Description165">Description</dt>
<dd>

<p>Data structures for gapfilling solution</p>

<p>list&lt;compound_id&gt; optionalNutrients - list of optional nutrients list&lt;compound_id&gt; essentialNutrients - list of essential nutrients</p>

</dd>
<dt id="Definition78">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
optionalNutrients has a value which is a reference to a list where each element is a compound_id
essentialNutrients has a value which is a reference to a list where each element is a compound_id</pre>

</dd>
</dl>

<h2 id="bound">bound</h2>

<dl>

<dt id="Description166">Description</dt>
<dd>

<p>Term of constraint or objective in FBA simulation</p>

<p>float min - minimum value of custom bound float max - maximum value of custom bound string varType - type of variable for custom bound string variable - variable ID for custom bound</p>

</dd>
<dt id="Definition79">Definition</dt>
<dd>



<pre>
a reference to a list containing 4 items:
0: (min) a float
1: (max) a float
2: (varType) a string
3: (variable) a string</pre>

</dd>
</dl>

<h2 id="term">term</h2>

<dl>

<dt id="Description167">Description</dt>
<dd>

<p>Term of constraint or objective in FBA simulation</p>

<p>float coefficient - coefficient of term in objective or constraint string varType - type of variable for term in objective or constraint string variable - variable ID for term in objective or constraint</p>

</dd>
<dt id="Definition80">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: (coefficient) a float
1: (varType) a string
2: (variable) a string</pre>

</dd>
</dl>

<h2 id="constraint">constraint</h2>

<dl>

<dt id="Description168">Description</dt>
<dd>

<p>Custom constraint in FBA simulation</p>

<p>float rhs - right hand side of custom constraint string sign - sign of custom constraint (e.g. &lt;, &gt;) list&lt;term&gt; terms - terms in custom constraint string name - name of custom constraint</p>

</dd>
<dt id="Definition81">Definition</dt>
<dd>



<pre>
a reference to a list containing 4 items:
0: (rhs) a float
1: (sign) a string
2: (terms) a reference to a list where each element is a term
3: (name) a string</pre>

</dd>
</dl>

<h2 id="FBAFormulation">FBAFormulation</h2>

<dl>

<dt id="Description169">Description</dt>
<dd>

<p>Data structures for gapfilling solution</p>

<p>media_id media - ID of media formulation to be used list&lt;compound_id&gt; additionalcpds - list of additional compounds to allow update promconstraint_id promconstraint - ID of promconstraint workspace_id promconstraint_workspace - workspace containing promconstraint sample_id eflux_sample - ID of gene expression sample to run eflux series_id eflux_series - ID of series that a sample belongs to. workspace_id eflux_workspace - workspace containing gene expression sample series for eflux workspace_id media_workspace - workspace containing media for FBA study float objfraction - fraction of objective to use for constraints bool allreversible - flag indicating if all reactions should be reversible bool maximizeObjective - flag indicating if objective should be maximized list&lt;term&gt; objectiveTerms - list of terms of objective function list&lt;feature_id&gt; geneko - list of gene knockouts list&lt;reaction_id&gt; rxnko - list of reaction knockouts list&lt;bound&gt; bounds - list of custom bounds list&lt;constraint&gt; constraints - list of custom constraints mapping&lt;string,float&gt; uptakelim - hash of maximum uptake for elements float defaultmaxflux - default maximum intracellular flux float defaultminuptake - default minimum nutrient uptake float defaultmaxuptake - default maximum nutrient uptake bool simplethermoconst - flag indicating if simple thermodynamic constraints should be used bool thermoconst - flag indicating if thermodynamic constraints should be used bool nothermoerror - flag indicating if no error should be allowed in thermodynamic constraints bool minthermoerror - flag indicating if error should be minimized in thermodynamic constraints</p>

</dd>
<dt id="Definition82">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
media has a value which is a media_id
additionalcpds has a value which is a reference to a list where each element is a compound_id
promconstraint has a value which is a promconstraint_id
promconstraint_workspace has a value which is a workspace_id
eflux_sample has a value which is a sample_id
eflux_series has a value which is a series_id
eflux_workspace has a value which is a workspace_id
media_workspace has a value which is a workspace_id
objfraction has a value which is a float
allreversible has a value which is a bool
maximizeObjective has a value which is a bool
objectiveTerms has a value which is a reference to a list where each element is a term
geneko has a value which is a reference to a list where each element is a feature_id
rxnko has a value which is a reference to a list where each element is a reaction_id
bounds has a value which is a reference to a list where each element is a bound
constraints has a value which is a reference to a list where each element is a constraint
uptakelim has a value which is a reference to a hash where the key is a string and the value is a float
defaultmaxflux has a value which is a float
defaultminuptake has a value which is a float
defaultmaxuptake has a value which is a float
simplethermoconst has a value which is a bool
thermoconst has a value which is a bool
nothermoerror has a value which is a bool
minthermoerror has a value which is a bool</pre>

</dd>
</dl>

<h2 id="FBA">FBA</h2>

<dl>

<dt id="Description170">Description</dt>
<dd>

<p>Data structures for gapfilling solution</p>

<p>fba_id id - ID of FBA study workspace_id workspace - workspace containing FBA study fbamodel_id model - ID of model FBA was run on workspace_id model_workspace - workspace with FBA model float objective - objective value of FBA study bool isComplete - flag indicating if job is complete FBAFormulation formulation - specs for FBA study list&lt;MinimalMediaPrediction&gt; minimalMediaPredictions - list of minimal media formulation list&lt;MetaboliteProduction&gt; metaboliteProductions - list of biomass component production list&lt;ReactionFlux&gt; reactionFluxes - list of reaction fluxes list&lt;CompoundFlux&gt; compoundFluxes - list of compound uptake fluxes list&lt;GeneAssertion&gt; geneAssertions - list of gene assertions</p>

</dd>
<dt id="Definition83">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a fba_id
workspace has a value which is a workspace_id
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
objective has a value which is a float
isComplete has a value which is a bool
formulation has a value which is an FBAFormulation
minimalMediaPredictions has a value which is a reference to a list where each element is a MinimalMediaPrediction
metaboliteProductions has a value which is a reference to a list where each element is a MetaboliteProduction
reactionFluxes has a value which is a reference to a list where each element is a ReactionFlux
compoundFluxes has a value which is a reference to a list where each element is a CompoundFlux
geneAssertions has a value which is a reference to a list where each element is a GeneAssertion</pre>

</dd>
</dl>

<h2 id="GapfillingFormulation">GapfillingFormulation</h2>

<dl>

<dt id="Description171">Description</dt>
<dd>

<p>******************************************************************************** Gapfilling type definition ********************************************************************************</p>

</dd>
<dt id="Definition84">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
formulation has a value which is an FBAFormulation
num_solutions has a value which is an int
nomediahyp has a value which is a bool
nobiomasshyp has a value which is a bool
nogprhyp has a value which is a bool
nopathwayhyp has a value which is a bool
allowunbalanced has a value which is a bool
activitybonus has a value which is a float
drainpen has a value which is a float
directionpen has a value which is a float
nostructpen has a value which is a float
unfavorablepen has a value which is a float
nodeltagpen has a value which is a float
biomasstranspen has a value which is a float
singletranspen has a value which is a float
transpen has a value which is a float
blacklistedrxns has a value which is a reference to a list where each element is a reaction_id
gauranteedrxns has a value which is a reference to a list where each element is a reaction_id
allowedcmps has a value which is a reference to a list where each element is a compartment_id
probabilisticAnnotation has a value which is a probanno_id
probabilisticAnnotation_workspace has a value which is a workspace_id</pre>

</dd>
</dl>

<h2 id="reactionAddition">reactionAddition</h2>

<dl>

<dt id="Description172">Description</dt>
<dd>

<p>Reactions removed in gapgen solution</p>

<p>modelreaction_id reaction - ID of the removed reaction string direction - direction of reaction removed in gapgen solution string equation - stoichiometry of removed reaction in terms of compound IDs string definition - stoichiometry of removed reaction in terms of compound names</p>

</dd>
<dt id="Definition85">Definition</dt>
<dd>



<pre>
a reference to a list containing 5 items:
0: (reaction) a reaction_id
1: (direction) a string
2: (compartment_id) a string
3: (equation) a string
4: (definition) a string</pre>

</dd>
</dl>

<h2 id="biomassRemoval">biomassRemoval</h2>

<dl>

<dt id="Description173">Description</dt>
<dd>

<p>Biomass component removed in gapfill solution</p>

<p>compound_id compound - ID of biomass component removed string name - name of biomass component removed</p>

</dd>
<dt id="Definition86">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: (compound) a compound_id
1: (name) a string</pre>

</dd>
</dl>

<h2 id="mediaAddition">mediaAddition</h2>

<dl>

<dt id="Description174">Description</dt>
<dd>

<p>Media component added in gapfill solution</p>

<p>compound_id compound - ID of media component added string name - name of media component added</p>

</dd>
<dt id="Definition87">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: (compound) a compound_id
1: (name) a string</pre>

</dd>
</dl>

<h2 id="GapFillSolution">GapFillSolution</h2>

<dl>

<dt id="Description175">Description</dt>
<dd>

<p>Data structures for gapfilling solution</p>

<p>gapfillsolution_id id - ID of gapfilling solution float objective - cost of gapfilling solution list&lt;biomassRemoval&gt; biomassRemovals - list of biomass components being removed list&lt;mediaAddition&gt; mediaAdditions - list of media components being added list&lt;reactionAddition&gt; reactionAdditions - list of reactions being added</p>

</dd>
<dt id="Definition88">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a gapfillsolution_id
objective has a value which is a float
integrated has a value which is a bool
biomassRemovals has a value which is a reference to a list where each element is a biomassRemoval
mediaAdditions has a value which is a reference to a list where each element is a mediaAddition
reactionAdditions has a value which is a reference to a list where each element is a reactionAddition</pre>

</dd>
</dl>

<h2 id="GapFill">GapFill</h2>

<dl>

<dt id="Description176">Description</dt>
<dd>

<p>Data structures for gapfilling analysis</p>

<p>gapfill_id id - ID of gapfill analysis workspace_id workspace - workspace containing gapfill analysis fbamodel_id model - ID of model being gapfilled workspace_id model_workspace - workspace containing model bool isComplete - indicates if gapfilling is complete GapfillingFormulation formulation - formulation of gapfilling analysis list&lt;GapFillSolution&gt; solutions - list of gapfilling solutions</p>

</dd>
<dt id="Definition89">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a gapfill_id
workspace has a value which is a workspace_id
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
isComplete has a value which is a bool
formulation has a value which is a GapfillingFormulation
solutions has a value which is a reference to a list where each element is a GapFillSolution</pre>

</dd>
</dl>

<h2 id="GapgenFormulation">GapgenFormulation</h2>

<dl>

<dt id="Description177">Description</dt>
<dd>

<p>******************************************************************************** Gap Generation type definition ********************************************************************************</p>

</dd>
<dt id="Definition90">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
formulation has a value which is an FBAFormulation
refmedia has a value which is a media_id
refmedia_workspace has a value which is a workspace_id
num_solutions has a value which is an int
nomediahyp has a value which is a bool
nobiomasshyp has a value which is a bool
nogprhyp has a value which is a bool
nopathwayhyp has a value which is a bool</pre>

</dd>
</dl>

<h2 id="reactionRemoval">reactionRemoval</h2>

<dl>

<dt id="Description178">Description</dt>
<dd>

<p>Reactions removed in gapgen solution</p>

<p>modelreaction_id reaction - ID of the removed reaction string direction - direction of reaction removed in gapgen solution string equation - stoichiometry of removed reaction in terms of compound IDs string definition - stoichiometry of removed reaction in terms of compound names</p>

</dd>
<dt id="Definition91">Definition</dt>
<dd>



<pre>
a reference to a list containing 4 items:
0: (reaction) a modelreaction_id
1: (direction) a string
2: (equation) a string
3: (definition) a string</pre>

</dd>
</dl>

<h2 id="biomassAddition">biomassAddition</h2>

<dl>

<dt id="Description179">Description</dt>
<dd>

<p>Compounds added to biomass in gapgen solution</p>

<p>compound_id compound - ID of biomass compound added string name - name of biomass compound added</p>

</dd>
<dt id="Definition92">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: (compound) a compound_id
1: (name) a string</pre>

</dd>
</dl>

<h2 id="mediaRemoval">mediaRemoval</h2>

<dl>

<dt id="Description180">Description</dt>
<dd>

<p>Media components removed in gapgen solution</p>

<p>compound_id compound - ID of media component removed string name - name of media component removed</p>

</dd>
<dt id="Definition93">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: (compound) a compound_id
1: (name) a string</pre>

</dd>
</dl>

<h2 id="GapgenSolution">GapgenSolution</h2>

<dl>

<dt id="Description181">Description</dt>
<dd>

<p>Data structures for gap generation solution</p>

<p>gapgensolution_id id - ID of gapgen solution float objective - cost of gapgen solution list&lt;biomassAddition&gt; biomassAdditions - list of components added to biomass list&lt;mediaRemoval&gt; mediaRemovals - list of media components removed list&lt;reactionRemoval&gt; reactionRemovals - list of reactions removed</p>

</dd>
<dt id="Definition94">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a gapgensolution_id
objective has a value which is a float
biomassAdditions has a value which is a reference to a list where each element is a biomassAddition
mediaRemovals has a value which is a reference to a list where each element is a mediaRemoval
reactionRemovals has a value which is a reference to a list where each element is a reactionRemoval</pre>

</dd>
</dl>

<h2 id="GapGen">GapGen</h2>

<dl>

<dt id="Description182">Description</dt>
<dd>

<p>Data structures for gap generation analysis</p>

<p>gapgen_id id - ID of gapgen object workspace_id workspace - workspace containing gapgen object fbamodel_id model - ID of model being gap generated workspace_id model_workspace - workspace containing model bool isComplete - flag indicating if gap generation is complete GapgenFormulation formulation - formulation of gap generation analysis list&lt;GapgenSolution&gt; solutions - list of gap generation solutions</p>

</dd>
<dt id="Definition95">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a gapgen_id
workspace has a value which is a workspace_id
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
isComplete has a value which is a bool
formulation has a value which is a GapgenFormulation
solutions has a value which is a reference to a list where each element is a GapgenSolution</pre>

</dd>
</dl>

<h2 id="Phenotype">Phenotype</h2>

<dl>

<dt id="Description183">Description</dt>
<dd>

<p>******************************************************************************** Phenotype type definitions ********************************************************************************</p>

</dd>
<dt id="Definition96">Definition</dt>
<dd>



<pre>
a reference to a list containing 6 items:
0: (geneKO) a reference to a list where each element is a feature_id
1: (baseMedia) a media_id
2: (media_workspace) a workspace_id
3: (additionalCpd) a reference to a list where each element is a compound_id
4: (normalizedGrowth) a float
5: (label) a string</pre>

</dd>
</dl>

<h2 id="PhenotypeSet">PhenotypeSet</h2>

<dl>

<dt id="Description184">Description</dt>
<dd>

<p>Data structures for set of growth phenotype observations</p>

<p>phenotype_set_id id - ID of the phenotype set genome_id genome - ID of the genome for the strain used with the growth phenotypes workspace_id genome_workspace - workspace containing the genome object list&lt;Phenotype&gt; phenotypes - list of phenotypes included in the phenotype set string importErrors - list of errors encountered during the import of the phenotype set string source - source of the phenotype set string name - name of the phenotype set</p>

</dd>
<dt id="Definition97">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a phenotype_set_id
genome has a value which is a genome_id
genome_workspace has a value which is a workspace_id
phenotypes has a value which is a reference to a list where each element is a Phenotype
source has a value which is a string
name has a value which is a string
importErrors has a value which is a string</pre>

</dd>
</dl>

<h2 id="phenotypeSimulationSet_id">phenotypeSimulationSet_id</h2>

<dl>

<dt id="Description185">Description</dt>
<dd>

<p>ID of the phenotype simulation object</p>

</dd>
<dt id="Definition98">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="PhenotypeSimulation">PhenotypeSimulation</h2>

<dl>

<dt id="Description186">Description</dt>
<dd>

<p>Data structures for a phenotype simulation</p>

<p>Phenotype phenotypeData - actual phenotype data simulated float simulatedGrowth - actual simulated growth rate float simulatedGrowthFraction - fraction of wildtype simulated growth rate string phenoclass - class of the phenotype simulation (i.e. &#39;CP&#39; - correct positive, &#39;CN&#39; - correct negative, &#39;FP&#39; - false positive, &#39;FN&#39; - false negative)</p>

</dd>
<dt id="Definition99">Definition</dt>
<dd>



<pre>
a reference to a list containing 4 items:
0: (phenotypeData) a Phenotype
1: (simulatedGrowth) a float
2: (simulatedGrowthFraction) a float
3: (phenoclass) a string</pre>

</dd>
</dl>

<h2 id="PhenotypeSimulationSet">PhenotypeSimulationSet</h2>

<dl>

<dt id="Description187">Description</dt>
<dd>

<p>Data structures for phenotype simulations of a set of phenotype data</p>

<p>phenotypeSimulationSet_id id - ID for the phenotype simulation set object fbamodel_id model - ID of the model used to simulate all phenotypes workspace_id model_workspace - workspace containing the model used for the simulation phenotype_set_id phenotypeSet - set of observed phenotypes that were simulated list&lt;PhenotypeSimulation&gt; phenotypeSimulations - list of simulated phenotypes</p>

</dd>
<dt id="Definition100">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a phenotypeSimulationSet_id
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
phenotypeSet has a value which is a phenotype_set_id
phenotypeSimulations has a value which is a reference to a list where each element is a PhenotypeSimulation</pre>

</dd>
</dl>

<h2 id="reactionSpecification">reactionSpecification</h2>

<dl>

<dt id="Description188">Description</dt>
<dd>

<p>Data structure for holding gapfill or gapgen solution reaction information</p>

<p>string direction - direction of gapfilled or gapgen reaction string reactionID - ID of gapfilled or gapgen reaction</p>

</dd>
<dt id="Definition101">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: (direction) a string
1: (reactionID) a string</pre>

</dd>
</dl>

<h2 id="PhenotypeSensitivityAnalysis">PhenotypeSensitivityAnalysis</h2>

<dl>

<dt id="Description189">Description</dt>
<dd>

<p>list&lt;string id, string solutionIndex, list&lt;reactionSpecification&gt; reactionList, list&lt;string&gt; biomassEdits,list&lt;tuple&lt;float simulatedGrowth,float simulatedGrowthFraction,string phenoclass&gt;&gt; PhenotypeSimulations&gt; reconciliationSolutionSimulations;</p>

</dd>
<dt id="Definition102">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
phenotypeSet has a value which is a phenotype_set_id
phenotypeSet_workspace has a value which is a workspace_id
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
phenotypes has a value which is a reference to a list where each element is a Phenotype
wildtypePhenotypeSimulations has a value which is a reference to a list where each element is a reference to a list containing 3 items:
0: (simulatedGrowth) a float
1: (simulatedGrowthFraction) a float
2: (phenoclass) a string</pre>

</dd>
</dl>

<h2 id="job_id">job_id</h2>

<dl>

<dt id="Description190">Description</dt>
<dd>

<p>******************************************************************************** Job object type definitions ********************************************************************************</p>

</dd>
<dt id="Definition103">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="CommandArguments">CommandArguments</h2>

<dl>

<dt id="Description191">Description</dt>
<dd>

<p>Object to hold the arguments to be submitted to the post process command</p>

</dd>
<dt id="Definition104">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="clusterjob">clusterjob</h2>

<dl>

<dt id="Description192">Description</dt>
<dd>

<p>Object to hold data required to run cluster job</p>

</dd>
<dt id="Definition105">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="JobObject">JobObject</h2>

<dl>

<dt id="Description193">Description</dt>
<dd>

<p>Data structures for a job object</p>

<p>job_id id - ID of the job object string type - type of the job string auth - authentication token of job owner string status - current status of job mapping&lt;string,string&gt; jobdata; string queuetime - time when job was queued string starttime - time when job started running string completetime - time when the job was completed string owner - owner of the job string queuecommand - command used to queue job</p>

</dd>
<dt id="Definition106">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a job_id
type has a value which is a string
auth has a value which is a string
status has a value which is a string
jobdata has a value which is a reference to a hash where the key is a string and the value is a string
queuetime has a value which is a string
starttime has a value which is a string
completetime has a value which is a string
owner has a value which is a string
queuecommand has a value which is a string</pre>

</dd>
</dl>

<h2 id="ETCNodes">ETCNodes</h2>

<dl>

<dt id="Description194">Description</dt>
<dd>

<p>******************************************************************************** ETC object type definitions ********************************************************************************</p>

</dd>
<dt id="Definition107">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
resp has a value which is a string
y has a value which is an int
x has a value which is an int
width has a value which is an int
height has a value which is an int
shape has a value which is a string
label has a value which is a string</pre>

</dd>
</dl>

<h2 id="ETCDiagramSpecs">ETCDiagramSpecs</h2>

<dl>

<dt id="Definition108">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
nodes has a value which is a reference to a list where each element is an ETCNodes
media has a value which is a string
growth has a value which is a string
organism has a value which is a string</pre>

</dd>
</dl>

<h2 id="TemplateReactions">TemplateReactions</h2>

<dl>

<dt id="Description195">Description</dt>
<dd>

<p>******************************************************************************** AutoRecon type definitions ********************************************************************************</p>

</dd>
<dt id="Definition109">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
reaction has a value which is a reaction_id
direction has a value which is a string
equation has a value which is a string
compartment has a value which is a compartment_id</pre>

</dd>
</dl>

<h2 id="ComplexReactions">ComplexReactions</h2>

<dl>

<dt id="Description196">Description</dt>
<dd>

<p>Information on complexes in a template model</p>

<pre><code>        complex_id complex - ID of the associated complex
        list&lt;TemplateReactions&gt; reactions - List of template models associated with complex</code></pre>

</dd>
<dt id="Definition110">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
complex has a value which is a complex_id
name has a value which is a string
reactions has a value which is a reference to a list where each element is a TemplateReactions</pre>

</dd>
</dl>

<h2 id="RoleComplexReactions">RoleComplexReactions</h2>

<dl>

<dt id="Description197">Description</dt>
<dd>

<p>Information on complexes in a template model</p>

<pre><code>        complex_id complex - ID of the associated complex
        list&lt;TemplateReactions&gt; reactions - List of template models associated with complex</code></pre>

</dd>
<dt id="Definition111">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
role has a value which is a role_id
name has a value which is a string
complexes has a value which is a reference to a list where each element is a ComplexReactions</pre>

</dd>
</dl>

<h2 id="ReactionDefinition">ReactionDefinition</h2>

<dl>

<dt id="Description198">Description</dt>
<dd>

<p>Reaction definition</p>

<pre><code>        reaction_id id - ID of reaction
        string name - name of reaction
        string definition - stoichiometric equation of reaction in terms of compound names</code></pre>

</dd>
<dt id="Definition112">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a reaction_id
name has a value which is a string
definition has a value which is a string</pre>

</dd>
</dl>

<h2 id="ReactionSynonyms">ReactionSynonyms</h2>

<dl>

<dt id="Description199">Description</dt>
<dd>

<p>Reaction synonyms</p>

<pre><code>        reaction_id primary - ID of primary reaction
        list&lt;ReactionDefinition&gt; synonyms - list of synonym reactions to the primary reaction (including itself)</code></pre>

</dd>
<dt id="Definition113">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
primary has a value which is a reaction_id
synonyms has a value which is a reference to a list where each element is a ReactionDefinition</pre>

</dd>
</dl>

<h2 id="ReactionSynonymsObject">ReactionSynonymsObject</h2>

<dl>

<dt id="Description200">Description</dt>
<dd>

<p>Reaction synonyms object</p>

<pre><code>        int version - version number of object
        biochemistry_id biochemistry - ID of associated biochemistry database
        workspace_id biochemistry_workspace - workspace with associated biochemistry database
        list&lt;ReactionSynonyms&gt; synonym_list - list of all reaction synonyms from a biochemistry database
        list&lt;ReactionDefinition&gt; excluded_list - list of reactions excluded because all compounds are cofactors</code></pre>

</dd>
<dt id="Definition114">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
version has a value which is an int
biochemistry has a value which is a biochemistry_id
biochemistry_workspace has a value which is a workspace_id
synonyms_list has a value which is a reference to a list where each element is a ReactionSynonyms
excluded_list has a value which is a reference to a list where each element is a ReactionDefinition</pre>

</dd>
</dl>

<h2 id="get_models_params">get_models_params</h2>

<dl>

<dt id="Description201">Description</dt>
<dd>

<p>******************************************************************************** Function definitions relating to data retrieval for Model Objects ********************************************************************************</p>

</dd>
<dt id="Definition115">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
models has a value which is a reference to a list where each element is a fbamodel_id
workspaces has a value which is a reference to a list where each element is a workspace_id
auth has a value which is a string
id_type has a value which is a string</pre>

</dd>
</dl>

<h2 id="get_fbas_params">get_fbas_params</h2>

<dl>

<dt id="Description202">Description</dt>
<dd>

<p>Input parameters for the &quot;get_fbas&quot; function.</p>

<pre><code>        list&lt;fba_id&gt; fbas - a list of the FBA study IDs for the FBA studies to be returned (a required argument)
        list&lt;workspace_id&gt; workspaces - a list of the workspaces contianing the FBA studies to be returned (a required argument)
string id_type - the type of ID that should be used in the output data (a optional argument; default is &#39;ModelSEED&#39;)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition116">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
fbas has a value which is a reference to a list where each element is a fba_id
workspaces has a value which is a reference to a list where each element is a workspace_id
auth has a value which is a string
id_type has a value which is a string</pre>

</dd>
</dl>

<h2 id="get_gapfills_params">get_gapfills_params</h2>

<dl>

<dt id="Description203">Description</dt>
<dd>

<p>Input parameters for the &quot;get_gapfills&quot; function.</p>

<pre><code>        list&lt;gapfill_id&gt; gapfills - a list of the gapfill study IDs for the gapfill studies to be returned (a required argument)
        list&lt;workspace_id&gt; workspaces - a list of the workspaces contianing the gapfill studies to be returned (a required argument)
string id_type - the type of ID that should be used in the output data (a optional argument; default is &#39;ModelSEED&#39;)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition117">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
gapfills has a value which is a reference to a list where each element is a gapfill_id
workspaces has a value which is a reference to a list where each element is a workspace_id
auth has a value which is a string
id_type has a value which is a string</pre>

</dd>
</dl>

<h2 id="get_gapgens_params">get_gapgens_params</h2>

<dl>

<dt id="Description204">Description</dt>
<dd>

<p>Input parameters for the &quot;get_gapgens&quot; function.</p>

<pre><code>        list&lt;gapgen_id&gt; gapgens - a list of the gapgen study IDs for the gapgen studies to be returned (a required argument)
        list&lt;workspace_id&gt; workspaces - a list of the workspaces contianing the gapgen studies to be returned (a required argument)
string id_type - the type of ID that should be used in the output data (a optional argument; default is &#39;ModelSEED&#39;)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition118">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
gapgens has a value which is a reference to a list where each element is a gapgen_id
workspaces has a value which is a reference to a list where each element is a workspace_id
auth has a value which is a string
id_type has a value which is a string</pre>

</dd>
</dl>

<h2 id="get_reactions_params">get_reactions_params</h2>

<dl>

<dt id="Description205">Description</dt>
<dd>

<p>Input parameters for the &quot;get_reactions&quot; function.</p>

<pre><code>        list&lt;reaction_id&gt; reactions - a list of the reaction IDs for the reactions to be returned (a required argument)
        string id_type - the type of ID that should be used in the output data (a optional argument; default is &#39;ModelSEED&#39;)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition119">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
reactions has a value which is a reference to a list where each element is a reaction_id
auth has a value which is a string
id_type has a value which is a string</pre>

</dd>
</dl>

<h2 id="get_compounds_params">get_compounds_params</h2>

<dl>

<dt id="Description206">Description</dt>
<dd>

<p>Input parameters for the &quot;get_compounds&quot; function. list&lt;compound_id&gt; compounds - a list of the compound IDs for the compounds to be returned (a required argument) string id_type - the type of ID that should be used in the output data (a optional argument; default is &#39;ModelSEED&#39;) string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</p>

</dd>
<dt id="Definition120">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
compounds has a value which is a reference to a list where each element is a compound_id
auth has a value which is a string
id_type has a value which is a string</pre>

</dd>
</dl>

<h2 id="get_alias_params">get_alias_params</h2>

<dl>

<dt id="Description207">Description</dt>
<dd>

<p>Input parameters for the get_alias function</p>

<pre><code>                string object_type    - The type of object (e.g. Compound or Reaction)
                string input_id_type - The type (e.g. ModelSEED) of alias to be inputted
                string output_id_type - The type (e.g. KEGG) of alias to be outputted
                list&lt;string&gt; input_ids - A list of input IDs
                string auth; - The authentication token of the KBase account (optional)</code></pre>

</dd>
<dt id="Definition121">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
object_type has a value which is a string
input_id_type has a value which is a string
output_id_type has a value which is a string
input_ids has a value which is a reference to a list where each element is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="get_alias_outputs">get_alias_outputs</h2>

<dl>

<dt id="Description208">Description</dt>
<dd>

<p>Output for get_alias function</p>

<pre><code>              string original_id - The original ID
              list&lt;string&gt; aliases - Aliases for the original ID in the new format</code></pre>

</dd>
<dt id="Definition122">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
original_id has a value which is a string
aliases has a value which is a reference to a list where each element is a string</pre>

</dd>
</dl>

<h2 id="get_aliassets_params">get_aliassets_params</h2>

<dl>

<dt id="Description209">Description</dt>
<dd>

<p>Input parameters for the get_aliassets function</p>

<pre><code>              string auth; - The authentication token of the KBase account (optional)
              string object_type; - The type of object (e.g. Compound or Reaction)</code></pre>

</dd>
<dt id="Definition123">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
object_type has a value which is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="get_media_params">get_media_params</h2>

<dl>

<dt id="Description210">Description</dt>
<dd>

<p>Input parameters for the &quot;get_media&quot; function.</p>

<pre><code>        list&lt;media_id&gt; medias - a list of the media IDs for the media to be returned (a required argument)
        string id_type - the type of ID that should be used in the output data (a optional argument; default is &#39;ModelSEED&#39;)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition124">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
medias has a value which is a reference to a list where each element is a media_id
workspaces has a value which is a reference to a list where each element is a workspace_id
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="get_biochemistry_params">get_biochemistry_params</h2>

<dl>

<dt id="Description211">Description</dt>
<dd>

<p>Input parameters for the &quot;get_biochemistry&quot; function.</p>

<pre><code>        biochemistry_id biochemistry - ID of the biochemistry database to be returned (a required argument)
        workspace_id biochemistry_workspace - workspace containing the biochemistry database to be returned (a required argument)
        string id_type - the type of ID that should be used in the output data (a optional argument; default is &#39;ModelSEED&#39;)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition125">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
biochemistry has a value which is a biochemistry_id
biochemistry_workspace has a value which is a workspace_id
id_type has a value which is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="import_probanno_params">import_probanno_params</h2>

<dl>

<dt id="Description212">Description</dt>
<dd>

<p>******************************************************************************** Code relating to reconstruction of metabolic models ********************************************************************************</p>

</dd>
<dt id="Definition126">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
probanno has a value which is a probanno_id
workspace has a value which is a workspace_id
genome has a value which is a genome_id
genome_workspace has a value which is a workspace_id
annotationProbabilities has a value which is a reference to a list where each element is an annotationProbability
ignore_errors has a value which is a bool
auth has a value which is a string
overwrite has a value which is a bool</pre>

</dd>
</dl>

<h2 id="Genome_uid1">Genome_uid</h2>

<dl>

<dt id="Description213">Description</dt>
<dd>

<p>Input parameters for the &quot;genome_object_to_workspace&quot; function.</p>

<pre><code>        Genome_uid uid - ID to use when saving genome to workspace
        GenomeObject genomeobj - full genome typed object to be loaded into the workspace (a required argument)
        workspace_id workspace - ID of the workspace into which the genome typed object is to be loaded (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition127">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="genome_object_to_workspace_params">genome_object_to_workspace_params</h2>

<dl>

<dt id="Definition128">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
uid has a value which is a Genome_uid
genomeobj has a value which is a GenomeObject
workspace has a value which is a workspace_id
auth has a value which is a string
overwrite has a value which is a bool</pre>

</dd>
</dl>

<h2 id="genome_to_workspace_params">genome_to_workspace_params</h2>

<dl>

<dt id="Description214">Description</dt>
<dd>

<p>Input parameters for the &quot;genome_to_workspace&quot; function.</p>

<pre><code>        genome_id genome - ID of the CDM genome that is to be loaded into the workspace (a required argument)
        string sourceLogin - login to pull private genome from source database
        string sourcePassword - password to pull private genome from source database
        string source - Source database for genome (i.e. seed, rast, kbase)
        workspace_id workspace - ID of the workspace into which the genome typed object is to be loaded (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)
        Genome_uid uid - ID to use when saving genome to workspace</code></pre>

</dd>
<dt id="Definition129">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
genome has a value which is a genome_id
workspace has a value which is a workspace_id
sourceLogin has a value which is a string
sourcePassword has a value which is a string
source has a value which is a string
auth has a value which is a string
overwrite has a value which is a bool
uid has a value which is a Genome_uid</pre>

</dd>
</dl>

<h2 id="domains_to_workspace_params">domains_to_workspace_params</h2>

<dl>

<dt id="Description215">Description</dt>
<dd>

<p>Input parameters for the &quot;domains_to_workspace&quot; function.</p>

<pre><code>        genome_id genome - ID of the workspace genome to fetch domains for (a required argument)
        string output_id - ID in which the domains are to be saved (default is genome ID plus &quot;.dom.0&quot;)
        workspace_id workspace - ID of the workspace into which the domains are to be loaded (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition130">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
genome has a value which is a genome_id
output_id has a value which is a string
workspace has a value which is a workspace_id
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="compute_domains_params">compute_domains_params</h2>

<dl>

<dt id="Description216">Description</dt>
<dd>

<p>Input parameters for the &quot;compute_domains_params&quot; function. string genome; string genome_workspace; list&lt;tuple&lt;string,string&gt;&gt; proteins; workspace_id workspace;</p>

</dd>
<dt id="Definition131">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
genome has a value which is a string
genome_workspace has a value which is a string
proteins has a value which is a reference to a list where each element is a reference to a list containing 2 items:
0: a string
1: a stringworkspace has a value which is a workspace_id</pre>

</dd>
</dl>

<h2 id="translation">translation</h2>

<dl>

<dt id="Description217">Description</dt>
<dd>

<p>A link between a KBase gene ID and the ID for the same gene in another database</p>

<pre><code>        string foreign_id - ID of the gene in another database
        feature_id feature - ID of the gene in KBase</code></pre>

</dd>
<dt id="Definition132">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: (foreign_id) a string
1: (feature) a feature_id</pre>

</dd>
</dl>

<h2 id="add_feature_translation_params">add_feature_translation_params</h2>

<dl>

<dt id="Description218">Description</dt>
<dd>

<p>Input parameters for the &quot;add_feature_translation&quot; function.</p>

<pre><code>        genome_id genome - ID of the genome into which the new aliases are to be loaded (a required argument)
        workspace_id workspace - ID of the workspace containing the target genome (a required argument)
        list&lt;translation&gt; translations - list of translations between KBase gene IDs and gene IDs in another database (a required argument)
        string id_type - type of the IDs being loaded (e.g. KEGG, NCBI) (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition133">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
genome has a value which is a genome_id
workspace has a value which is a workspace_id
translations has a value which is a reference to a list where each element is a translation
id_type has a value which is a string
auth has a value which is a string
overwrite has a value which is a bool</pre>

</dd>
</dl>

<h2 id="genome_to_fbamodel_params">genome_to_fbamodel_params</h2>

<dl>

<dt id="Description219">Description</dt>
<dd>

<p>Input parameters for the &quot;genome_to_fbamodel&quot; function.</p>

<pre><code>        genome_id genome - ID of the genome for which a model is to be built (a required argument)
        workspace_id genome_workspace - ID of the workspace containing the target genome (an optional argument; default is the workspace argument)
        template_id templatemodel - 
        workspace_id templatemodel_workspace - 
        bool probannoOnly - a boolean indicating if only the probabilistic annotation should be used in building the model (an optional argument; default is &#39;0&#39;)
        fbamodel_id model - ID that should be used for the newly constructed model (an optional argument; default is &#39;undef&#39;)
        bool coremodel - indicates that a core model should be constructed instead of a genome scale model (an optional argument; default is &#39;0&#39;)
        workspace_id workspace - ID of the workspace where the newly developed model will be stored; also the default assumed workspace for input objects (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition134">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
genome has a value which is a genome_id
genome_workspace has a value which is a workspace_id
templatemodel has a value which is a template_id
templatemodel_workspace has a value which is a workspace_id
model has a value which is a fbamodel_id
coremodel has a value which is a bool
workspace has a value which is a workspace_id
auth has a value which is a string
fulldb has a value which is a bool</pre>

</dd>
</dl>

<h2 id="translate_fbamodel_params">translate_fbamodel_params</h2>

<dl>

<dt id="Description220">Description</dt>
<dd>

<p>Input parameters for the &quot;translate_fbamodel&quot; function.</p>

<pre><code>        gencomp
        gencomp_workspace
        fbamodel_id model;
        fbamodel_id model_workspace;</code></pre>

</dd>
<dt id="Definition135">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
protcomp has a value which is a string
protcomp_workspace has a value which is a string
model has a value which is a string
model_workspace has a value which is a string
workspace has a value which is a workspace_id</pre>

</dd>
</dl>

<h2 id="build_pangenome_params">build_pangenome_params</h2>

<dl>

<dt id="Description221">Description</dt>
<dd>

<p>Input parameters for the &quot;translate_fbamodel&quot; function.</p>

<pre><code>        gencomp
        gencomp_workspace
        fbamodel_id model;
        fbamodel_id model_workspace;</code></pre>

</dd>
<dt id="Definition136">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
genomes has a value which is a reference to a list where each element is a string
genome_workspace has a value which is a reference to a list where each element is a string
workspace has a value which is a workspace_id</pre>

</dd>
</dl>

<h2 id="heat_map_matrix">heat_map_matrix</h2>

<dl>

<dt id="Definition137">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
is_refs has a value which is a bool
labels has a value which is a reference to a list where each element is a string
matrix has a value which is a reference to a list where each element is a reference to a list where each element is a float</pre>

</dd>
</dl>

<h2 id="genome_heatmap_from_pangenome_params">genome_heatmap_from_pangenome_params</h2>

<dl>

<dt id="Definition138">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
pangenome has a value which is a string
pangenome_workspace has a value which is a string
workspace has a value which is a string</pre>

</dd>
</dl>

<h2 id="ortholog_data">ortholog_data</h2>

<dl>

<dt id="Description222">Description</dt>
<dd>

<p>gene ID,gene ref,protein sequence,function,score</p>

</dd>
<dt id="Definition139">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
gene_data has a value which is a reference to a list where each element is a reference to a list containing 5 items:
0: a string
1: a string
2: a string
3: a string
4: a floatprotein_heatmap has a value which is a heat_map_matrix</pre>

</dd>
</dl>

<h2 id="ortholog_family_from_pangenome_params">ortholog_family_from_pangenome_params</h2>

<dl>

<dt id="Definition140">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
pangenome has a value which is a string
pangenome_workspace has a value which is a string
orthologid has a value which is a string
workspace has a value which is a string</pre>

</dd>
</dl>

<h2 id="pangenome_to_proteome_comparison_params">pangenome_to_proteome_comparison_params</h2>

<dl>

<dt id="Definition141">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
pangenome has a value which is a string
pangenome_workspace has a value which is a string
outputid has a value which is a string
workspace has a value which is a string</pre>

</dd>
</dl>

<h2 id="import_fbamodel_params">import_fbamodel_params</h2>

<dl>

<dt id="Description223">Description</dt>
<dd>

<p>Input parameters for the &quot;import_fbamodel&quot; function.</p>

<pre><code>        genome_id genome - ID of the genome for which a model is to be built (a required argument)
        workspace_id genome_workspace - ID of the workspace containing the target genome (an optional argument; default is the workspace argument)
        string biomass - biomass equation for model (an essential argument)
        list&lt;tuple&lt;string id,string direction,string compartment,string gpr&gt; reactions - list of reactions to appear in imported model (an essential argument)
        fbamodel_id model - ID that should be used for the newly imported model (an optional argument; default is &#39;undef&#39;)
        workspace_id workspace - ID of the workspace where the newly developed model will be stored; also the default assumed workspace for input objects (a required argument)
        bool ignore_errors - ignores missing genes or reactions and imports model anyway
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition142">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
genome has a value which is a genome_id
genome_workspace has a value which is a workspace_id
biomass has a value which is a string
reactions has a value which is a reference to a list where each element is a reference to a list containing 4 items:
0: (id) a string
1: (direction) a string
2: (compartment) a string
3: (gpr) a stringmodel has a value which is a fbamodel_id
workspace has a value which is a workspace_id
ignore_errors has a value which is a bool
auth has a value which is a string
overwrite has a value which is a bool</pre>

</dd>
</dl>

<h2 id="export_fbamodel_params">export_fbamodel_params</h2>

<dl>

<dt id="Description224">Description</dt>
<dd>

<p>Input parameters for the &quot;export_fbamodel&quot; function.</p>

<pre><code>        fbamodel_id model - ID of the model to be exported (a required argument)
        workspace_id workspace - workspace containing the model to be exported (a required argument)
        fba_id fba - A FBA object related to the model. (an optional argument)
        string format - format to which the model should be exported (sbml, html, json, readable, cytoseed) (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition143">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
workspace has a value which is a workspace_id
fbas has a value which is a reference to a list where each element is a fba_id
format has a value which is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="export_object_params">export_object_params</h2>

<dl>

<dt id="Description225">Description</dt>
<dd>

<p>Input parameters for the &quot;export_object&quot; function.</p>

<pre><code>        workspace_ref reference - reference of object to print in html (a required argument)
        string type - type of the object to be exported (a required argument)
        string format - format to which data should be exported (an optional argument; default is html)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition144">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
reference has a value which is a workspace_ref
type has a value which is a string
format has a value which is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="export_genome_params">export_genome_params</h2>

<dl>

<dt id="Description226">Description</dt>
<dd>

<p>Input parameters for the &quot;export_genome&quot; function.</p>

<pre><code>        genome_id genome - ID of the genome to be exported (a required argument)
        workspace_id workspace - workspace containing the model to be exported (a required argument)
        string format - format to which the model should be exported (sbml, html, json, readable, cytoseed) (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition145">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
genome has a value which is a genome_id
workspace has a value which is a workspace_id
format has a value which is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="adjust_model_reaction_params">adjust_model_reaction_params</h2>

<dl>

<dt id="Description227">Description</dt>
<dd>

<p>Input parameters for the &quot;adjust_model_reaction&quot; function.</p>

<pre><code>        fbamodel_id model - ID of model to be adjusted
        workspace_id workspace - workspace containing model to be adjusted
        list&lt;reaction_id&gt; reaction - List of IDs of reactions to be added, removed, or adjusted
        list&lt;string&gt; direction - directions to set for reactions being added or adjusted
        list&lt;compartment_id&gt; compartment - IDs of compartment containing reaction being added or adjusted
        list&lt;int&gt; compartmentIndex - indexes of compartment containing reaction being altered or adjusted
        list&lt;string&gt; gpr - array specifying gene-protein-reaction association(s)
        bool removeReaction - boolean indicating listed reaction(s) should be removed
        bool addReaction - boolean indicating reaction(s) should be added
        bool overwrite - boolean indicating whether or not to overwrite model object in the workspace
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)

        For all of the lists above, if only one element is specified it is assumed the user wants to apply the same
        to all the listed reactions.</code></pre>

</dd>
<dt id="Definition146">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
workspace has a value which is a workspace_id
reaction has a value which is a reference to a list where each element is a reaction_id
direction has a value which is a reference to a list where each element is a string
compartment has a value which is a reference to a list where each element is a compartment_id
compartmentIndex has a value which is a reference to a list where each element is an int
gpr has a value which is a reference to a list where each element is a string
removeReaction has a value which is a bool
addReaction has a value which is a bool
overwrite has a value which is a bool
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="adjust_biomass_reaction_params">adjust_biomass_reaction_params</h2>

<dl>

<dt id="Description228">Description</dt>
<dd>

<p>Input parameters for the &quot;adjust_biomass_reaction&quot; function.</p>

<pre><code>        fbamodel_id model - ID of model to be adjusted
        workspace_id workspace - workspace containing model to be adjusted
        biomass_id biomass - ID of biomass reaction to adjust
        list&lt;float&gt; coefficients - coefficient of biomass compound
        list&lt;compound_id&gt; compounds - ID of biomass compound to adjust in biomass
        list&lt;compartment_id&gt; compartments - ID of compartment containing compound to adjust in biomass
        list&lt;int&gt; compartmentIndecies - index of compartment containing compound to adjust in biomass
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition147">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
workspace has a value which is a workspace_id
biomass has a value which is a biomass_id
coefficients has a value which is a reference to a list where each element is a float
compounds has a value which is a reference to a list where each element is a compound_id
compartments has a value which is a reference to a list where each element is a compartment_id
compartmentIndecies has a value which is a reference to a list where each element is an int
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="addmedia_params">addmedia_params</h2>

<dl>

<dt id="Description229">Description</dt>
<dd>

<p>******************************************************************************** Code relating to flux balance analysis ********************************************************************************</p>

</dd>
<dt id="Definition148">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
media has a value which is a media_id
workspace has a value which is a workspace_id
name has a value which is a string
isDefined has a value which is a bool
isMinimal has a value which is a bool
type has a value which is a string
compounds has a value which is a reference to a list where each element is a string
concentrations has a value which is a reference to a list where each element is a float
maxflux has a value which is a reference to a list where each element is a float
minflux has a value which is a reference to a list where each element is a float
overwrite has a value which is a bool
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="export_media_params">export_media_params</h2>

<dl>

<dt id="Description230">Description</dt>
<dd>

<p>Input parameters for the &quot;export_media&quot; function.</p>

<pre><code>        media_id media - ID of the media to be exported (a required argument)
        workspace_id workspace - workspace containing the media to be exported (a required argument)
        string format - format to which the media should be exported (html, json, readable) (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition149">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
media has a value which is a media_id
workspace has a value which is a workspace_id
format has a value which is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="runfba_params">runfba_params</h2>

<dl>

<dt id="Description231">Description</dt>
<dd>

<p>Input parameters for the &quot;addmedia&quot; function.</p>

<pre><code>        fbamodel_id model - ID of the model that FBA should be run on (a required argument)
        workspace_id model_workspace - workspace where model for FBA should be run (an optional argument; default is the value of the workspace argument)
        FBAFormulation formulation - a hash specifying the parameters for the FBA study (an optional argument)
        bool fva - a flag indicating if flux variability should be run (an optional argument: default is &#39;0&#39;)
        bool simulateko - a flag indicating if flux variability should be run (an optional argument: default is &#39;0&#39;)
        bool minimizeflux - a flag indicating if flux variability should be run (an optional argument: default is &#39;0&#39;)
        bool findminmedia - a flag indicating if flux variability should be run (an optional argument: default is &#39;0&#39;)
        string notes - a string of notes to attach to the FBA study (an optional argument; defaul is &#39;&#39;)
        fba_id fba - ID under which the FBA results should be saved (an optional argument; defaul is &#39;undef&#39;)
        workspace_id workspace - workspace where FBA results will be saved (a required argument)
        bool add_to_model - a flag indicating if the FBA study should be attached to the model to support viewing results (an optional argument: default is &#39;0&#39;)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition150">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
formulation has a value which is an FBAFormulation
fva has a value which is a bool
simulateko has a value which is a bool
minimizeflux has a value which is a bool
findminmedia has a value which is a bool
notes has a value which is a string
fba has a value which is a fba_id
workspace has a value which is a workspace_id
auth has a value which is a string
overwrite has a value which is a bool
add_to_model has a value which is a bool</pre>

</dd>
</dl>

<h2 id="quantitative_optimization_params">quantitative_optimization_params</h2>

<dl>

<dt id="Description232">Description</dt>
<dd>

<p>Input parameters for the &quot;addmedia&quot; function.</p>

<pre><code>        fbamodel_id model - ID of the model that FBA should be run on (a required argument)
        workspace_id model_workspace - workspace where model for FBA should be run (an optional argument; default is the value of the workspace argument)
        FBAFormulation formulation - a hash specifying the parameters for the FBA study (an optional argument)
        fbamodel_id outputid - ID of model to be saved with quantitative optimization solution (an optional argument)
        workspace_id workspace - workspace where all output objects will be saved (a required argument)
        string biomass - ID of biomass reaction as target for quantitative optimization (an optional argument)</code></pre>

</dd>
<dt id="Definition151">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
formulation has a value which is an FBAFormulation
outputid has a value which is a fbamodel_id
workspace has a value which is a workspace_id
biomass has a value which is a string</pre>

</dd>
</dl>

<h2 id="generate_model_stats_params">generate_model_stats_params</h2>

<dl>

<dt id="Description233">Description</dt>
<dd>

<p>Input parameters for the &quot;generate_model_stats&quot; function.</p>

<pre><code>        fbamodel_id model - ID of the models that FBA should be run on (a required argument)
        workspace_id model_workspace - workspaces where model for FBA should be run (an optional argument; default is the value of the workspace argument)</code></pre>

</dd>
<dt id="Definition152">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id</pre>

</dd>
</dl>

<h2 id="subsystem_statistics">subsystem_statistics</h2>

<dl>

<dt id="Definition153">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
name has a value which is a string
class has a value which is a string
subclass has a value which is a string
genes has a value which is an int
reactions has a value which is an int
model_genes has a value which is an int
minimal_essential_genes has a value which is an int
complete_essential_genes has a value which is an int
minimal_essential_reactions has a value which is an int
complete_essential_reactions has a value which is an int
minimal_blocked_reactions has a value which is an int
complete_blocked_reactions has a value which is an int
minimal_variable_reactions has a value which is an int
complete_variable_reactions has a value which is an int</pre>

</dd>
</dl>

<h2 id="model_statistics">model_statistics</h2>

<dl>

<dt id="Definition154">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
total_reactions has a value which is an int
total_genes has a value which is an int
total_compounds has a value which is an int
extracellular_compounds has a value which is an int
intracellular_compounds has a value which is an int
transport_reactions has a value which is an int
subsystem_reactions has a value which is an int
subsystem_genes has a value which is an int
spontaneous_reactions has a value which is an int
reactions_with_genes has a value which is an int
gapfilled_reactions has a value which is an int
model_genes has a value which is an int
minimal_essential_genes has a value which is an int
complete_essential_genes has a value which is an int
minimal_essential_reactions has a value which is an int
complete_essential_reactions has a value which is an int
minimal_blocked_reactions has a value which is an int
complete_blocked_reactions has a value which is an int
minimal_variable_reactions has a value which is an int
complete_variable_reactions has a value which is an int
growth_complete_media has a value which is a bool
growth_minimal_media has a value which is a bool
subsystems has a value which is a reference to a list where each element is a subsystem_statistics</pre>

</dd>
</dl>

<h2 id="minimize_reactions_params">minimize_reactions_params</h2>

<dl>

<dt id="Description234">Description</dt>
<dd>

<p>Input parameters for the &quot;minimize_reactions&quot; function.</p>

<pre><code>        fbamodel_id model - ID of the model that FBA should be run on (a required argument)
        workspace_id model_workspace - workspace where model for FBA should be run (an optional argument; default is the value of the workspace argument)
        workspace_id workspace - workspace where FBA results will be saved (a required argument)
        FBAFormulation formulation - a hash specifying the parameters for the FBA study (an optional argument)
        list&lt;string&gt; reactions - list of model reactions to be minimized (an optional argument)
        bool all_model_reactions - minimize all reactions in the model (default is &#39;false&#39; unless &#39;reactions&#39; list is empty)
        mapping&lt;string,float&gt; reaction_costs - hash of costs for each reaction to be minimized (default is &#39;1&#39; for every reaction)
        fba_id output_id - id to which FBA result should be saved</code></pre>

</dd>
<dt id="Definition155">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
workspace has a value which is a workspace_id
formulation has a value which is an FBAFormulation
reactions has a value which is a reference to a list where each element is a string
all_model_reactions has a value which is a bool
reaction_costs has a value which is a reference to a hash where the key is a string and the value is a float
output_id has a value which is a fba_id</pre>

</dd>
</dl>

<h2 id="export_fba_params">export_fba_params</h2>

<dl>

<dt id="Description235">Description</dt>
<dd>

<p>Input parameters for the &quot;addmedia&quot; function.</p>

<pre><code>        fba_id fba - ID of the FBA study to be exported (a required argument)
        workspace_id workspace - workspace where FBA study is stored (a required argument)
        string format - format to which the FBA study should be exported (i.e. html, json, readable) (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition156">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
fba has a value which is a fba_id
workspace has a value which is a workspace_id
format has a value which is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="import_phenotypes_params">import_phenotypes_params</h2>

<dl>

<dt id="Description236">Description</dt>
<dd>

<p>******************************************************************************** Code relating to phenotype simulation and reconciliation ********************************************************************************</p>

</dd>
<dt id="Definition157">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
phenotypeSet has a value which is a phenotype_set_id
workspace has a value which is a workspace_id
genome has a value which is a genome_id
genome_workspace has a value which is a workspace_id
phenotypes has a value which is a reference to a list where each element is a Phenotype
name has a value which is a string
source has a value which is a string
ignore_errors has a value which is a bool
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="simulate_phenotypes_params">simulate_phenotypes_params</h2>

<dl>

<dt id="Description237">Description</dt>
<dd>

<p>Input parameters for the &quot;simulate_phenotypes&quot; function.</p>

<pre><code>        fbamodel_id model - ID of the model to be used for the simulation (a required argument)
        workspace_id model_workspace - workspace containing the model for the simulation (an optional argument: default is value of workspace argument)
        phenotype_set_id phenotypeSet - ID of the phenotypes set to be simulated (a required argument)
        workspace_id phenotypeSet_workspace - workspace containing the phenotype set to be simulated (an optional argument: default is value of workspace argument)
        FBAFormulation formulation - parameters for the simulation flux balance analysis (an optional argument: default is &#39;undef&#39;)
        string notes - string of notes to associate with the phenotype simulation (an optional argument: default is &#39;&#39;)
        phenotypeSimulationSet_id phenotypeSimulationSet - ID of the phenotype simulation set to be generated (an optional argument: default is &#39;undef&#39;)
        workspace_id workspace - workspace where the phenotype simulation set should be saved (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)
        bool all_transporters - Set to TRUE if you want to add transporters for ALL media in the phenotypeset before simulating
        bool positive_transporters - Set to TRUE if you want to add transporters for POSITIVE (non-zero growth) media only before simulating</code></pre>

</dd>
<dt id="Definition158">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
phenotypeSet has a value which is a phenotype_set_id
phenotypeSet_workspace has a value which is a workspace_id
formulation has a value which is an FBAFormulation
notes has a value which is a string
phenotypeSimulationSet has a value which is a phenotypeSimulationSet_id
workspace has a value which is a workspace_id
overwrite has a value which is a bool
auth has a value which is a string
all_transporters has a value which is a bool
positive_transporters has a value which is a bool</pre>

</dd>
</dl>

<h2 id="add_media_transporters_params">add_media_transporters_params</h2>

<dl>

<dt id="Description238">Description</dt>
<dd>

<p>Input parameters for the add_media_transporters function.</p>

<pre><code>            phenotype_set_id phenotypeSet - ID for a phenotype set (required)
            workspace_id phenotypeSet_workspace - ID for the workspace in which the phenotype set is found
                fbamodel_id model - Model to which to add the transport reactions (required)
                workspace_id model_workspace - workspace containing the input model
                fbamodel_id outmodel - Name of output model (with transporters added)
                workspace_id workspace - workspace where the modified model should be saved
                bool overwrite - Overwrite or not
                string auth - Auth string
                bool all_transporters - Add transporters for ALL media in the phenotypeset
                bool positive_transporters - Add transporters for only POSITIVE (non-zero growth) media in the phenotype set</code></pre>

</dd>
<dt id="Definition159">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
phenotypeSet has a value which is a phenotype_set_id
phenotypeSet_workspace has a value which is a workspace_id
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
outmodel has a value which is a fbamodel_id
workspace has a value which is a workspace_id
overwrite has a value which is a bool
auth has a value which is a string
all_transporters has a value which is a bool
positive_transporters has a value which is a bool</pre>

</dd>
</dl>

<h2 id="export_phenotypeSimulationSet_params">export_phenotypeSimulationSet_params</h2>

<dl>

<dt id="Description239">Description</dt>
<dd>

<p>Input parameters for the &quot;export_phenotypeSimulationSet&quot; function.</p>

<pre><code>        phenotypeSimulationSet_id phenotypeSimulationSet - ID of the phenotype simulation set to be exported (a required argument)
        workspace_id workspace - workspace where the phenotype simulation set is stored (a required argument)
        string format - format to which phenotype simulation set should be exported (html, json)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition160">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
phenotypeSimulationSet has a value which is a phenotypeSimulationSet_id
workspace has a value which is a workspace_id
format has a value which is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="integrate_reconciliation_solutions_params">integrate_reconciliation_solutions_params</h2>

<dl>

<dt id="Description240">Description</dt>
<dd>

<p>Input parameters for the &quot;integrate_reconciliation_solutions&quot; function.</p>

<pre><code>        fbamodel_id model - ID of model for which reconciliation solutions should be integrated (a required argument)
        workspace_id model_workspace - workspace containing model for which solutions should be integrated (an optional argument: default is value of workspace argument)
        list&lt;gapfillsolution_id&gt; gapfillSolutions - list of gapfill solutions to be integrated (a required argument)
        list&lt;gapgensolution_id&gt; gapgenSolutions - list of gapgen solutions to be integrated (a required argument)
        fbamodel_id out_model - ID to which modified model should be saved (an optional argument: default is value of workspace argument)
        workspace_id workspace - workspace where modified model should be saved (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition161">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
gapfillSolutions has a value which is a reference to a list where each element is a gapfillsolution_id
gapgenSolutions has a value which is a reference to a list where each element is a gapgensolution_id
out_model has a value which is a fbamodel_id
workspace has a value which is a workspace_id
auth has a value which is a string
overwrite has a value which is a bool</pre>

</dd>
</dl>

<h2 id="queue_runfba_params">queue_runfba_params</h2>

<dl>

<dt id="Description241">Description</dt>
<dd>

<p>******************************************************************************** Code relating to queuing long running jobs ********************************************************************************</p>

</dd>
<dt id="Definition162">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
formulation has a value which is an FBAFormulation
fva has a value which is a bool
simulateko has a value which is a bool
minimizeflux has a value which is a bool
findminmedia has a value which is a bool
notes has a value which is a string
fba has a value which is a fba_id
workspace has a value which is a workspace_id
auth has a value which is a string
overwrite has a value which is a bool
add_to_model has a value which is a bool</pre>

</dd>
</dl>

<h2 id="gapfill_model_params">gapfill_model_params</h2>

<dl>

<dt id="Description242">Description</dt>
<dd>

<p>Input parameters for the &quot;queue_gapfill_model&quot; function.</p>

<pre><code>        fbamodel_id model - ID of the model that gapfill should be run on (a required argument)
        workspace_id model_workspace - workspace where model for gapfill should be run (an optional argument; default is the value of the workspace argument)
        GapfillingFormulation formulation - a hash specifying the parameters for the gapfill study (an optional argument)
        phenotype_set_id phenotypeSet - ID of a phenotype set against which gapfilled model should be simulated (an optional argument: default is &#39;undef&#39;)
        workspace_id phenotypeSet_workspace - workspace containing phenotype set to be simulated (an optional argument; default is the value of the workspace argument)
        bool integrate_solution - a flag indicating if the first solution should be integrated in the model (an optional argument: default is &#39;0&#39;)
        list&lt;string&gt; target_reactions - a list of reactions to activate with gapfilling
        fbamodel_id out_model - ID where the gapfilled model will be saved (an optional argument: default is &#39;undef&#39;)
        gapfill_id gapFill - ID to which gapfill solution will be saved (an optional argument: default is &#39;undef&#39;)
        workspace_id workspace - workspace where gapfill results will be saved (a required argument)
        int timePerSolution - maximum time to spend to obtain each solution
        int totalTimeLimit - maximum time to spend to obtain all solutions
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)
        bool completeGapfill - boolean indicating that all inactive reactions should be gapfilled</code></pre>

</dd>
<dt id="Definition163">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
formulation has a value which is a GapfillingFormulation
phenotypeSet has a value which is a phenotype_set_id
phenotypeSet_workspace has a value which is a workspace_id
integrate_solution has a value which is a bool
target_reactions has a value which is a reference to a list where each element is a string
out_model has a value which is a fbamodel_id
workspace has a value which is a workspace_id
gapFill has a value which is a gapfill_id
timePerSolution has a value which is an int
totalTimeLimit has a value which is an int
auth has a value which is a string
overwrite has a value which is a bool
completeGapfill has a value which is a bool</pre>

</dd>
</dl>

<h2 id="gapgen_model_params">gapgen_model_params</h2>

<dl>

<dt id="Description243">Description</dt>
<dd>

<p>Input parameters for the &quot;queue_gapgen_model&quot; function.</p>

<pre><code>        fbamodel_id model - ID of the model that gapgen should be run on (a required argument)
        workspace_id model_workspace - workspace where model for gapgen should be run (an optional argument; default is the value of the workspace argument)
        GapgenFormulation formulation - a hash specifying the parameters for the gapgen study (an optional argument)
        phenotype_set_id phenotypeSet - ID of a phenotype set against which gapgened model should be simulated (an optional argument: default is &#39;undef&#39;)
        workspace_id phenotypeSet_workspace - workspace containing phenotype set to be simulated (an optional argument; default is the value of the workspace argument)
        bool integrate_solution - a flag indicating if the first solution should be integrated in the model (an optional argument: default is &#39;0&#39;)
        fbamodel_id out_model - ID where the gapgened model will be saved (an optional argument: default is &#39;undef&#39;)
        gapgen_id gapGen - ID to which gapgen solution will be saved (an optional argument: default is &#39;undef&#39;)
        workspace_id workspace - workspace where gapgen results will be saved (a required argument)
        int timePerSolution - maximum time to spend to obtain each solution
        int totalTimeLimit - maximum time to spend to obtain all solutions
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition164">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
formulation has a value which is a GapgenFormulation
phenotypeSet has a value which is a phenotype_set_id
phenotypeSet_workspace has a value which is a workspace_id
integrate_solution has a value which is a bool
out_model has a value which is a fbamodel_id
workspace has a value which is a workspace_id
gapGen has a value which is a gapgen_id
auth has a value which is a string
timePerSolution has a value which is an int
totalTimeLimit has a value which is an int
overwrite has a value which is a bool</pre>

</dd>
</dl>

<h2 id="wildtype_phenotype_reconciliation_params">wildtype_phenotype_reconciliation_params</h2>

<dl>

<dt id="Description244">Description</dt>
<dd>

<p>Input parameters for the &quot;queue_wildtype_phenotype_reconciliation&quot; function.</p>

<pre><code>        fbamodel_id model - ID of the model that reconciliation should be run on (a required argument)
        workspace_id model_workspace - workspace where model for reconciliation should be run (an optional argument; default is the value of the workspace argument)
        FBAFormulation formulation - a hash specifying the parameters for the reconciliation study (an optional argument)
        GapfillingFormulation gapfill_formulation - a hash specifying the parameters for the gapfill study (an optional argument)
        GapgenFormulation gapgen_formulation - a hash specifying the parameters for the gapgen study (an optional argument)
        phenotype_set_id phenotypeSet - ID of a phenotype set against which reconciled model should be simulated (an optional argument: default is &#39;undef&#39;)
        workspace_id phenotypeSet_workspace - workspace containing phenotype set to be simulated (an optional argument; default is the value of the workspace argument)
        fbamodel_id out_model - ID where the reconciled model will be saved (an optional argument: default is &#39;undef&#39;)
        list&lt;gapgen_id&gt; gapGens - IDs of gapgen solutions (an optional argument: default is &#39;undef&#39;)
        list&lt;gapfill_id&gt; gapFills - IDs of gapfill solutions (an optional argument: default is &#39;undef&#39;)
        bool queueSensitivityAnalysis - flag indicating if sensitivity analysis should be queued to run on solutions (an optional argument: default is &#39;0&#39;)
        bool queueReconciliationCombination - flag indicating if reconcilication combination should be queued to run on solutions (an optional argument: default is &#39;0&#39;)
        workspace_id workspace - workspace where reconciliation results will be saved (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition165">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
fba_formulation has a value which is an FBAFormulation
gapfill_formulation has a value which is a GapfillingFormulation
gapgen_formulation has a value which is a GapgenFormulation
phenotypeSet has a value which is a phenotype_set_id
phenotypeSet_workspace has a value which is a workspace_id
out_model has a value which is a fbamodel_id
workspace has a value which is a workspace_id
gapFills has a value which is a reference to a list where each element is a gapfill_id
gapGens has a value which is a reference to a list where each element is a gapgen_id
queueSensitivityAnalysis has a value which is a bool
queueReconciliationCombination has a value which is a bool
auth has a value which is a string
overwrite has a value which is a bool</pre>

</dd>
</dl>

<h2 id="queue_reconciliation_sensitivity_analysis_params">queue_reconciliation_sensitivity_analysis_params</h2>

<dl>

<dt id="Description245">Description</dt>
<dd>

<p>Input parameters for the &quot;queue_reconciliation_sensitivity_analysis&quot; function.</p>

<pre><code>        fbamodel_id model - ID of the model that sensitivity analysis should be run on (a required argument)
        workspace_id model_workspace - workspace where model for sensitivity analysis should be run (an optional argument; default is the value of the workspace argument)
        FBAFormulation formulation - a hash specifying the parameters for the sensitivity analysis study (an optional argument)
        GapfillingFormulation gapfill_formulation - a hash specifying the parameters for the gapfill study (an optional argument)
        GapgenFormulation gapgen_formulation - a hash specifying the parameters for the gapgen study (an optional argument)
        phenotype_set_id phenotypeSet - ID of a phenotype set against which sensitivity analysis model should be simulated (an optional argument: default is &#39;undef&#39;)
        workspace_id phenotypeSet_workspace - workspace containing phenotype set to be simulated (an optional argument; default is the value of the workspace argument)
        fbamodel_id out_model - ID where the sensitivity analysis model will be saved (an optional argument: default is &#39;undef&#39;)
        list&lt;gapgen_id&gt; gapGens - IDs of gapgen solutions (an optional argument: default is &#39;undef&#39;)
        list&lt;gapfill_id&gt; gapFills - IDs of gapfill solutions (an optional argument: default is &#39;undef&#39;)
        bool queueReconciliationCombination - flag indicating if sensitivity analysis combination should be queued to run on solutions (an optional argument: default is &#39;0&#39;)
        workspace_id workspace - workspace where sensitivity analysis results will be saved (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition166">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
workspace has a value which is a workspace_id
phenotypeSet has a value which is a phenotype_set_id
fba_formulation has a value which is an FBAFormulation
model_workspace has a value which is a workspace_id
phenotypeSet_workspace has a value which is a workspace_id
gapFills has a value which is a reference to a list where each element is a gapfill_id
gapGens has a value which is a reference to a list where each element is a gapgen_id
queueReconciliationCombination has a value which is a bool
auth has a value which is a string
overwrite has a value which is a bool</pre>

</dd>
</dl>

<h2 id="combine_wildtype_phenotype_reconciliation_params">combine_wildtype_phenotype_reconciliation_params</h2>

<dl>

<dt id="Description246">Description</dt>
<dd>

<p>Input parameters for the &quot;queue_combine_wildtype_phenotype_reconciliation&quot; function.</p>

<pre><code>        fbamodel_id model - ID of the model that solution combination should be run on (a required argument)
        workspace_id model_workspace - workspace where model for solution combination should be run (an optional argument; default is the value of the workspace argument)
        FBAFormulation formulation - a hash specifying the parameters for the solution combination study (an optional argument)
        GapfillingFormulation gapfill_formulation - a hash specifying the parameters for the gapfill study (an optional argument)
        GapgenFormulation gapgen_formulation - a hash specifying the parameters for the gapgen study (an optional argument)
        phenotype_set_id phenotypeSet - ID of a phenotype set against which solution combination model should be simulated (an optional argument: default is &#39;undef&#39;)
        workspace_id phenotypeSet_workspace - workspace containing phenotype set to be simulated (an optional argument; default is the value of the workspace argument)
        fbamodel_id out_model - ID where the solution combination model will be saved (an optional argument: default is &#39;undef&#39;)
        list&lt;gapgen_id&gt; gapGens - IDs of gapgen solutions (an optional argument: default is &#39;undef&#39;)
        list&lt;gapfill_id&gt; gapFills - IDs of gapfill solutions (an optional argument: default is &#39;undef&#39;)
        workspace_id workspace - workspace where solution combination results will be saved (a required argument)
        int timePerSolution - maximum time spent per solution
        int totalTimeLimit - maximum time allowed to work on problem
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition167">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
model_workspace has a value which is a workspace_id
fba_formulation has a value which is an FBAFormulation
gapfill_formulation has a value which is a GapfillingFormulation
gapgen_formulation has a value which is a GapgenFormulation
phenotypeSet has a value which is a phenotype_set_id
phenotypeSet_workspace has a value which is a workspace_id
out_model has a value which is a fbamodel_id
workspace has a value which is a workspace_id
gapFills has a value which is a reference to a list where each element is a gapfill_id
gapGens has a value which is a reference to a list where each element is a gapgen_id
auth has a value which is a string
overwrite has a value which is a bool</pre>

</dd>
</dl>

<h2 id="run_job_params">run_job_params</h2>

<dl>

<dt id="Description247">Description</dt>
<dd>

<p>Input parameters for the &quot;run_job&quot; function.</p>

<pre><code>        job_id job - ID of the job object (a required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition168">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
job has a value which is a job_id
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="queue_job_params">queue_job_params</h2>

<dl>

<dt id="Description248">Description</dt>
<dd>

<p>Input parameters for the &quot;queue_job&quot; function.</p>

<pre><code>        string method;
        mapping&lt;string,string&gt; parameters;</code></pre>

</dd>
<dt id="Definition169">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
method has a value which is a string
parameters has a value which is a reference to a hash where the key is a string and the value is a string</pre>

</dd>
</dl>

<h2 id="set_cofactors_params">set_cofactors_params</h2>

<dl>

<dt id="Description249">Description</dt>
<dd>

<p>Input parameters for the &quot;set_cofactors&quot; function.</p>

<pre><code>        list&lt;compound_id&gt; cofactors - list of compounds that are universal cofactors (required)
        biochemistry_id biochemistry - ID of biochemistry database (optional, default is &quot;default&quot;) 
        workspace_id biochemistry_workspace - ID of workspace containing biochemistry database (optional, default is current workspace)
        bool reset - true to reset (turn off) compounds as universal cofactors (optional, default is false)
        bool overwrite - true to overwrite existing object (optional, default is false)
        string auth - the authentication token of the KBase account (optional, default user is &quot;public&quot;)</code></pre>

</dd>
<dt id="Definition170">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
cofactors has a value which is a reference to a list where each element is a compound_id
biochemistry has a value which is a biochemistry_id
biochemistry_workspace has a value which is a workspace_id
reset has a value which is a bool
overwrite has a value which is a bool
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="find_reaction_synonyms_params">find_reaction_synonyms_params</h2>

<dl>

<dt id="Description250">Description</dt>
<dd>

<p>Input parameters for the &quot;find_reaction_synonyms&quot; function.</p>

<pre><code>        reaction_synonyms - ID of reaction synonyms object (required argument)
        workspace_id workspace - ID of workspace for storing objects (optional argument, default is current workspace)
        biochemistry_id biochemistry - ID of the biochemistry database (optional argument, default is default)
        workspace_id biochemistry_workspace - ID of workspace containing biochemistry database (optional argument, default is kbase)
        overwrite - True to overwrite existing object (optional argument, default is false)
        string auth - the authentication token of the KBase account (optional argument, default user is &quot;public&quot;)</code></pre>

</dd>
<dt id="Definition171">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
reaction_synonyms has a value which is a reaction_synonyms_id
workspace has a value which is a workspace_id
biochemistry has a value which is a biochemistry_id
biochemistry_workspace has a value which is a workspace_id
overwrite has a value which is a bool
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="role_to_reactions_params">role_to_reactions_params</h2>

<dl>

<dt id="Description251">Description</dt>
<dd>

<p>Input parameters for the &quot;role_to_reactions&quot; function.</p>

<pre><code>        template_id templateModel - ID of the template model to be used to determine mapping (default is &#39;&#39;)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition172">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
templateModel has a value which is a template_id
workspace has a value which is a workspace_id
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="kb_sub_id">kb_sub_id</h2>

<dl>

<dt id="Description252">Description</dt>
<dd>

<p>******************************************************************************** Code relating to assessing model sensitivity to reaction knockouts ********************************************************************************</p>

</dd>
<dt id="Definition173">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="kb_id">kb_id</h2>

<dl>

<dt id="Description253">Description</dt>
<dd>

<p>@id kb</p>

</dd>
<dt id="Definition174">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="ws_id">ws_id</h2>

<dl>

<dt id="Description254">Description</dt>
<dd>

<p>@id ws</p>

</dd>
<dt id="Definition175">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="ws_sub_id">ws_sub_id</h2>

<dl>

<dt id="Description255">Description</dt>
<dd>

<p>@id wssub</p>

</dd>
<dt id="Definition176">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="ReactionSensitivityAnalysisCorrectedReaction">ReactionSensitivityAnalysisCorrectedReaction</h2>

<dl>

<dt id="Description256">Description</dt>
<dd>

<p>ReactionSensitivityAnalysisCorrectedReaction object</p>

<p>kb_sub_id kbid - KBase ID for reaction knockout corrected reaction ws_sub_id model_reaction_wsid - ID of model reaction float normalized_required_reaction_count - Normalized count of reactions required for this reaction to function list&lt;ws_sub_id&gt; required_reactions - list of reactions required for this reaction to function</p>

<p>@searchable ws_subset kbid model_reaction_kbid required_reactions @optional</p>

</dd>
<dt id="Definition177">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
kbid has a value which is a kb_sub_id
model_reaction_wsid has a value which is a ws_sub_id
normalized_required_reaction_count has a value which is a float
required_reactions has a value which is a reference to a list where each element is a ws_sub_id</pre>

</dd>
</dl>

<h2 id="ReactionSensitivityAnalysisReaction">ReactionSensitivityAnalysisReaction</h2>

<dl>

<dt id="Description257">Description</dt>
<dd>

<p>Object for holding reaction knockout sensitivity reaction data</p>

<p>kb_sub_id kbid - KBase ID for reaction knockout sensitivity reaction ws_sub_id model_reaction_wsid - ID of model reaction bool delete - indicates if reaction is to be deleted bool deleted - indicates if the reaction has been deleted string direction - Direction of reaction that was tested (&gt; is forward, &lt; backward and = both) float growth_fraction - Fraction of wild-type growth after knockout float normalized_activated_reaction_count - Normalized number of activated reactions list&lt;ws_sub_id&gt; biomass_compounds - List of biomass compounds that depend on the reaction list&lt;ws_sub_id&gt; new_inactive_rxns - List of new reactions dependant upon reaction KO list&lt;ws_sub_id&gt; new_essentials - List of new essential genes with reaction knockout</p>

</dd>
<dt id="Definition178">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
kbid has a value which is a kb_sub_id
model_reaction_wsid has a value which is a ws_sub_id
growth_fraction has a value which is a float
delete has a value which is a bool
deleted has a value which is a bool
direction has a value which is a string
normalized_activated_reaction_count has a value which is a float
biomass_compounds has a value which is a reference to a list where each element is a ws_sub_id
new_inactive_rxns has a value which is a reference to a list where each element is a ws_sub_id
new_essentials has a value which is a reference to a list where each element is a ws_sub_id</pre>

</dd>
</dl>

<h2 id="ReactionSensitivityAnalysis">ReactionSensitivityAnalysis</h2>

<dl>

<dt id="Description258">Description</dt>
<dd>

<p>Object for holding reaction knockout sensitivity results</p>

<pre><code>        kb_id kbid - KBase ID of reaction sensitivity object
        ws_id model_wsid - Workspace reference to associated model
        string type - type of reaction KO sensitivity object
        bool deleted_noncontributing_reactions - boolean indicating if noncontributing reactions were deleted
        bool integrated_deletions_in_model - boolean indicating if deleted reactions were implemented in the model
        list&lt;ReactionSensitivityAnalysisReaction&gt; reactions - list of sensitivity data for tested reactions
        list&lt;ReactionSensitivityAnalysisCorrectedReaction&gt; corrected_reactions - list of reactions dependant upon tested reactions</code></pre>

</dd>
<dt id="Definition179">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
kbid has a value which is a kb_id
model_wsid has a value which is a ws_id
type has a value which is a string
deleted_noncontributing_reactions has a value which is a bool
integrated_deletions_in_model has a value which is a bool
reactions has a value which is a reference to a list where each element is a ReactionSensitivityAnalysisReaction
corrected_reactions has a value which is a reference to a list where each element is a ReactionSensitivityAnalysisCorrectedReaction</pre>

</dd>
</dl>

<h2 id="rxnprob_id">rxnprob_id</h2>

<dl>

<dt id="Description259">Description</dt>
<dd>

<p>ID for a RxnProbs T.O. (defined in the probabilistic annotation spec)</p>

</dd>
<dt id="Definition180">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="reaction_sensitivity_analysis_params">reaction_sensitivity_analysis_params</h2>

<dl>

<dt id="Description260">Description</dt>
<dd>

<p>Input parameters for the &quot;reaction_sensitivity_analysis&quot; function.</p>

<pre><code>        fbamodel_id model - ID of model to be analyzed (a required argument)
        workspace_id model_ws - ID of workspace with model to be analyzed (an optional argument - default is value of workspace argument)
        string rxnsens_uid - Name of RxnSensitivity object in workspace (an optional argument - default is KBase ID)
        workspace_id workspace - ID of workspace where output and default inputs will be selected from (a required argument)
        list&lt;reaction_id&gt; reactions_to_delete - list of reactions to delete in sensitiviity analysis; note, order of the reactions matters (a required argument unless gapfill solution ID is provided)                
        gapfillsolution_id gapfill_solution_id - A Gapfill solution ID. If provided, all reactions in the provided solution will be tested for deletion.
        bool delete_noncontributing_reactions - a boolean indicating if unuseful reactions should be deleted when running the analysis (an optional argument - default is &quot;0&quot;)
        rxnprob_id rxnprobs_id - ID for a RxnProbs object in a workspace. If provided less likely reactions will be tested for deletion first in the sensitivity analysis (optional).
        workspace_id rxnprobs_ws - Workspace in which the RxnProbs object is located (optional - default is the value of the workspace argument).
        string type - type of Reaction sensitivity analysis (an optional argument - default is &quot;unknown&quot;)
        string auth  - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument)</code></pre>

</dd>
<dt id="Definition181">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
model_ws has a value which is a workspace_id
rxnsens_uid has a value which is a string
workspace has a value which is a workspace_id
reactions_to_delete has a value which is a reference to a list where each element is a reaction_id
gapfill_solution_id has a value which is a gapfillsolution_id
delete_noncontributing_reactions has a value which is a bool
rxnprobs_id has a value which is a rxnprob_id
rxnprobs_ws has a value which is a workspace_id
type has a value which is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="filter_iterative_solutions_params">filter_iterative_solutions_params</h2>

<dl>

<dt id="Description261">Description</dt>
<dd>

<p>Input parameters for the &quot;filter_iterative_solutions&quot; function. fbamodel_id model - Model ID for which to filter iterative gapfill solutions (a required argument) fbamodel_id outmodel - ModelID to which to save the filtered results (by default the filtered model is given the same ID as the input model) float cutoff - Cutoff for cost per reaction above which to remove iterative gapfill solution reactions (a required argument) gapfillsolution_id gapfillsln - Gapfill_solution ID (UUID.solution.#) containing the iterative gapfill solutions to filter (a required argument) string auth - The authorization token of the KBase account with workspace permissions. workspace_id workspace - ID of workspace where output and default inputs will be selected from (a required argument) workspace_id input_model_ws - ID of workspace containing the input model</p>

</dd>
<dt id="Definition182">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
outmodel has a value which is a fbamodel_id
cutoff has a value which is a float
gapfillsln has a value which is a gapfillsolution_id
workspace has a value which is a workspace_id
input_model_ws has a value which is a workspace_id
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="delete_noncontributing_reactions_params">delete_noncontributing_reactions_params</h2>

<dl>

<dt id="Description262">Description</dt>
<dd>

<p>Input parameters for the &quot;delete_noncontributing_reactions&quot; function. workspace_id workspae - Workspace for outputs and default inputs (a required argument) workspace_id rxn_sensitivity_ws - Workspace for reaction sensitivity object used as input string rxn_sensitivity - Reaction sensitivity ID fbamodel_id new_model_uid - ID for output model with noncontributing reactions deleted string new_rxn_sensitivity_uid - ID for rxnsensitivity object with bits set to indicate reactions were deleted string auth - Authorization token for user (must have appropriate permissions to read and write objects)</p>

</dd>
<dt id="Definition183">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
rxn_sensitivity_ws has a value which is a workspace_id
rxn_sensitivity has a value which is a string
workspace has a value which is a workspace_id
new_model_uid has a value which is a fbamodel_id
new_rxn_sensitivity_uid has a value which is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="AnnotationParameters">AnnotationParameters</h2>

<dl>

<dt id="Description263">Description</dt>
<dd>

<p>******************************************************************************** Code relating to workspace versions of genome analysis algorithms ********************************************************************************</p>

</dd>
<dt id="Definition184">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
call_genes has a value which is a bool
annotate_genes has a value which is a bool</pre>

</dd>
</dl>

<h2 id="annotate_workspace_Genome_params">annotate_workspace_Genome_params</h2>

<dl>

<dt id="Description264">Description</dt>
<dd>

<p>Input parameters for the &quot;annotate_workspace_Genome&quot; function.</p>

<p>string Genome_uid - user ID to be assigned to the Genome (required argument) string Genome_ws - workspace with genome for annotation (optional; workspace argument will be used if no genome workspace is provided) string new_uid - new ID to assign to annotated genome (optional; original genome will be overwritten if no new uid is provided) workspace_id workspace - ID of workspace with Genome (required argument) AnnotationParameters parameters - parameters for running annotation job string auth - the authentication token of the KBase account changing workspace permissions</p>

</dd>
<dt id="Definition185">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
Genome_uid has a value which is a string
Genome_ws has a value which is a string
new_uid has a value which is a string
workspace has a value which is a workspace_id
annotation_parameters has a value which is an AnnotationParameters
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="gtf_to_genome_params">gtf_to_genome_params</h2>

<dl>

<dt id="Description265">Description</dt>
<dd>

<p>******************************************************************************** Code relating to import and analysis of ProteinSets ********************************************************************************</p>

</dd>
<dt id="Definition186">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
contigset has a value which is a string
contigset_ws has a value which is a workspace_id
workspace has a value which is a workspace_id
genome_uid has a value which is a string
source_id has a value which is a string
source has a value which is a string
scientific_name has a value which is a string
domain has a value which is a string
genetic_code has a value which is an int
taxonomy has a value which is a string
gtf_file has a value which is a string</pre>

</dd>
</dl>

<h2 id="fasta_to_ProteinSet_params">fasta_to_ProteinSet_params</h2>

<dl>

<dt id="Description266">Description</dt>
<dd>

<p>Input parameters for the &quot;fasta_to_ProteinSet&quot; function.</p>

<pre><code>        string uid - user assigned ID for the protein set (optional)
        string fasta - string with sequence data from fasta file (required argument)
        workspace_id workspace - ID of workspace for storing objects (required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)
        string name - name of the protein data (optional)
        string sourceid - source ID of the protein data (optional)
        string source - source of the protein data (optional)
        string type - type of the protein set (optional)</code></pre>

</dd>
<dt id="Definition187">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
uid has a value which is a string
fasta has a value which is a string
workspace has a value which is a workspace_id
auth has a value which is a string
name has a value which is a string
sourceid has a value which is a string
source has a value which is a string
type has a value which is a string</pre>

</dd>
</dl>

<h2 id="ProteinSet_to_Genome_params">ProteinSet_to_Genome_params</h2>

<dl>

<dt id="Description267">Description</dt>
<dd>

<p>Input parameters for the &quot;ProteinSet_to_Genome&quot; function.</p>

<pre><code>        string ProteinSet_uid - ID to be assigned to the ProteinSet (required argument)
        workspace_id ProteinSet_ws - ID of workspace with the ProteinSet (optional argument; default is value of workspace argument)
        string uid - user assigned ID for the Genome (optional)
        workspace_id workspace - ID of workspace for storing objects (required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)
        string scientific_name - scientific name to assign to genome
        string domain - domain of life for genome
        int genetic_code - genetic code to assign to genome</code></pre>

</dd>
<dt id="Definition188">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
ProteinSet_uid has a value which is a string
ProteinSet_ws has a value which is a workspace_id
workspace has a value which is a workspace_id
uid has a value which is a string
auth has a value which is a string
scientific_name has a value which is a string
domain has a value which is a string
annotation_parameters has a value which is an AnnotationParameters</pre>

</dd>
</dl>

<h2 id="fasta_to_ContigSet_params">fasta_to_ContigSet_params</h2>

<dl>

<dt id="Description268">Description</dt>
<dd>

<p>******************************************************************************** Code relating to import and analysis of Contigs ********************************************************************************</p>

</dd>
<dt id="Definition189">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
uid has a value which is a string
fasta has a value which is a string
workspace has a value which is a workspace_id
auth has a value which is a string
name has a value which is a string
sourceid has a value which is a string
source has a value which is a string
type has a value which is a string</pre>

</dd>
</dl>

<h2 id="ContigSet_to_Genome_params">ContigSet_to_Genome_params</h2>

<dl>

<dt id="Description269">Description</dt>
<dd>

<p>Input parameters for the &quot;ContigSet_to_Genome&quot; function.</p>

<pre><code>        string ContigSet_uid - ID to be assigned to the ContigSet (required argument)
        workspace_id ContigSet_ws - ID of workspace with the ContigSet (optional argument; default is value of workspace argument)
        string uid - user assigned ID for the Genome (optional)
        workspace_id workspace - ID of workspace for storing objects (required argument)
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)
        string scientific_name - scientific name to assign to genome
        string domain - domain of life for genome
        int genetic_code - genetic code to assign to genome
        AnnotationParameters annotation_parameters - parameters for annotation of the genome</code></pre>

</dd>
<dt id="Definition190">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
ContigSet_uid has a value which is a string
ContigSet_ws has a value which is a workspace_id
workspace has a value which is a workspace_id
uid has a value which is a string
auth has a value which is a string
scientific_name has a value which is a string
domain has a value which is a string
genetic_code has a value which is an int
annotation_parameters has a value which is an AnnotationParameters</pre>

</dd>
</dl>

<h2 id="probanno_to_genome_params">probanno_to_genome_params</h2>

<dl>

<dt id="Description270">Description</dt>
<dd>

<p>******************************************************************************** Code relating to analysis of probabilistic annotations ********************************************************************************</p>

</dd>
<dt id="Definition191">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
pa_id has a value which is a probanno_id
pa_ws has a value which is a workspace_id
workspace has a value which is a workspace_id
g_id has a value which is a genome_id
threshold has a value which is a float
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="FunctionalRole">FunctionalRole</h2>

<dl>

<dt id="Description271">Description</dt>
<dd>

<p>******************************************************************************** Code relating to loading, retrieval, and curation of mappings ********************************************************************************</p>

</dd>
<dt id="Definition192">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a role_id
name has a value which is a string
feature has a value which is a string
aliases has a value which is a reference to a list where each element is a string
complexes has a value which is a reference to a list where each element is a complex_id</pre>

</dd>
</dl>

<h2 id="ComplexRole">ComplexRole</h2>

<dl>

<dt id="Definition193">Definition</dt>
<dd>



<pre>
a reference to a list containing 4 items:
0: (id) a role_id
1: (roleType) a string
2: (optional_role) a bool
3: (triggering) a bool</pre>

</dd>
</dl>

<h2 id="Complex">Complex</h2>

<dl>

<dt id="Definition194">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a complex_id
name has a value which is a string
aliases has a value which is a reference to a list where each element is a string
roles has a value which is a reference to a list where each element is a ComplexRole</pre>

</dd>
</dl>

<h2 id="subsystem_id">subsystem_id</h2>

<dl>

<dt id="Definition195">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="Subsystem">Subsystem</h2>

<dl>

<dt id="Definition196">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a subsystem_id
name has a value which is a string
phenoclass has a value which is a string
subclass has a value which is a string
type has a value which is a string
aliases has a value which is a reference to a list where each element is a string
roles has a value which is a reference to a list where each element is a role_id</pre>

</dd>
</dl>

<h2 id="Mapping">Mapping</h2>

<dl>

<dt id="Definition197">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a mapping_id
name has a value which is a string
subsystems has a value which is a reference to a list where each element is a Subsystem
roles has a value which is a reference to a list where each element is a FunctionalRole
complexes has a value which is a reference to a list where each element is a Complex</pre>

</dd>
</dl>

<h2 id="get_mapping_params">get_mapping_params</h2>

<dl>

<dt id="Definition198">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
map has a value which is a mapping_id
workspace has a value which is a workspace_id
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="subsysclass">subsysclass</h2>

<dl>

<dt id="Definition199">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: a string
1: a string</pre>

</dd>
</dl>

<h2 id="subsysclasses">subsysclasses</h2>

<dl>

<dt id="Definition200">Definition</dt>
<dd>



<pre>
a reference to a hash where the key is a string and the value is a subsysclass
</pre>

</dd>
</dl>

<h2 id="subsystem_of_roles_params">subsystem_of_roles_params</h2>

<dl>

<dt id="Definition201">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
roles has a value which is a reference to a list where each element is a string
map has a value which is a string
map_workspace has a value which is a string</pre>

</dd>
</dl>

<h2 id="adjust_mapping_role_params">adjust_mapping_role_params</h2>

<dl>

<dt id="Description272">Description</dt>
<dd>

<p>Input parameters for the &quot;adjust_mapping_role&quot; function.</p>

<pre><code>        mapping_id map - ID of the mapping object to be edited
        workspace_id workspace - ID of workspace containing mapping to be edited
        string role - identifier for role to be edited
        bool new - boolean indicating that a new role is being added
        string name - new name for the role
        string feature - representative feature MD5
        list&lt;string&gt; aliasesToAdd - list of new aliases for the role
        list&lt;string&gt; aliasesToRemove - list of aliases to remove for role
        bool delete - boolean indicating that role should be deleted
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition202">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
map has a value which is a mapping_id
workspace has a value which is a workspace_id
role has a value which is a string
new has a value which is a bool
name has a value which is a string
feature has a value which is a string
aliasesToAdd has a value which is a reference to a list where each element is a string
aliasesToRemove has a value which is a reference to a list where each element is a string
delete has a value which is a bool
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="adjust_mapping_complex_params">adjust_mapping_complex_params</h2>

<dl>

<dt id="Description273">Description</dt>
<dd>

<p>Input parameters for the &quot;adjust_mapping_complex&quot; function.</p>

<pre><code>        mapping_id map - ID of the mapping object to be edited
        workspace_id workspace - ID of workspace containing mapping to be edited
        string complex - identifier for complex to be edited
        bool new - boolean indicating that a new complex is being added
        string name - new name for the role
        string feature - representative feature MD5
        list&lt;string&gt; rolesToAdd - roles to add to the complex
        list&lt;string&gt; rolesToRemove - roles to remove from the complex
        bool delete - boolean indicating that complex should be deleted
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition203">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
map has a value which is a mapping_id
workspace has a value which is a workspace_id
complex has a value which is a string
new has a value which is a bool
name has a value which is a string
rolesToAdd has a value which is a reference to a list where each element is a string
rolesToRemove has a value which is a reference to a list where each element is a string
delete has a value which is a bool
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="adjust_mapping_subsystem_params">adjust_mapping_subsystem_params</h2>

<dl>

<dt id="Description274">Description</dt>
<dd>

<p>Input parameters for the &quot;adjust_mapping_subsystem&quot; function.</p>

<pre><code>        mapping_id map - ID of the mapping object to be edited
        workspace_id workspace - ID of workspace containing mapping to be edited
        string subsystem - identifier for subsystem to be edited
        bool new - boolean indicating that a new subsystem is being added
        string name - new name for the subsystem
        string type - new type for the subsystem
        string primclass - new class for the subsystem
        string subclass - new subclass for the subsystem
        list&lt;string&gt; rolesToAdd - roles to add to the subsystem
        list&lt;string&gt; rolesToRemove - roles to remove from the subsystem
        bool delete - boolean indicating that subsystem should be deleted
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition204">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
map has a value which is a mapping_id
workspace has a value which is a workspace_id
subsystem has a value which is a string
new has a value which is a bool
name has a value which is a string
type has a value which is a string
primclass has a value which is a string
subclass has a value which is a string
rolesToAdd has a value which is a reference to a list where each element is a string
rolesToRemove has a value which is a reference to a list where each element is a string
delete has a value which is a bool
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="temprxn_id">temprxn_id</h2>

<dl>

<dt id="Description275">Description</dt>
<dd>

<p>******************************************************************************** Code relating to loading, retrieval, and curation of template models ********************************************************************************</p>

</dd>
<dt id="Definition205">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="TemplateReaction">TemplateReaction</h2>

<dl>

<dt id="Definition206">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a temprxn_id
compartment has a value which is a compartment_id
reaction has a value which is a reaction_id
complexes has a value which is a reference to a list where each element is a complex_id
direction has a value which is a string
type has a value which is a string</pre>

</dd>
</dl>

<h2 id="TemplateBiomassCompounds">TemplateBiomassCompounds</h2>

<dl>

<dt id="Definition207">Definition</dt>
<dd>



<pre>
a reference to a list containing 7 items:
0: (compound) a compound_id
1: (compartment) a compartment_id
2: (cpdclass) a string
3: (universal) a string
4: (coefficientType) a string
5: (coefficient) a string
6: (linkedCompounds) a reference to a list where each element is a reference to a list containing 2 items:
	0: (coeffficient) a string
	1: (compound) a compound_id</pre>

</dd>
</dl>

<h2 id="tempbiomass_id">tempbiomass_id</h2>

<dl>

<dt id="Definition208">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="TemplateBiomass">TemplateBiomass</h2>

<dl>

<dt id="Definition209">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a tempbiomass_id
name has a value which is a string
type has a value which is a string
other has a value which is a string
protein has a value which is a string
dna has a value which is a string
rna has a value which is a string
cofactor has a value which is a string
energy has a value which is a string
cellwall has a value which is a string
lipid has a value which is a string
compounds has a value which is a reference to a list where each element is a TemplateBiomassCompounds</pre>

</dd>
</dl>

<h2 id="TemplateModel">TemplateModel</h2>

<dl>

<dt id="Definition210">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
id has a value which is a template_id
name has a value which is a string
type has a value which is a string
domain has a value which is a string
map has a value which is a mapping_id
mappingws has a value which is a workspace_id
mapping_ref has a value which is a string
biochemistry_ref has a value which is a string
reactions has a value which is a reference to a list where each element is a TemplateReaction
biomasses has a value which is a reference to a list where each element is a TemplateBiomass</pre>

</dd>
</dl>

<h2 id="get_template_model_params">get_template_model_params</h2>

<dl>

<dt id="Definition211">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
templateModel has a value which is a template_id
workspace has a value which is a workspace_id
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="import_template_fbamodel_params">import_template_fbamodel_params</h2>

<dl>

<dt id="Description276">Description</dt>
<dd>

<p>Input parameters for the &quot;import_template_fbamodel&quot; function.</p>

<pre><code>        mapping_id map - ID of the mapping to associate the template model with (an optional argument; default is &#39;default&#39;)
        workspace_id mapping_workspace - ID of the workspace where the associated mapping is found (an optional argument; default is &#39;kbase&#39;)
        list&lt;tuple&lt;string id,string compartment,string direction,string type,list&lt;string complex&gt; complexes&gt;&gt; templateReactions - list of reactions to include in template model
        list&lt;tuple&lt;string name,string type,float dna,float rna,float protein,float lipid,float cellwall,float cofactor,float energy,float other,list&lt;tuple&lt;string id,string compartment,string cpdclass,string coefficientType,float coefficient,string conditions&gt;&gt; compounds&gt;&gt; templateBiomass - list of template biomass reactions for template model
        string name - name for template model
        string modelType - type of model constructed by template
        string domain - domain of template model
        template_id id - ID that should be used for the newly imported template model (an optional argument; default is &#39;undef&#39;)
        workspace_id workspace - ID of the workspace where the newly developed template model will be stored; also the default assumed workspace for input objects (a required argument)
        bool ignore_errors - ignores missing roles or reactions and imports template model anyway
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition212">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
map has a value which is a mapping_id
mapping_workspace has a value which is a workspace_id
templateReactions has a value which is a reference to a list where each element is a reference to a list containing 5 items:
0: (id) a string
1: (compartment) a string
2: (direction) a string
3: (type) a string
4: (complexes) a reference to a list where each element is a stringtemplateBiomass has a value which is a reference to a list where each element is a reference to a list containing 11 items:
0: (name) a string
1: (type) a string
2: (dna) a float
3: (rna) a float
4: (protein) a float
5: (lipid) a float
6: (cellwall) a float
7: (cofactor) a float
8: (energy) a float
9: (other) a float
10: (compounds) a reference to a list where each element is a reference to a list containing 6 items:
	0: (id) a string
	1: (compartment) a string
	2: (cpdclass) a string
	3: (coefficientType) a string
	4: (coefficient) a float
	5: (conditions) a stringname has a value which is a string
modelType has a value which is a string
domain has a value which is a string
id has a value which is a template_id
workspace has a value which is a workspace_id
ignore_errors has a value which is a bool
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="adjust_template_reaction_params">adjust_template_reaction_params</h2>

<dl>

<dt id="Definition213">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
templateModel has a value which is a template_id
workspace has a value which is a workspace_id
reaction has a value which is a string
clearComplexes has a value which is a bool
new has a value which is a bool
delete has a value which is a bool
compartment has a value which is a compartment_id
complexesToAdd has a value which is a reference to a list where each element is a complex_id
complexesToRemove has a value which is a reference to a list where each element is a complex_id
direction has a value which is a string
type has a value which is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="adjust_template_biomass_params">adjust_template_biomass_params</h2>

<dl>

<dt id="Definition214">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
templateModel has a value which is a template_id
workspace has a value which is a workspace_id
biomass has a value which is a string
new has a value which is a bool
delete has a value which is a bool
clearBiomassCompounds has a value which is a bool
name has a value which is a string
type has a value which is a string
other has a value which is a string
protein has a value which is a string
dna has a value which is a string
rna has a value which is a string
cofactor has a value which is a string
energy has a value which is a string
cellwall has a value which is a string
lipid has a value which is a string
compoundsToRemove has a value which is a reference to a list where each element is a reference to a list containing 2 items:
0: (compound) a compound_id
1: (compartment) a compartment_idcompoundsToAdd has a value which is a reference to a list where each element is a reference to a list containing 7 items:
0: (compound) a compound_id
1: (compartment) a compartment_id
2: (cpdclass) a string
3: (universal) a string
4: (coefficientType) a string
5: (coefficient) a string
6: (linkedCompounds) a reference to a list where each element is a reference to a list containing 2 items:
	0: (coeffficient) a string
	1: (compound) a compound_idauth has a value which is a string</pre>

</dd>
</dl>

<h2 id="add_stimuli_params">add_stimuli_params</h2>

<dl>

<dt id="Description277">Description</dt>
<dd>

<p>******************************************************************************** Code relating to reconstruction, import, and analysis of regulatory models ********************************************************************************</p>

</dd>
<dt id="Definition215">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
biochemid has a value which is a string
biochem_workspace has a value which is a string
stimuliid has a value which is a string
name has a value which is a string
abbreviation has a value which is a string
type has a value which is a string
description has a value which is a string
compounds has a value which is a reference to a list where each element is a string
workspace has a value which is a string
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="Stimuli">Stimuli</h2>

<dl>

<dt id="Definition216">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
kbid has a value which is a kbase_id
name has a value which is a string
abbreviation has a value which is a string
description has a value which is a string
type has a value which is a string
compound_kbids has a value which is a reference to a list where each element is a kbase_id</pre>

</dd>
</dl>

<h2 id="RegulatoryModelRegulonStimuli">RegulatoryModelRegulonStimuli</h2>

<dl>

<dt id="Definition217">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
kbid has a value which is a kbase_id
stimuli_kbid has a value which is a kbase_id
is_inhibitor has a value which is a bool
strength has a value which is a float
min_concentration has a value which is a float
max_concentration has a value which is a float
regulator_kbids has a value which is a reference to a list where each element is a kbase_id</pre>

</dd>
</dl>

<h2 id="RegulatoryModelRegulon">RegulatoryModelRegulon</h2>

<dl>

<dt id="Definition218">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
kbid has a value which is a kbase_id
name has a value which is a string
feature_kbids has a value which is a reference to a list where each element is a kbase_id
stimuli has a value which is a reference to a list where each element is a RegulatoryModelRegulonStimuli</pre>

</dd>
</dl>

<h2 id="RegulatoryModel">RegulatoryModel</h2>

<dl>

<dt id="Definition219">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
kbid has a value which is a kbase_id
name has a value which is a string
type has a value which is a string
genome_wsid has a value which is a ws_ref
regulons has a value which is a reference to a list where each element is a RegulatoryModelRegulon</pre>

</dd>
</dl>

<h2 id="import_regulatory_model_params">import_regulatory_model_params</h2>

<dl>

<dt id="Definition220">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
regmodel_uid has a value which is a string
workspace has a value which is a workspace_id
genome has a value which is a string
genome_ws has a value which is a workspace_id
name has a value which is a string
type has a value which is a string
regulons has a value which is a reference to a list where each element is a reference to a list containing 3 items:
0: (name) a string
1: (features) a reference to a list where each element is a string
2: (stimuli) a reference to a list where each element is a reference to a list containing 6 items:
	0: (stimuli) a string
	1: (in_inhibitor) a bool
	2: (strength) a float
	3: (min_conc) a float
	4: (max_conc) a float
	5: (regulators) a reference to a list where each element is a kbase_idauth has a value which is a string</pre>

</dd>
</dl>

<h2 id="compare_models_params">compare_models_params</h2>

<dl>

<dt id="Description278">Description</dt>
<dd>

<p>******************************************************************************** Functions relating to comparison of models ********************************************************************************</p>

</dd>
<dt id="Definition221">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
models has a value which is a reference to a list where each element is a fbamodel_id
workspaces has a value which is a reference to a list where each element is a workspace_id
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="ModelComparisonModel">ModelComparisonModel</h2>

<dl>

<dt id="Description279">Description</dt>
<dd>

<p>Data structure to hold model comparison data</p>

<pre><code>        fbamodel_id model - id of the fba model
        workspace_id workspace - id of workspace with model
        string model_name - name of the fba model
        genome_id genome - id of the genome for the fba model
        string genome_name - name of the genome for the fba model
        int core_reactions - number of core reactions in the fba model
        int unique_reactions - number of unique reactions in the fba model</code></pre>

</dd>
<dt id="Definition222">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a fbamodel_id
workspace has a value which is a workspace_id
model_name has a value which is a string
genome has a value which is a genome_id
genome_name has a value which is a string
gapfilled_reactions has a value which is an int
core_reactions has a value which is an int
noncore_reactions has a value which is an int</pre>

</dd>
</dl>

<h2 id="ModelCompareReaction">ModelCompareReaction</h2>

<dl>

<dt id="Description280">Description</dt>
<dd>

<p>Data structure to hold model reaction comparison data</p>

<pre><code>        reaction_id reaction - id of the reaction
        compartment_id compartment - id of the reaction compartment
        string equation - equation for the reaction
        bool core - boolean indicating if the reaction is core
        mapping&lt;fbamodel_id,list&lt;feature_id&gt; &gt; model_features - map of models and features for reaction
        string role - role associated with the reaction
        string subsytem - subsystem associated with role
        string primclass - class one of the subsystem
        string subclass - class two of the subsystem
        int number_models - number of models with reaction
        float fraction_models - fraction of models with reaction</code></pre>

</dd>
<dt id="Definition223">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
reaction has a value which is a reaction_id
compartment has a value which is a string
equation has a value which is a string
core has a value which is a bool
model_features has a value which is a reference to a hash where the key is a fbamodel_id and the value is a reference to a list where each element is a feature_id
role has a value which is a string
subsystem has a value which is a string
primclass has a value which is a string
subclass has a value which is a string
number_models has a value which is an int
fraction_models has a value which is a float</pre>

</dd>
</dl>

<h2 id="ModelComparisonData">ModelComparisonData</h2>

<dl>

<dt id="Description281">Description</dt>
<dd>

<p>Output structure for the &quot;compare_models&quot; function.</p>

<pre><code>        list&lt;ModelComparisonModel&gt; model_comparisons;
        list&lt;ModelCompareReaction&gt; reaction_comparisons;</code></pre>

</dd>
<dt id="Definition224">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model_comparisons has a value which is a reference to a list where each element is a ModelComparisonModel
reaction_comparisons has a value which is a reference to a list where each element is a ModelCompareReaction
auth has a value which is a string</pre>

</dd>
</dl>

<h2 id="compare_fbas_params">compare_fbas_params</h2>

<dl>

<dt id="Description282">Description</dt>
<dd>

<p>******************************************************************************** Functions relating to comparison of FBAs ********************************************************************************</p>

</dd>
<dt id="Definition225">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
output_id has a value which is a string
fbas has a value which is a reference to a list where each element is a reference to a list containing 2 items:
0: a workspace_id
1: a fba_idworkspace has a value which is a workspace_id</pre>

</dd>
</dl>

<h2 id="compare_genomes_params">compare_genomes_params</h2>

<dl>

<dt id="Description283">Description</dt>
<dd>

<p>******************************************************************************** Functions relating to comparison of genomes ********************************************************************************</p>

</dd>
<dt id="Definition226">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
pangenome_id has a value which is a string
pangenome_ws has a value which is a string
protcomp_id has a value which is a string
protcomp_ws has a value which is a string
output_id has a value which is a string
workspace has a value which is a string</pre>

</dd>
</dl>

<h2 id="MetagenomeAnnotationOTUFunction">MetagenomeAnnotationOTUFunction</h2>

<dl>

<dt id="Description284">Description</dt>
<dd>

<p>******************************************************************************** Functions relating to construction of community models ********************************************************************************</p>

</dd>
<dt id="Definition227">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
kbid has a value which is a string
reference_genes has a value which is a reference to a list where each element is a string
functional_role has a value which is a string
abundance has a value which is an int
confidence has a value which is a float</pre>

</dd>
</dl>

<h2 id="MetagenomeAnnotationOTU">MetagenomeAnnotationOTU</h2>

<dl>

<dt id="Description285">Description</dt>
<dd>

<p>Structure for the &quot;MetagenomeAnnotationOTU&quot; object</p>

<pre><code>        string name - name of metagenome OTU
        string kbid - KBase ID of OTU of metagenome object
        string source_id - ID used for OTU in metagenome source
        string source - source OTU ID
        list&lt;MetagenomeAnnotationOTUFunction&gt; functions - list of functions in OTU</code></pre>

</dd>
<dt id="Definition228">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
ave_confidence has a value which is a float
ave_coverage has a value which is a float
kbid has a value which is a string
name has a value which is a string
source_id has a value which is a string
source has a value which is a string
functions has a value which is a reference to a list where each element is a MetagenomeAnnotationOTUFunction</pre>

</dd>
</dl>

<h2 id="MetagenomeAnnotation">MetagenomeAnnotation</h2>

<dl>

<dt id="Description286">Description</dt>
<dd>

<p>Structure for the &quot;MetagenomeAnnotation&quot; object</p>

<pre><code>        string type - type of metagenome object
        string name - name of metagenome object
        string kbid - KBase ID of metagenome object
        string source_id - ID used in metagenome source
        string source - source of metagenome data
        string confidence_type - type of confidence score
        list&lt;MetagenomeAnnotationOTU&gt; otus - list of otus in metagenome</code></pre>

</dd>
<dt id="Definition229">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
type has a value which is a string
name has a value which is a string
kbid has a value which is a string
source_id has a value which is a string
source has a value which is a string
confidence_type has a value which is a string
otus has a value which is a reference to a list where each element is a MetagenomeAnnotationOTU</pre>

</dd>
</dl>

<h2 id="import_metagenome_annotation_params">import_metagenome_annotation_params</h2>

<dl>

<dt id="Description287">Description</dt>
<dd>

<p>Input parameters for the &quot;import_metagenome_annotation&quot; function.</p>

<pre><code>        string metaanno_uid - ID to save metagenome in workspace
        workspace_id workspace - ID of workspace for metagenome object
        string source_id - ID used in metagenome data source
        string source - metagenome data source
        string type - type of metagenome
        string confidence_type - type of confidence score
        string name - name of metagenome
        list&lt;tuple&lt;list&lt;string&gt; genes,string functional_role,string otu,int abundance,float confidence,string confidence_type&gt;&gt; annotations;
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition230">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
metaanno_uid has a value which is a string
workspace has a value which is a workspace_id
source_id has a value which is a string
source has a value which is a string
type has a value which is a string
confidence_type has a value which is a string
name has a value which is a string
annotations has a value which is a reference to a list where each element is a reference to a list containing 5 items:
0: (genes) a reference to a list where each element is a string
1: (functional_role) a string
2: (otu) a string
3: (abundance) an int
4: (confidence) a floatauth has a value which is a string</pre>

</dd>
</dl>

<h2 id="models_to_community_model_params">models_to_community_model_params</h2>

<dl>

<dt id="Description288">Description</dt>
<dd>

<p>Input parameters for the &quot;models_to_community_model&quot; function.</p>

<pre><code>        string model_uid - ID of community model
        workspace_id workspace - workspace where community model should be saved
        string name - name of community model
        list&lt;tuple&lt;string model_uid,string model_ws,float abundance&gt;&gt; models - models to be merged into community model
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition231">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model_uid has a value which is a string
workspace has a value which is a workspace_id
name has a value which is a string
models has a value which is a reference to a list where each element is a reference to a list containing 3 items:
0: (model_uid) a string
1: (model_ws) a string
2: (abundance) a floatauth has a value which is a string</pre>

</dd>
</dl>

<h2 id="metagenome_to_fbamodels_params">metagenome_to_fbamodels_params</h2>

<dl>

<dt id="Description289">Description</dt>
<dd>

<p>Input parameters for the &quot;metagenome_to_fbamodel&quot; function.</p>

<pre><code>        string model_uid - ID of community model
        workspace_id workspace - workspace where community model should be saved
        string name - name of community model
        list&lt;tuple&lt;string model_uid,workspace_id model_ws,float abundance&gt;&gt; models - models to be merged into community model
        string auth - the authentication token of the KBase account changing workspace permissions; must have &#39;admin&#39; privelages to workspace (an optional argument; user is &quot;public&quot; if auth is not provided)</code></pre>

</dd>
<dt id="Definition232">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model_uids has a value which is a reference to a hash where the key is a string and the value is a string
workspace has a value which is a workspace_id
metaanno_uid has a value which is a string
metaanno_ws has a value which is a workspace_id
min_abundance has a value which is a float
confidence_threshold has a value which is a float
max_otu_models has a value which is an int
min_reactions has a value which is an int
templates has a value which is a reference to a hash where the key is a string and the value is a reference to a list containing 2 items:
0: (template_ws) a workspace_id
1: (template_uid) a template_idauth has a value which is a string</pre>

</dd>
</dl>

<h2 id="sample_id1">sample_id</h2>

<dl>

<dt id="Description290">Description</dt>
<dd>

<p>ID of gene expression sample</p>

</dd>
<dt id="Definition233">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="series_id1">series_id</h2>

<dl>

<dt id="Description291">Description</dt>
<dd>

<p>ID of gene expression sample series</p>

</dd>
<dt id="Definition234">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="measurement">measurement</h2>

<dl>

<dt id="Description292">Description</dt>
<dd>

<p>Normilized gene expression value</p>

</dd>
<dt id="Definition235">Definition</dt>
<dd>



<pre>
a float
</pre>

</dd>
</dl>

<h2 id="ExpressionDataSample">ExpressionDataSample</h2>

<dl>

<dt id="Definition236">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
sample_id has a value which is a string
data_expression_levels_for_sample has a value which is a reference to a hash where the key is a feature_id and the value is a measurement</pre>

</dd>
</dl>

<h2 id="import_expression_params">import_expression_params</h2>

<dl>

<dt id="Description293">Description</dt>
<dd>

<p>Input parameters for the &quot;simulate_expression&quot; function.</p>

<pre><code>               mapping&lt;sample_id, ExpressionDataSample&gt; expression_data_sample_series - gene expression data (a required argument)
        series_id series -  ID of series (a required argument)
        string source_id - ID of the source (an optional argument: default is &#39;&#39;)
        string source_date - Date of the source (an optional argument: default is &#39;&#39;)
        string processing_comments - comment (an optional argument: default is &#39;&#39;)
        string description - description (an optional argument: default is &#39;&#39;)
        workspace_id workspace - workspace to contain the data (an optional argument: default is value of workspace argument)                
        string numerical_interpretation - Numerical interpretation</code></pre>

</dd>
<dt id="Definition237">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
expression_data_sample_series has a value which is a reference to a hash where the key is a sample_id and the value is an ExpressionDataSample
series has a value which is a series_id
source_id has a value which is a string
source_date has a value which is a string
description has a value which is a string
processing_comments has a value which is a string
workspace has a value which is a workspace_id
genome_id has a value which is a genome_id
numerical_interpretation has a value which is a string</pre>

</dd>
</dl>

<h2 id="effector">effector</h2>

<dl>

<dt id="Description294">Description</dt>
<dd>

<p>Import RegPrecise regulome.</p>

</dd>
<dt id="Definition238">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
name has a value which is a string
class has a value which is a string</pre>

</dd>
</dl>

<h2 id="locus">locus</h2>

<dl>

<dt id="Definition239">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
name has a value which is a string
locus has a value which is a string</pre>

</dd>
</dl>

<h2 id="operon">operon</h2>

<dl>

<dt id="Definition240">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a locus
</pre>

</dd>
</dl>

<h2 id="regulon">regulon</h2>

<dl>

<dt id="Definition241">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
operons has a value which is a reference to a list where each element is an operon
transcription_factor has a value which is a locus
effectors has a value which is a reference to a list where each element is an effector
sign has a value which is a string</pre>

</dd>
</dl>

<h2 id="import_regulome_params">import_regulome_params</h2>

<dl>

<dt id="Definition242">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
regulons has a value which is a reference to a list where each element is a regulon
workspace has a value which is a workspace_id
genome_workspace has a value which is a workspace_id
genome_id has a value which is a genome_id</pre>

</dd>
</dl>

<h2 id="CreatePromConstraintParameters">CreatePromConstraintParameters</h2>

<dl>

<dt id="Description295">Description</dt>
<dd>

<p>Named parameters for &#39;create_promconstraint&#39; method. Currently all options are required.</p>

<pre><code>    genome_id genome_id             - the workspace ID of the genome to link to the prom object
    series_id series_id     - the workspace ID of the expression data collection needed to
                                                   build the PROM constraints.
    regulome_id  regulome_id        - the workspace ID of the regulatory network data to use
    promconstraint_id promconstraint_id - the the workspace ID for the new PROM constraint</code></pre>

</dd>
<dt id="Definition243">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
genome_id has a value which is a genome_id
series_id has a value which is a series_id
regulome_id has a value which is a regulome_id
promconstraint_id has a value which is a promconstraint_id</pre>

</dd>
</dl>

<h2 id="add_biochemistry_compounds_params">add_biochemistry_compounds_params</h2>

<dl>

<dt id="Description296">Description</dt>
<dd>

<p>Add specified compounds to specified biochemistry</p>

</dd>
<dt id="Definition244">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
compounds has a value which is a reference to a list where each element is a reference to a list containing 9 items:
0: (abbreviation) a string
1: (name) a string
2: (aliases) a reference to a list where each element is a string
3: (formula) a string
4: (charge) a float
5: (isCofactor) a bool
6: (structureString) a string
7: (structureType) a string
8: (id) a stringworkspace has a value which is a string
biochemistry has a value which is a string
biochemistry_ws has a value which is a string
output_id has a value which is a string</pre>

</dd>
</dl>

<h2 id="update_object_references_params">update_object_references_params</h2>

<dl>

<dt id="Description297">Description</dt>
<dd>

<p>Update object references</p>

</dd>
<dt id="Definition245">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
object has a value which is a string
object_workspace has a value which is a string
original_object has a value which is a string
original_workspace has a value which is a string
original_instance has a value which is a string
reference_field has a value which is a string
newobject has a value which is a string
newobject_workspace has a value which is a string
newobject_instance has a value which is a string
create_newobject has a value which is a bool
update_subrefs has a value which is a bool
output_id has a value which is a string
workspace has a value which is a string</pre>

</dd>
</dl>

<h2 id="add_reactions_params">add_reactions_params</h2>

<dl>

<dt id="Description298">Description</dt>
<dd>

<p>******************************************************************************** Functions relating to editing of genomes and models ********************************************************************************</p>

</dd>
<dt id="Definition246">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a string
model_workspace has a value which is a string
output_id has a value which is a string
workspace has a value which is a string
reactions has a value which is a reference to a list where each element is a reference to a list containing 9 items:
0: (reaction_id) a string
1: (compartment) a string
2: (direction) a string
3: (gpr) a string
4: (pathway) a string
5: (name) a string
6: (reference) a string
7: (enzyme) a string
8: (equation) a string</pre>

</dd>
</dl>

<h2 id="remove_reactions_params">remove_reactions_params</h2>

<dl>

<dt id="Description299">Description</dt>
<dd>

<p>Input parameters for the &quot;remove_reactions&quot; function.</p>

</dd>
<dt id="Definition247">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a string
model_workspace has a value which is a string
output_id has a value which is a string
workspace has a value which is a string
reactions has a value which is a reference to a list where each element is a string</pre>

</dd>
</dl>

<h2 id="modify_reactions_params">modify_reactions_params</h2>

<dl>

<dt id="Description300">Description</dt>
<dd>

<p>Input parameters for the &quot;modify_reactions&quot; function.</p>

</dd>
<dt id="Definition248">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
model has a value which is a string
model_workspace has a value which is a string
output_id has a value which is a string
workspace has a value which is a string
reactions has a value which is a reference to a list where each element is a reference to a list containing 7 items:
0: (reaction_id) a string
1: (direction) a string
2: (gpr) a string
3: (pathway) a string
4: (name) a string
5: (reference) a string
6: (enzyme) a string</pre>

</dd>
</dl>

<h2 id="add_features_params">add_features_params</h2>

<dl>

<dt id="Description301">Description</dt>
<dd>

<p>Input parameters for the &quot;add_features&quot; function.</p>

</dd>
<dt id="Definition249">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
genome has a value which is a string
genome_workspace has a value which is a string
output_id has a value which is a string
workspace has a value which is a string
genes has a value which is a reference to a list where each element is a reference to a list containing 9 items:
0: (feature) a feature_id
1: (function) a string
2: (type) a string
3: (aliases) a reference to a list where each element is a string
4: (publications) a reference to a list where each element is a string
5: (annotations) a reference to a list where each element is a string
6: (protein_translation) a string
7: (dna_sequence) a string
8: (locations) a reference to a list where each element is a reference to a list containing 4 items:
	0: a string
	1: an int
	2: a string
	3: an int</pre>

</dd>
</dl>

<h2 id="remove_features_params">remove_features_params</h2>

<dl>

<dt id="Description302">Description</dt>
<dd>

<p>Input parameters for the &quot;remove_features&quot; function.</p>

</dd>
<dt id="Definition250">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
genome has a value which is a string
genome_workspace has a value which is a string
output_id has a value which is a string
workspace has a value which is a string
features has a value which is a reference to a list where each element is a string</pre>

</dd>
</dl>

<h2 id="modify_features_params">modify_features_params</h2>

<dl>

<dt id="Description303">Description</dt>
<dd>

<p>Input parameters for the &quot;modify_genes&quot; function.</p>

</dd>
<dt id="Definition251">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
genome has a value which is a string
genome_workspace has a value which is a string
output_id has a value which is a string
workspace has a value which is a string
genes has a value which is a reference to a list where each element is a reference to a list containing 9 items:
0: (feature) a feature_id
1: (function) a string
2: (type) a string
3: (aliases) a reference to a list where each element is a string
4: (publications) a reference to a list where each element is a string
5: (annotations) a reference to a list where each element is a string
6: (protein_translation) a string
7: (dna_sequence) a string
8: (locations) a reference to a list where each element is a reference to a list containing 4 items:
	0: a string
	1: an int
	2: a string
	3: an int</pre>

</dd>
</dl>

<h2 id="import_trainingset_params">import_trainingset_params</h2>

<dl>

<dt id="Description304">Description</dt>
<dd>

<p>******************************************************************************** Functions relating to classification of genomes ********************************************************************************</p>

</dd>
<dt id="Definition252">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
workspace_training_set has a value which is a reference to a list where each element is a reference to a list containing 3 items:
0: (workspace_id) a string
1: (genome_id) a string
2: (class) a stringexternal_training_set has a value which is a reference to a list where each element is a reference to a list containing 4 items:
0: (database) a string
1: (genome_id) a string
2: (class) a string
3: (attributes) a reference to a list where each element is a stringdescription has a value which is a string
class_data has a value which is a reference to a list where each element is a reference to a list containing 2 items:
0: (class) a string
1: (description) a stringattribute_type has a value which is a string
preload_attributes has a value which is a bool
workspace has a value which is a string
output_id has a value which is a string</pre>

</dd>
</dl>

<h2 id="preload_trainingset_params">preload_trainingset_params</h2>

<dl>

<dt id="Description305">Description</dt>
<dd>

<p>Input parameters for the &quot;preload_trainingset&quot; function.</p>

</dd>
<dt id="Definition253">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
trainingset has a value which is a string
trainingset_ws has a value which is a string
attribute_type has a value which is a string
workspace has a value which is a string
output_id has a value which is a string</pre>

</dd>
</dl>

<h2 id="build_classifier_params">build_classifier_params</h2>

<dl>

<dt id="Description306">Description</dt>
<dd>

<p>Input parameters for the &quot;build_classifier&quot; function.</p>

</dd>
<dt id="Definition254">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
trainingset has a value which is a string
trainingset_ws has a value which is a string
attribute_type has a value which is a string
classifier_type has a value which is a string
workspace has a value which is a string
output_id has a value which is a string</pre>

</dd>
</dl>

<h2 id="classify_genomes_params">classify_genomes_params</h2>

<dl>

<dt id="Description307">Description</dt>
<dd>

<p>Input parameters for the &quot;classify_genomes&quot; function.</p>

</dd>
<dt id="Definition255">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
workspace_genomes has a value which is a reference to a list where each element is a reference to a list containing 2 items:
0: (workspace_id) a string
1: (genome_id) a stringexternal_genomes has a value which is a reference to a list where each element is a reference to a list containing 2 items:
0: (database) a string
1: (genome_id) a stringworkspace has a value which is a string
output_id has a value which is a string
classifier_ws has a value which is a string
classifier has a value which is a string</pre>

</dd>
</dl>

<h2 id="build_tissue_model_params">build_tissue_model_params</h2>

<dl>

<dt id="Description308">Description</dt>
<dd>

<p>******************************************************************************** Functions relating to modeling of expression data ********************************************************************************</p>

</dd>
<dt id="Definition256">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
expsample_ws has a value which is a string
expsample has a value which is a string
model_ws has a value which is a string
model has a value which is a string
workspace has a value which is a string
output_id has a value which is a string</pre>

</dd>
</dl>


</body>

</html>


