<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>Central Store Application API</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:root@localhost" />
</head>

<body style="background-color: white">



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#Basic-Themes">Basic Themes</a></li>
      <li><a href="#Batching-queries">Batching queries</a></li>
    </ul>
  </li>
  <li><a href="#METHODS">METHODS</a>
    <ul>
      <li><a href="#fids_to_annotations">fids_to_annotations</a></li>
      <li><a href="#fids_to_functions">fids_to_functions</a></li>
      <li><a href="#fids_to_literature">fids_to_literature</a></li>
      <li><a href="#fids_to_protein_families">fids_to_protein_families</a></li>
      <li><a href="#fids_to_roles">fids_to_roles</a></li>
      <li><a href="#fids_to_subsystems">fids_to_subsystems</a></li>
      <li><a href="#fids_to_co_occurring_fids">fids_to_co_occurring_fids</a></li>
      <li><a href="#fids_to_locations">fids_to_locations</a></li>
      <li><a href="#locations_to_fids">locations_to_fids</a></li>
      <li><a href="#alleles_to_bp_locs">alleles_to_bp_locs</a></li>
      <li><a href="#region_to_fids">region_to_fids</a></li>
      <li><a href="#region_to_alleles">region_to_alleles</a></li>
      <li><a href="#alleles_to_traits">alleles_to_traits</a></li>
      <li><a href="#traits_to_alleles">traits_to_alleles</a></li>
      <li><a href="#ous_with_trait">ous_with_trait</a></li>
      <li><a href="#locations_to_dna_sequences">locations_to_dna_sequences</a></li>
      <li><a href="#proteins_to_fids">proteins_to_fids</a></li>
      <li><a href="#proteins_to_protein_families">proteins_to_protein_families</a></li>
      <li><a href="#proteins_to_literature">proteins_to_literature</a></li>
      <li><a href="#proteins_to_functions">proteins_to_functions</a></li>
      <li><a href="#proteins_to_roles">proteins_to_roles</a></li>
      <li><a href="#roles_to_proteins">roles_to_proteins</a></li>
      <li><a href="#roles_to_subsystems">roles_to_subsystems</a></li>
      <li><a href="#roles_to_protein_families">roles_to_protein_families</a></li>
      <li><a href="#fids_to_coexpressed_fids">fids_to_coexpressed_fids</a></li>
      <li><a href="#protein_families_to_fids">protein_families_to_fids</a></li>
      <li><a href="#protein_families_to_proteins">protein_families_to_proteins</a></li>
      <li><a href="#protein_families_to_functions">protein_families_to_functions</a></li>
      <li><a href="#protein_families_to_co_occurring_families">protein_families_to_co_occurring_families</a></li>
      <li><a href="#co_occurrence_evidence">co_occurrence_evidence</a></li>
      <li><a href="#contigs_to_sequences">contigs_to_sequences</a></li>
      <li><a href="#contigs_to_lengths">contigs_to_lengths</a></li>
      <li><a href="#contigs_to_md5s">contigs_to_md5s</a></li>
      <li><a href="#md5s_to_genomes">md5s_to_genomes</a></li>
      <li><a href="#genomes_to_md5s">genomes_to_md5s</a></li>
      <li><a href="#genomes_to_contigs">genomes_to_contigs</a></li>
      <li><a href="#genomes_to_fids">genomes_to_fids</a></li>
      <li><a href="#genomes_to_taxonomies">genomes_to_taxonomies</a></li>
      <li><a href="#genomes_to_subsystems">genomes_to_subsystems</a></li>
      <li><a href="#subsystems_to_genomes">subsystems_to_genomes</a></li>
      <li><a href="#subsystems_to_fids">subsystems_to_fids</a></li>
      <li><a href="#subsystems_to_roles">subsystems_to_roles</a></li>
      <li><a href="#subsystems_to_spreadsheets">subsystems_to_spreadsheets</a></li>
      <li><a href="#all_roles_used_in_models">all_roles_used_in_models</a></li>
      <li><a href="#complexes_to_complex_data">complexes_to_complex_data</a></li>
      <li><a href="#genomes_to_genome_data">genomes_to_genome_data</a></li>
      <li><a href="#fids_to_regulon_data">fids_to_regulon_data</a></li>
      <li><a href="#regulons_to_fids">regulons_to_fids</a></li>
      <li><a href="#fids_to_feature_data">fids_to_feature_data</a></li>
      <li><a href="#equiv_sequence_assertions">equiv_sequence_assertions</a></li>
      <li><a href="#fids_to_atomic_regulons">fids_to_atomic_regulons</a></li>
      <li><a href="#atomic_regulons_to_fids">atomic_regulons_to_fids</a></li>
      <li><a href="#fids_to_protein_sequences">fids_to_protein_sequences</a></li>
      <li><a href="#fids_to_proteins">fids_to_proteins</a></li>
      <li><a href="#fids_to_dna_sequences">fids_to_dna_sequences</a></li>
      <li><a href="#roles_to_fids">roles_to_fids</a></li>
      <li><a href="#reactions_to_complexes">reactions_to_complexes</a></li>
      <li><a href="#aliases_to_fids">aliases_to_fids</a></li>
      <li><a href="#aliases_to_fids_by_source">aliases_to_fids_by_source</a></li>
      <li><a href="#source_ids_to_fids">source_ids_to_fids</a></li>
      <li><a href="#external_ids_to_fids">external_ids_to_fids</a></li>
      <li><a href="#reaction_strings">reaction_strings</a></li>
      <li><a href="#roles_to_complexes">roles_to_complexes</a></li>
      <li><a href="#complexes_to_roles">complexes_to_roles</a></li>
      <li><a href="#fids_to_subsystem_data">fids_to_subsystem_data</a></li>
      <li><a href="#representative">representative</a></li>
      <li><a href="#otu_members">otu_members</a></li>
      <li><a href="#otus_to_representatives">otus_to_representatives</a></li>
      <li><a href="#fids_to_genomes">fids_to_genomes</a></li>
      <li><a href="#text_search">text_search</a></li>
      <li><a href="#corresponds">corresponds</a></li>
      <li><a href="#corresponds_from_sequences">corresponds_from_sequences</a></li>
      <li><a href="#close_genomes">close_genomes</a></li>
      <li><a href="#representative_sequences">representative_sequences</a></li>
      <li><a href="#align_sequences">align_sequences</a></li>
      <li><a href="#build_tree">build_tree</a></li>
      <li><a href="#alignment_by_id">alignment_by_id</a></li>
      <li><a href="#tree_by_id">tree_by_id</a></li>
      <li><a href="#all_entities">all_entities</a></li>
      <li><a href="#all_relationships">all_relationships</a></li>
      <li><a href="#get_entity">get_entity</a></li>
      <li><a href="#get_relationship">get_relationship</a></li>
      <li><a href="#version">version</a></li>
    </ul>
  </li>
  <li><a href="#TYPES">TYPES</a>
    <ul>
      <li><a href="#annotator">annotator</a></li>
      <li><a href="#annotation_time">annotation_time</a></li>
      <li><a href="#comment">comment</a></li>
      <li><a href="#fid">fid</a></li>
      <li><a href="#protein_family">protein_family</a></li>
      <li><a href="#role">role</a></li>
      <li><a href="#subsystem">subsystem</a></li>
      <li><a href="#variant">variant</a></li>
      <li><a href="#variant_of_subsystem">variant_of_subsystem</a></li>
      <li><a href="#variant_subsystem_pairs">variant_subsystem_pairs</a></li>
      <li><a href="#type_of_fid">type_of_fid</a></li>
      <li><a href="#types_of_fids">types_of_fids</a></li>
      <li><a href="#length">length</a></li>
      <li><a href="#begin">begin</a></li>
      <li><a href="#strand">strand</a></li>
      <li><a href="#contig">contig</a></li>
      <li><a href="#region_of_dna">region_of_dna</a></li>
      <li><a href="#location">location</a></li>
      <li><a href="#locations">locations</a></li>
      <li><a href="#region_of_dna_string">region_of_dna_string</a></li>
      <li><a href="#region_of_dna_strings">region_of_dna_strings</a></li>
      <li><a href="#location_string">location_string</a></li>
      <li><a href="#dna">dna</a></li>
      <li><a href="#function">function</a></li>
      <li><a href="#protein">protein</a></li>
      <li><a href="#md5">md5</a></li>
      <li><a href="#genome">genome</a></li>
      <li><a href="#taxonomic_group">taxonomic_group</a></li>
      <li><a href="#annotation">annotation</a></li>
      <li><a href="#pubref">pubref</a></li>
      <li><a href="#scored_fid">scored_fid</a></li>
      <li><a href="#annotations">annotations</a></li>
      <li><a href="#pubrefs">pubrefs</a></li>
      <li><a href="#roles">roles</a></li>
      <li><a href="#optional">optional</a></li>
      <li><a href="#role_with_flag">role_with_flag</a></li>
      <li><a href="#roles_with_flags">roles_with_flags</a></li>
      <li><a href="#scored_fids">scored_fids</a></li>
      <li><a href="#proteins">proteins</a></li>
      <li><a href="#functions">functions</a></li>
      <li><a href="#taxonomic_groups">taxonomic_groups</a></li>
      <li><a href="#subsystems">subsystems</a></li>
      <li><a href="#contigs">contigs</a></li>
      <li><a href="#md5s">md5s</a></li>
      <li><a href="#genomes">genomes</a></li>
      <li><a href="#pair_of_fids">pair_of_fids</a></li>
      <li><a href="#pairs_of_fids">pairs_of_fids</a></li>
      <li><a href="#protein_families">protein_families</a></li>
      <li><a href="#score">score</a></li>
      <li><a href="#evidence">evidence</a></li>
      <li><a href="#fids">fids</a></li>
      <li><a href="#row">row</a></li>
      <li><a href="#fid_function_pair">fid_function_pair</a></li>
      <li><a href="#fid_function_pairs">fid_function_pairs</a></li>
      <li><a href="#fc_protein_family">fc_protein_family</a></li>
      <li><a href="#fc_protein_families">fc_protein_families</a></li>
      <li><a href="#allele">allele</a></li>
      <li><a href="#alleles">alleles</a></li>
      <li><a href="#trait">trait</a></li>
      <li><a href="#traits">traits</a></li>
      <li><a href="#ou">ou</a></li>
      <li><a href="#ous">ous</a></li>
      <li><a href="#bp_loc">bp_loc</a></li>
      <li><a href="#measurement_type">measurement_type</a></li>
      <li><a href="#measurement_value">measurement_value</a></li>
      <li><a href="#aux">aux</a></li>
      <li><a href="#fields">fields</a></li>
      <li><a href="#complex">complex</a></li>
      <li><a href="#complex_with_flag">complex_with_flag</a></li>
      <li><a href="#complexes_with_flags">complexes_with_flags</a></li>
      <li><a href="#complexes">complexes</a></li>
      <li><a href="#name">name</a></li>
      <li><a href="#reaction">reaction</a></li>
      <li><a href="#reactions">reactions</a></li>
      <li><a href="#complex_data">complex_data</a></li>
      <li><a href="#genome_data">genome_data</a></li>
      <li><a href="#regulon">regulon</a></li>
      <li><a href="#regulons">regulons</a></li>
      <li><a href="#regulon_data">regulon_data</a></li>
      <li><a href="#regulons_data">regulons_data</a></li>
      <li><a href="#feature_data">feature_data</a></li>
      <li><a href="#expert">expert</a></li>
      <li><a href="#source">source</a></li>
      <li><a href="#id">id</a></li>
      <li><a href="#function_assertion">function_assertion</a></li>
      <li><a href="#function_assertions">function_assertions</a></li>
      <li><a href="#atomic_regulon">atomic_regulon</a></li>
      <li><a href="#atomic_regulon_size">atomic_regulon_size</a></li>
      <li><a href="#atomic_regulon_size_pair">atomic_regulon_size_pair</a></li>
      <li><a href="#atomic_regulon_size_pairs">atomic_regulon_size_pairs</a></li>
      <li><a href="#atomic_regulons">atomic_regulons</a></li>
      <li><a href="#protein_sequence">protein_sequence</a></li>
      <li><a href="#dna_sequence">dna_sequence</a></li>
      <li><a href="#alias">alias</a></li>
      <li><a href="#aliases">aliases</a></li>
      <li><a href="#external_id">external_id</a></li>
      <li><a href="#external_ids">external_ids</a></li>
      <li><a href="#name_parameter">name_parameter</a></li>
      <li><a href="#ss_var_role_tuple">ss_var_role_tuple</a></li>
      <li><a href="#ss_var_role_tuples">ss_var_role_tuples</a></li>
      <li><a href="#genome_name">genome_name</a></li>
      <li><a href="#entity_name">entity_name</a></li>
      <li><a href="#weight">weight</a></li>
      <li><a href="#field_name">field_name</a></li>
      <li><a href="#search_hit">search_hit</a></li>
      <li><a href="#correspondence">correspondence</a></li>
      <li><a href="#sequence">sequence</a></li>
      <li><a href="#seq_triple">seq_triple</a></li>
      <li><a href="#seq_set">seq_set</a></li>
      <li><a href="#id_set">id_set</a></li>
      <li><a href="#alignment">alignment</a></li>
      <li><a href="#rep_seq_parms">rep_seq_parms</a></li>
      <li><a href="#muscle_parms_t">muscle_parms_t</a></li>
      <li><a href="#mafft_parms_t">mafft_parms_t</a></li>
      <li><a href="#align_seq_parms">align_seq_parms</a></li>
      <li><a href="#newick_tree">newick_tree</a></li>
      <li><a href="#build_tree_parms">build_tree_parms</a></li>
      <li><a href="#aln_id">aln_id</a></li>
      <li><a href="#tree_id">tree_id</a></li>
      <li><a href="#entity_name1">entity_name</a></li>
      <li><a href="#entity_names">entity_names</a></li>
      <li><a href="#relationship_name">relationship_name</a></li>
      <li><a href="#relationship_names">relationship_names</a></li>
      <li><a href="#field_name1">field_name</a></li>
      <li><a href="#boolean">boolean</a></li>
      <li><a href="#field_info">field_info</a></li>
      <li><a href="#entity_info">entity_info</a></li>
      <li><a href="#relationship_info">relationship_info</a></li>
    </ul>
  </li>
</ul>

<h1 id="NAME">NAME</h1>

<p>CDMI_API</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>The CDMI_API defines the component of the Kbase API that supports interaction with instances of the CDM (Central Data Model). A basic familiarity with these routines will allow the user to extract data from the CS (Central Store). We anticipate supporting numerous sparse CDMIs in the PS (Persistent Store).</p>

<h2 id="Basic-Themes">Basic Themes</h2>

<p>There are several broad categories of routines supported in the CDMI-API.</p>

<p>The simplest is set of &quot;get entity&quot; routines -- each returning data extracted from instances of a single entity type. These routines all take as input a list of ids referencing instances of a single type of entity. They construct as output a mapping which takes as input an id and associates as output a set of fields from that instance of the entity. Each routine allows the user to specify which fields are desired.</p>

<p>For example, assume you have an input file &quot;Staphylococci,&quot; which is a list of genome IDs for each species of Staphylococcus in the database. The get_entity_Genome command is used to retrieve detailed information about each genome in the file. By using different modifiers, you can specify what kind of information you want to display. In this example, the modifier &quot;contigs&quot; was used. Thus, the number next to the genome ID in the output file indicates the number of contigs each Staphylococcus genome has. For a list of available modifiers relating to each identity, please refer to the ER model.</p>

<pre><code>        &gt; / cat Staphylococci | cut -f 1 | get_entity_Genome - f contigs
        kb|g.134        2
        kb|g.636        1
        kb|g.2506        15
        kb|g.9303        1
        kb|g.3801        87
        kb|g.2025        46
        kb|g.2516        13
        kb|g.2603        33
        kb|g.19928        2
        kb|g.1852        131
        kb|g.8476        1
        kb|g.2742        46</code></pre>

<p>To use these routines effectively, a user will need to gradually become familiar with the entities supported in the CDM. We suggest perusing the entity-relationship model that underlies the CDM to get a good introduction.</p>

<p>The next simplest set of routines provide the &quot;get relationship&quot; routines. These take as input a list of ids for a specific entity type, and the give access to the relationship nodes associated with each entity. Thus, get_relationship_WasSubmittedBy takes the input genome ID and outputs the ID with an added column showing the source of that particular genome. It is essential to be able to navigate the ER model to successfully implement these commands, since not all relationship types are applicable to each entity.</p>

<pre><code>        &gt; / echo &#39;kb|g.0&#39; | get_relationship_WasSubmittedBy -to id
        kb|g.0        SEED</code></pre>

<p>Of the remaining CDMI-API routines, most are used to extract data by &quot;crossing one or more relationships&quot;. Thus,</p>

<pre><code>        my $references = $kbO-&gt;fids_to_literature($fids)</code></pre>

<p>takes as input a list of feature ids referenced by the variable $fids. It creates a hash ($references) which maps each input key to a list of literature references. The construction of the literature references for a given ID involves crossing relationships from the entity &#39;Feature&#39; to &#39;ProteinSequence&#39; to &#39;Publication&#39;. We have attempted to package this specific search in a convenient form. We anticipate that the number of queries of this last class will grow (especially as new entities are added to the model).</p>

<h2 id="Batching-queries">Batching queries</h2>

<p>A majority of the CS-API routines take a list of ids as input. Each id may be thought of as input to a query that produces an output result. We support processing an input list, since the performance (which is usually governed by network interactions) is much better if you process a batch of items, rather than invoking the API repeatedly for each of the ids. Normally, the output would be a mapping (a hash for Perl versions) from the input ids to the output results. Thus, a routine like</p>

<pre><code>             fids_to_literature</code></pre>

<p>will take a list of feature ids as input. The returned value will be a mapping from feature ids (fids) to publication references.</p>

<p>It is a little inconvenient to batch your requests by supplying a list of fids, but the performance will be much better in most cases. Please note that you are controlling the granularity of each request, and in most cases the size of the input list is not critical. However, you should note that while batching up hundreds or thousands of input ids at a time should work just fine, millions may well cause things to break (e.g., you may exhaust local memory in your machine as the output results are returned). As machines get larger, the appropriate size of the input lists may become largely irrelevant. For now, we recommend that you experiment a bit and use common sense.</p>

<h1 id="METHODS">METHODS</h1>

<h2 id="fids_to_annotations">fids_to_annotations</h2>

<pre><code>  $return = $obj-&gt;fids_to_annotations($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is an annotations
fids is a reference to a list where each element is a fid
fid is a string
annotations is a reference to a list where each element is an annotation
annotation is a reference to a list containing 3 items:
	0: a comment
	1: an annotator
	2: an annotation_time
comment is a string
annotator is a string
annotation_time is an int</pre>

</dd>
<dt id="Description">Description</dt>
<dd>

<p>This routine takes as input a list of fids. It retrieves the existing annotations for each fid, including the text of the annotation, who made the annotation and when (as seconds from the epoch).</p>

</dd>
</dl>

<h2 id="fids_to_functions">fids_to_functions</h2>

<pre><code>  $return = $obj-&gt;fids_to_functions($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types1">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a function
fids is a reference to a list where each element is a fid
fid is a string
function is a string</pre>

</dd>
<dt id="Description1">Description</dt>
<dd>

<p>This routine takes as input a list of fids and returns a mapping from the fids to their assigned functions.</p>

</dd>
</dl>

<h2 id="fids_to_literature">fids_to_literature</h2>

<pre><code>  $return = $obj-&gt;fids_to_literature($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types2">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a pubrefs
fids is a reference to a list where each element is a fid
fid is a string
pubrefs is a reference to a list where each element is a pubref
pubref is a reference to a list containing 3 items:
	0: (id) a string
	1: (link) a string
	2: (title) a string</pre>

</dd>
<dt id="Description2">Description</dt>
<dd>

<p>We try to associate features and publications, when the publications constitute supporting evidence of the function. We connect a paper to a feature when we believe that an &quot;expert&quot; has asserted that the function of the feature is basically what we have associated with the feature. Thus, we might attach a paper reporting the crystal structure of a protein, even though the paper is clearly not the paper responsible for the original characterization. Our position in this matter is somewhat controversial, but we are seeking to characterize some assertions as relatively solid, and this strategy seems to support that goal. Please note that we certainly wish we could also capture original publications, and when experts can provide those connections, we hope that they will help record the associations.</p>

</dd>
</dl>

<h2 id="fids_to_protein_families">fids_to_protein_families</h2>

<pre><code>  $return = $obj-&gt;fids_to_protein_families($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types3">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a protein_families
fids is a reference to a list where each element is a fid
fid is a string
protein_families is a reference to a list where each element is a protein_family
protein_family is a string</pre>

</dd>
<dt id="Description3">Description</dt>
<dd>

<p>Kbase supports the creation and maintence of protein families. Each family is intended to contain a set of isofunctional homologs. Currently, the families are collections of translations of features, rather than of just protein sequences (represented by md5s, for example). fids_to_protein_families supports access to the features that have been grouped into a family. Ideally, each feature in a family would have the same assigned function. This is not always true, but probably should be.</p>

</dd>
</dl>

<h2 id="fids_to_roles">fids_to_roles</h2>

<pre><code>  $return = $obj-&gt;fids_to_roles($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types4">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a roles
fids is a reference to a list where each element is a fid
fid is a string
roles is a reference to a list where each element is a role
role is a string</pre>

</dd>
<dt id="Description4">Description</dt>
<dd>

<p>Given a feature, one can get the set of roles it implements using fid_to_roles. Remember, a protein can be multifunctional -- implementing several roles. This can occur due to fusions or to broad specificity of substrate.</p>

</dd>
</dl>

<h2 id="fids_to_subsystems">fids_to_subsystems</h2>

<pre><code>  $return = $obj-&gt;fids_to_subsystems($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types5">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a subsystems
fids is a reference to a list where each element is a fid
fid is a string
subsystems is a reference to a list where each element is a subsystem
subsystem is a string</pre>

</dd>
<dt id="Description5">Description</dt>
<dd>

<p>fids in subsystems normally have somewhat more reliable assigned functions than those not in subsystems. Hence, it is common to ask &quot;Is this protein-encoding gene included in any subsystems?&quot; fids_to_subsystems can be used to see which subsystems contain a fid (or, you can submit as input a set of fids and get the subsystems for each).</p>

</dd>
</dl>

<h2 id="fids_to_co_occurring_fids">fids_to_co_occurring_fids</h2>

<pre><code>  $return = $obj-&gt;fids_to_co_occurring_fids($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types6">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a scored_fids
fids is a reference to a list where each element is a fid
fid is a string
scored_fids is a reference to a list where each element is a scored_fid
scored_fid is a reference to a list containing 2 items:
	0: a fid
	1: (score) a float</pre>

</dd>
<dt id="Description6">Description</dt>
<dd>

<p>One of the most powerful clues to function relates to conserved clusters of genes on the chromosome (in prokaryotic genomes). We have attempted to record pairs of genes that tend to occur close to one another on the chromosome. To meaningfully do this, we need to construct similarity-based mappings between genes in distinct genomes. We have constructed such mappings for many (but not all) genomes maintained in the Kbase CS. The prokaryotic geneomes in the CS are grouped into OTUs by ribosomal RNA (genomes within a single OTU have SSU rRNA that is greater than 97% identical). If two genes occur close to one another (i.e., corresponding genes occur close to one another), then we assign a score, which is the number of distinct OTUs in which such clustering is detected. This allows one to normalize for situations in which hundreds of corresponding genes are detected, but they all come from very closely related genomes.</p>

<p>The significance of the score relates to the number of genomes in the database. We recommend that you take the time to look at a set of scored pairs and determine approximately what percentage appear to be actually related for a few cutoff values.</p>

</dd>
</dl>

<h2 id="fids_to_locations">fids_to_locations</h2>

<pre><code>  $return = $obj-&gt;fids_to_locations($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types7">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a location
fids is a reference to a list where each element is a fid
fid is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig
	1: a begin
	2: a strand
	3: a length
contig is a string
begin is an int
strand is a string
length is an int</pre>

</dd>
<dt id="Description7">Description</dt>
<dd>

<p>A &quot;location&quot; is a sequence of &quot;regions&quot;. A region is a contiguous set of bases in a contig. We work with locations in both the string form and as structures. fids_to_locations takes as input a list of fids. For each fid, a structured location is returned. The location is a list of regions; a region is given as a pointer to a list containing</p>

<pre><code>             the contig,
             the beginning base in the contig (from 1).
             the strand (+ or -), and
             the length</code></pre>

<p>Note that specifying a region using these 4 values allows you to represent a single base-pair region on either strand unambiguously (which giving begin/end pairs does not achieve).</p>

</dd>
</dl>

<h2 id="locations_to_fids">locations_to_fids</h2>

<pre><code>  $return = $obj-&gt;locations_to_fids($region_of_dna_strings)</code></pre>

<dl>

<dt id="Parameter-and-return-types8">Parameter and return types</dt>
<dd>



<pre>
$region_of_dna_strings is a region_of_dna_strings
$return is a reference to a hash where the key is a region_of_dna_string and the value is a fids
region_of_dna_strings is a reference to a list where each element is a region_of_dna_string
region_of_dna_string is a string
fids is a reference to a list where each element is a fid
fid is a string</pre>

</dd>
<dt id="Description8">Description</dt>
<dd>

<p>It is frequently the case that one wishes to look up the genes that occur in a given region of a contig. Location_to_fids can be used to extract such sets of genes for each region in the input set of regions. We define a gene as &quot;occuring&quot; in a region if the location of the gene overlaps the designated region.</p>

</dd>
</dl>

<h2 id="alleles_to_bp_locs">alleles_to_bp_locs</h2>

<pre><code>  $return = $obj-&gt;alleles_to_bp_locs($alleles)</code></pre>

<dl>

<dt id="Parameter-and-return-types9">Parameter and return types</dt>
<dd>



<pre>
$alleles is an alleles
$return is a reference to a hash where the key is an allele and the value is a bp_loc
alleles is a reference to a list where each element is an allele
allele is a string
bp_loc is a reference to a list containing 2 items:
	0: a contig
	1: (position) an int
contig is a string</pre>

</dd>
<dt id="Description9">Description</dt>
<dd>

</dd>
</dl>

<h2 id="region_to_fids">region_to_fids</h2>

<pre><code>  $return = $obj-&gt;region_to_fids($region_of_dna)</code></pre>

<dl>

<dt id="Parameter-and-return-types10">Parameter and return types</dt>
<dd>



<pre>
$region_of_dna is a region_of_dna
$return is a fids
region_of_dna is a reference to a list containing 4 items:
	0: a contig
	1: a begin
	2: a strand
	3: a length
contig is a string
begin is an int
strand is a string
length is an int
fids is a reference to a list where each element is a fid
fid is a string</pre>

</dd>
<dt id="Description10">Description</dt>
<dd>

</dd>
</dl>

<h2 id="region_to_alleles">region_to_alleles</h2>

<pre><code>  $return = $obj-&gt;region_to_alleles($region_of_dna)</code></pre>

<dl>

<dt id="Parameter-and-return-types11">Parameter and return types</dt>
<dd>



<pre>
$region_of_dna is a region_of_dna
$return is a reference to a list where each element is a reference to a list containing 2 items:
	0: an allele
	1: (position) an int
region_of_dna is a reference to a list containing 4 items:
	0: a contig
	1: a begin
	2: a strand
	3: a length
contig is a string
begin is an int
strand is a string
length is an int
allele is a string</pre>

</dd>
<dt id="Description11">Description</dt>
<dd>

</dd>
</dl>

<h2 id="alleles_to_traits">alleles_to_traits</h2>

<pre><code>  $return = $obj-&gt;alleles_to_traits($alleles)</code></pre>

<dl>

<dt id="Parameter-and-return-types12">Parameter and return types</dt>
<dd>



<pre>
$alleles is an alleles
$return is a reference to a hash where the key is an allele and the value is a traits
alleles is a reference to a list where each element is an allele
allele is a string
traits is a reference to a list where each element is a trait
trait is a string</pre>

</dd>
<dt id="Description12">Description</dt>
<dd>

</dd>
</dl>

<h2 id="traits_to_alleles">traits_to_alleles</h2>

<pre><code>  $return = $obj-&gt;traits_to_alleles($traits)</code></pre>

<dl>

<dt id="Parameter-and-return-types13">Parameter and return types</dt>
<dd>



<pre>
$traits is a traits
$return is a reference to a hash where the key is a trait and the value is an alleles
traits is a reference to a list where each element is a trait
trait is a string
alleles is a reference to a list where each element is an allele
allele is a string</pre>

</dd>
<dt id="Description13">Description</dt>
<dd>

</dd>
</dl>

<h2 id="ous_with_trait">ous_with_trait</h2>

<pre><code>  $return = $obj-&gt;ous_with_trait($genome, $trait, $measurement_type, $min_value, $max_value)</code></pre>

<dl>

<dt id="Parameter-and-return-types14">Parameter and return types</dt>
<dd>



<pre>
$genome is a genome
$trait is a trait
$measurement_type is a measurement_type
$min_value is a float
$max_value is a float
$return is a reference to a list where each element is a reference to a list containing 2 items:
	0: an ou
	1: a measurement_value
genome is a string
trait is a string
measurement_type is a string
ou is a string
measurement_value is a float</pre>

</dd>
<dt id="Description14">Description</dt>
<dd>

</dd>
</dl>

<h2 id="locations_to_dna_sequences">locations_to_dna_sequences</h2>

<pre><code>  $dna_seqs = $obj-&gt;locations_to_dna_sequences($locations)</code></pre>

<dl>

<dt id="Parameter-and-return-types15">Parameter and return types</dt>
<dd>



<pre>
$locations is a locations
$dna_seqs is a reference to a list where each element is a reference to a list containing 2 items:
	0: a location
	1: a dna
locations is a reference to a list where each element is a location
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig
	1: a begin
	2: a strand
	3: a length
contig is a string
begin is an int
strand is a string
length is an int
dna is a string</pre>

</dd>
<dt id="Description15">Description</dt>
<dd>

<p>locations_to_dna_sequences takes as input a list of locations (each in the form of a list of regions). The routine constructs 2-tuples composed of</p>

<pre><code>     [the input location,the dna string]</code></pre>

<p>The returned DNA string is formed by concatenating the DNA for each of the regions that make up the location.</p>

</dd>
</dl>

<h2 id="proteins_to_fids">proteins_to_fids</h2>

<pre><code>  $return = $obj-&gt;proteins_to_fids($proteins)</code></pre>

<dl>

<dt id="Parameter-and-return-types16">Parameter and return types</dt>
<dd>



<pre>
$proteins is a proteins
$return is a reference to a hash where the key is a protein and the value is a fids
proteins is a reference to a list where each element is a protein
protein is a string
fids is a reference to a list where each element is a fid
fid is a string</pre>

</dd>
<dt id="Description16">Description</dt>
<dd>

<p>proteins_to_fids takes as input a list of proteins (i.e., a list of md5s) and returns for each a set of protein-encoding fids that have the designated sequence as their translation. That is, for each sequence, the returned fids will be the entire set (within Kbase) that have the sequence as a translation.</p>

</dd>
</dl>

<h2 id="proteins_to_protein_families">proteins_to_protein_families</h2>

<pre><code>  $return = $obj-&gt;proteins_to_protein_families($proteins)</code></pre>

<dl>

<dt id="Parameter-and-return-types17">Parameter and return types</dt>
<dd>



<pre>
$proteins is a proteins
$return is a reference to a hash where the key is a protein and the value is a protein_families
proteins is a reference to a list where each element is a protein
protein is a string
protein_families is a reference to a list where each element is a protein_family
protein_family is a string</pre>

</dd>
<dt id="Description17">Description</dt>
<dd>

<p>Protein families contain a set of isofunctional homologs. proteins_to_protein_families can be used to look up is used to get the set of protein_families containing a specified protein. For performance reasons, you can submit a batch of proteins (i.e., a list of proteins), and for each input protein, you get back a set (possibly empty) of protein_families. Specific collections of families (e.g., FIGfams) usually require that a protein be in at most one family. However, we will be integrating protein families from a number of sources, and so a protein can be in multiple families.</p>

</dd>
</dl>

<h2 id="proteins_to_literature">proteins_to_literature</h2>

<pre><code>  $return = $obj-&gt;proteins_to_literature($proteins)</code></pre>

<dl>

<dt id="Parameter-and-return-types18">Parameter and return types</dt>
<dd>



<pre>
$proteins is a proteins
$return is a reference to a hash where the key is a protein and the value is a pubrefs
proteins is a reference to a list where each element is a protein
protein is a string
pubrefs is a reference to a list where each element is a pubref
pubref is a reference to a list containing 3 items:
	0: (id) a string
	1: (link) a string
	2: (title) a string</pre>

</dd>
<dt id="Description18">Description</dt>
<dd>

<p>The routine proteins_to_literature can be used to extract the list of papers we have associated with specific protein sequences. The user should note that in many cases the association of a paper with a protein sequence is not precise. That is, the paper may actually describe a closely-related protein (that may not yet even be in a sequenced genome). Annotators attempt to use best judgement when associating literature and proteins. Publication references include [pubmed ID,URL for the paper, title of the paper]. In some cases, the URL and title are omitted. In theory, we can extract them from PubMed and we will attempt to do so.</p>

</dd>
</dl>

<h2 id="proteins_to_functions">proteins_to_functions</h2>

<pre><code>  $return = $obj-&gt;proteins_to_functions($proteins)</code></pre>

<dl>

<dt id="Parameter-and-return-types19">Parameter and return types</dt>
<dd>



<pre>
$proteins is a proteins
$return is a reference to a hash where the key is a protein and the value is a fid_function_pairs
proteins is a reference to a list where each element is a protein
protein is a string
fid_function_pairs is a reference to a list where each element is a fid_function_pair
fid_function_pair is a reference to a list containing 2 items:
	0: a fid
	1: a function
fid is a string
function is a string</pre>

</dd>
<dt id="Description19">Description</dt>
<dd>

<p>The routine proteins_to_functions allows users to access functions associated with specific protein sequences. The input proteins are given as a list of MD5 values (these MD5 values each correspond to a specific protein sequence). For each input MD5 value, a list of [feature-id,function] pairs is constructed and returned. Note that there are many cases in which a single protein sequence corresponds to the translation associated with multiple protein-encoding genes, and each may have distinct functions (an undesirable situation, we grant).</p>

<p>This function allows you to access all of the functions assigned (by all annotation groups represented in Kbase) to each of a set of sequences.</p>

</dd>
</dl>

<h2 id="proteins_to_roles">proteins_to_roles</h2>

<pre><code>  $return = $obj-&gt;proteins_to_roles($proteins)</code></pre>

<dl>

<dt id="Parameter-and-return-types20">Parameter and return types</dt>
<dd>



<pre>
$proteins is a proteins
$return is a reference to a hash where the key is a protein and the value is a roles
proteins is a reference to a list where each element is a protein
protein is a string
roles is a reference to a list where each element is a role
role is a string</pre>

</dd>
<dt id="Description20">Description</dt>
<dd>

<p>The routine proteins_to_roles allows a user to gather the set of functional roles that are associated with specifc protein sequences. A single protein sequence (designated by an MD5 value) may have numerous associated functions, since functions are treated as an attribute of the feature, and multiple features may have precisely the same translation. In our experience, it is not uncommon, even for the best annotation teams, to assign distinct functions (and, hence, functional roles) to identical protein sequences.</p>

<p>For each input MD5 value, this routine gathers the set of features (fids) that share the same sequence, collects the associated functions, expands these into functional roles (for multi-functional proteins), and returns the set of roles that results.</p>

<p>Note that, if the user wishes to see the specific features that have the assigned fiunctional roles, they should use proteins_to_functions instead (it returns the fids associated with each assigned function).</p>

</dd>
</dl>

<h2 id="roles_to_proteins">roles_to_proteins</h2>

<pre><code>  $return = $obj-&gt;roles_to_proteins($roles)</code></pre>

<dl>

<dt id="Parameter-and-return-types21">Parameter and return types</dt>
<dd>



<pre>
$roles is a roles
$return is a reference to a hash where the key is a role and the value is a proteins
roles is a reference to a list where each element is a role
role is a string
proteins is a reference to a list where each element is a protein
protein is a string</pre>

</dd>
<dt id="Description21">Description</dt>
<dd>

<p>roles_to_proteins can be used to extract the set of proteins (designated by MD5 values) that currently are believed to implement a given role. Note that the proteins may be multifunctional, meaning that they may be implementing other roles, as well.</p>

</dd>
</dl>

<h2 id="roles_to_subsystems">roles_to_subsystems</h2>

<pre><code>  $return = $obj-&gt;roles_to_subsystems($roles)</code></pre>

<dl>

<dt id="Parameter-and-return-types22">Parameter and return types</dt>
<dd>



<pre>
$roles is a roles
$return is a reference to a hash where the key is a role and the value is a subsystems
roles is a reference to a list where each element is a role
role is a string
subsystems is a reference to a list where each element is a subsystem
subsystem is a string</pre>

</dd>
<dt id="Description22">Description</dt>
<dd>

<p>roles_to_subsystems can be used to access the set of subsystems that include specific roles. The input is a list of roles (i.e., role descriptions), and a mapping is returned as a hash with key role description and values composed of sets of susbsystem names.</p>

</dd>
</dl>

<h2 id="roles_to_protein_families">roles_to_protein_families</h2>

<pre><code>  $return = $obj-&gt;roles_to_protein_families($roles)</code></pre>

<dl>

<dt id="Parameter-and-return-types23">Parameter and return types</dt>
<dd>



<pre>
$roles is a roles
$return is a reference to a hash where the key is a role and the value is a protein_families
roles is a reference to a list where each element is a role
role is a string
protein_families is a reference to a list where each element is a protein_family
protein_family is a string</pre>

</dd>
<dt id="Description23">Description</dt>
<dd>

<p>roles_to_protein_families can be used to locate the protein families containing features that have assigned functions implying that they implement designated roles. Note that for any input role (given as a role description), you may have a set of distinct protein_families returned.</p>

</dd>
</dl>

<h2 id="fids_to_coexpressed_fids">fids_to_coexpressed_fids</h2>

<pre><code>  $return = $obj-&gt;fids_to_coexpressed_fids($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types24">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a scored_fids
fids is a reference to a list where each element is a fid
fid is a string
scored_fids is a reference to a list where each element is a scored_fid
scored_fid is a reference to a list containing 2 items:
	0: a fid
	1: (score) a float</pre>

</dd>
<dt id="Description24">Description</dt>
<dd>

<p>The routine fids_to_coexpressed_fids returns (for each input fid) a list of features that appear to be coexpressed. That is, for an input fid, we determine the set of fids from the same genome that have Pearson Correlation Coefficients (based on normalized expression data) greater than 0.5 or less than -0.5.</p>

</dd>
</dl>

<h2 id="protein_families_to_fids">protein_families_to_fids</h2>

<pre><code>  $return = $obj-&gt;protein_families_to_fids($protein_families)</code></pre>

<dl>

<dt id="Parameter-and-return-types25">Parameter and return types</dt>
<dd>



<pre>
$protein_families is a protein_families
$return is a reference to a hash where the key is a protein_family and the value is a fids
protein_families is a reference to a list where each element is a protein_family
protein_family is a string
fids is a reference to a list where each element is a fid
fid is a string</pre>

</dd>
<dt id="Description25">Description</dt>
<dd>

<p>protein_families_to_fids can be used to access the set of fids represented by each of a set of protein_families. We define protein_families as sets of fids (rather than sets of MD5s. This may, or may not, be a mistake.</p>

</dd>
</dl>

<h2 id="protein_families_to_proteins">protein_families_to_proteins</h2>

<pre><code>  $return = $obj-&gt;protein_families_to_proteins($protein_families)</code></pre>

<dl>

<dt id="Parameter-and-return-types26">Parameter and return types</dt>
<dd>



<pre>
$protein_families is a protein_families
$return is a reference to a hash where the key is a protein_family and the value is a proteins
protein_families is a reference to a list where each element is a protein_family
protein_family is a string
proteins is a reference to a list where each element is a protein
protein is a string</pre>

</dd>
<dt id="Description26">Description</dt>
<dd>

<p>protein_families_to_proteins can be used to access the set of proteins (i.e., the set of MD5 values) represented by each of a set of protein_families. We define protein_families as sets of fids (rather than sets of MD5s. This may, or may not, be a mistake.</p>

</dd>
</dl>

<h2 id="protein_families_to_functions">protein_families_to_functions</h2>

<pre><code>  $return = $obj-&gt;protein_families_to_functions($protein_families)</code></pre>

<dl>

<dt id="Parameter-and-return-types27">Parameter and return types</dt>
<dd>



<pre>
$protein_families is a protein_families
$return is a reference to a hash where the key is a protein_family and the value is a function
protein_families is a reference to a list where each element is a protein_family
protein_family is a string
function is a string</pre>

</dd>
<dt id="Description27">Description</dt>
<dd>

<p>protein_families_to_functions can be used to extract the set of functions assigned to the fids that make up the family. Each input protein_family is mapped to a family function.</p>

</dd>
</dl>

<h2 id="protein_families_to_co_occurring_families">protein_families_to_co_occurring_families</h2>

<pre><code>  $return = $obj-&gt;protein_families_to_co_occurring_families($protein_families)</code></pre>

<dl>

<dt id="Parameter-and-return-types28">Parameter and return types</dt>
<dd>



<pre>
$protein_families is a protein_families
$return is a reference to a hash where the key is a protein_family and the value is a fc_protein_families
protein_families is a reference to a list where each element is a protein_family
protein_family is a string
fc_protein_families is a reference to a list where each element is a fc_protein_family
fc_protein_family is a reference to a list containing 3 items:
	0: a protein_family
	1: a score
	2: a function
score is a float
function is a string</pre>

</dd>
<dt id="Description28">Description</dt>
<dd>

<p>Since we accumulate data relating to the co-occurrence (i.e., chromosomal clustering) of genes in prokaryotic genomes, we can note which pairs of genes tend to co-occur. From this data, one can compute the protein families that tend to co-occur (i.e., tend to cluster on the chromosome). This allows one to formulate conjectures for unclustered pairs, based on clustered pairs from the same protein_families.</p>

</dd>
</dl>

<h2 id="co_occurrence_evidence">co_occurrence_evidence</h2>

<pre><code>  $return = $obj-&gt;co_occurrence_evidence($pairs_of_fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types29">Parameter and return types</dt>
<dd>



<pre>
$pairs_of_fids is a pairs_of_fids
$return is a reference to a list where each element is a reference to a list containing 2 items:
	0: a pair_of_fids
	1: an evidence
pairs_of_fids is a reference to a list where each element is a pair_of_fids
pair_of_fids is a reference to a list containing 2 items:
	0: a fid
	1: a fid
fid is a string
evidence is a reference to a list where each element is a pair_of_fids</pre>

</dd>
<dt id="Description29">Description</dt>
<dd>

<p>co-occurence_evidence is used to retrieve the detailed pairs of genes that go into the computation of co-occurence scores. The scores reflect an estimate of the number of distinct OTUs that contain an instance of a co-occuring pair. This routine returns as evidence a list of all the pairs that went into the computation.</p>

<p>The input to the computation is a list of pairs for which evidence is desired.</p>

<p>The returned output is a list of elements. one for each input pair. Each output element is a 2-tuple: the input pair and the evidence for the pair. The evidence is a list of pairs of fids that are believed to correspond to the input pair.</p>

</dd>
</dl>

<h2 id="contigs_to_sequences">contigs_to_sequences</h2>

<pre><code>  $return = $obj-&gt;contigs_to_sequences($contigs)</code></pre>

<dl>

<dt id="Parameter-and-return-types30">Parameter and return types</dt>
<dd>



<pre>
$contigs is a contigs
$return is a reference to a hash where the key is a contig and the value is a dna
contigs is a reference to a list where each element is a contig
contig is a string
dna is a string</pre>

</dd>
<dt id="Description30">Description</dt>
<dd>

<p>contigs_to_sequences is used to access the DNA sequence associated with each of a set of input contigs. It takes as input a set of contig IDs (from which the genome can be determined) and produces a mapping from the input IDs to the returned DNA sequence in each case.</p>

</dd>
</dl>

<h2 id="contigs_to_lengths">contigs_to_lengths</h2>

<pre><code>  $return = $obj-&gt;contigs_to_lengths($contigs)</code></pre>

<dl>

<dt id="Parameter-and-return-types31">Parameter and return types</dt>
<dd>



<pre>
$contigs is a contigs
$return is a reference to a hash where the key is a contig and the value is a length
contigs is a reference to a list where each element is a contig
contig is a string
length is an int</pre>

</dd>
<dt id="Description31">Description</dt>
<dd>

<p>In some cases, one wishes to know just the lengths of the contigs, rather than their actual DNA sequence (e.g., suppose that you wished to know if a gene boundary occured within 100 bp of the end of the contig). To avoid requiring a user to access the entire DNA sequence, we offer the ability to retrieve just the contig lengths. Input to the routine is a list of contig IDs. The routine returns a mapping from contig IDs to lengths</p>

</dd>
</dl>

<h2 id="contigs_to_md5s">contigs_to_md5s</h2>

<pre><code>  $return = $obj-&gt;contigs_to_md5s($contigs)</code></pre>

<dl>

<dt id="Parameter-and-return-types32">Parameter and return types</dt>
<dd>



<pre>
$contigs is a contigs
$return is a reference to a hash where the key is a contig and the value is a md5
contigs is a reference to a list where each element is a contig
contig is a string
md5 is a string</pre>

</dd>
<dt id="Description32">Description</dt>
<dd>

<p>contigs_to_md5s can be used to acquire MD5 values for each of a list of contigs. The quickest way to determine whether two contigs are identical is to compare their associated MD5 values, eliminating the need to retrieve the sequence of each and compare them.</p>

<p>The routine takes as input a list of contig IDs. The output is a mapping from contig ID to MD5 value.</p>

</dd>
</dl>

<h2 id="md5s_to_genomes">md5s_to_genomes</h2>

<pre><code>  $return = $obj-&gt;md5s_to_genomes($md5s)</code></pre>

<dl>

<dt id="Parameter-and-return-types33">Parameter and return types</dt>
<dd>



<pre>
$md5s is a md5s
$return is a reference to a hash where the key is a md5 and the value is a genomes
md5s is a reference to a list where each element is a md5
md5 is a string
genomes is a reference to a list where each element is a genome
genome is a string</pre>

</dd>
<dt id="Description33">Description</dt>
<dd>

<p>md5s to genomes is used to get the genomes associated with each of a list of input md5 values.</p>

<pre><code>           The routine takes as input a list of MD5 values.  It constructs a mapping from each input
           MD5 value to a list of genomes that share the same MD5 value.

           The MD5 value for a genome is independent of the names of contigs and the case of the DNA sequence
           data.</code></pre>

</dd>
</dl>

<h2 id="genomes_to_md5s">genomes_to_md5s</h2>

<pre><code>  $return = $obj-&gt;genomes_to_md5s($genomes)</code></pre>

<dl>

<dt id="Parameter-and-return-types34">Parameter and return types</dt>
<dd>



<pre>
$genomes is a genomes
$return is a reference to a hash where the key is a genome and the value is a md5
genomes is a reference to a list where each element is a genome
genome is a string
md5 is a string</pre>

</dd>
<dt id="Description34">Description</dt>
<dd>

<p>The routine genomes_to_md5s can be used to look up the MD5 value associated with each of a set of genomes. The MD5 values are computed when the genome is loaded, so this routine just retrieves the precomputed values.</p>

<p>Note that the MD5 value of a genome is independent of the contig names and case of the DNA sequences that make up the genome.</p>

</dd>
</dl>

<h2 id="genomes_to_contigs">genomes_to_contigs</h2>

<pre><code>  $return = $obj-&gt;genomes_to_contigs($genomes)</code></pre>

<dl>

<dt id="Parameter-and-return-types35">Parameter and return types</dt>
<dd>



<pre>
$genomes is a genomes
$return is a reference to a hash where the key is a genome and the value is a contigs
genomes is a reference to a list where each element is a genome
genome is a string
contigs is a reference to a list where each element is a contig
contig is a string</pre>

</dd>
<dt id="Description35">Description</dt>
<dd>

<p>The routine genomes_to_contigs can be used to retrieve the IDs of the contigs associated with each of a list of input genomes. The routine constructs a mapping from genome ID to the list of contigs included in the genome.</p>

</dd>
</dl>

<h2 id="genomes_to_fids">genomes_to_fids</h2>

<pre><code>  $return = $obj-&gt;genomes_to_fids($genomes, $types_of_fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types36">Parameter and return types</dt>
<dd>



<pre>
$genomes is a genomes
$types_of_fids is a types_of_fids
$return is a reference to a hash where the key is a genome and the value is a fids
genomes is a reference to a list where each element is a genome
genome is a string
types_of_fids is a reference to a list where each element is a type_of_fid
type_of_fid is a string
fids is a reference to a list where each element is a fid
fid is a string</pre>

</dd>
<dt id="Description36">Description</dt>
<dd>

<p>genomes_to_fids bis used to get the fids included in specific genomes. It is often the case that you want just one or two types of fids -- hence, the types_of_fids argument.</p>

</dd>
</dl>

<h2 id="genomes_to_taxonomies">genomes_to_taxonomies</h2>

<pre><code>  $return = $obj-&gt;genomes_to_taxonomies($genomes)</code></pre>

<dl>

<dt id="Parameter-and-return-types37">Parameter and return types</dt>
<dd>



<pre>
$genomes is a genomes
$return is a reference to a hash where the key is a genome and the value is a taxonomic_groups
genomes is a reference to a list where each element is a genome
genome is a string
taxonomic_groups is a reference to a list where each element is a taxonomic_group
taxonomic_group is a string</pre>

</dd>
<dt id="Description37">Description</dt>
<dd>

<p>The routine genomes_to_taxonomies can be used to retrieve taxonomic information for each of a list of input genomes. For each genome in the input list of genomes, a list of taxonomic groups is returned. Kbase will use the groups maintained by NCBI. For an NCBI taxonomic string like</p>

<pre><code>     cellular organisms;
     Bacteria;
     Proteobacteria;
     Gammaproteobacteria;
     Enterobacteriales;
     Enterobacteriaceae;
     Escherichia;
     Escherichia coli</code></pre>

<p>associated with the strain &#39;Escherichia coli 1412&#39;, this routine would return a list of these taxonomic groups:</p>

<pre><code>     [&#39;Bacteria&#39;,
      &#39;Proteobacteria&#39;,
      &#39;Gammaproteobacteria&#39;,
      &#39;Enterobacteriales&#39;,
      &#39;Enterobacteriaceae&#39;,
      &#39;Escherichia&#39;,
      &#39;Escherichia coli&#39;,
      &#39;Escherichia coli 1412&#39;
     ]</code></pre>

<p>That is, the initial &quot;cellular organisms&quot; has been deleted, and the strain ID has been added as the last &quot;grouping&quot;.</p>

<p>The output is a mapping from genome IDs to lists of the form shown above.</p>

</dd>
</dl>

<h2 id="genomes_to_subsystems">genomes_to_subsystems</h2>

<pre><code>  $return = $obj-&gt;genomes_to_subsystems($genomes)</code></pre>

<dl>

<dt id="Parameter-and-return-types38">Parameter and return types</dt>
<dd>



<pre>
$genomes is a genomes
$return is a reference to a hash where the key is a genome and the value is a variant_subsystem_pairs
genomes is a reference to a list where each element is a genome
genome is a string
variant_subsystem_pairs is a reference to a list where each element is a variant_of_subsystem
variant_of_subsystem is a reference to a list containing 2 items:
	0: a subsystem
	1: a variant
subsystem is a string
variant is a string</pre>

</dd>
<dt id="Description38">Description</dt>
<dd>

<p>A user can invoke genomes_to_subsystems to rerieve the names of the subsystems relevant to each genome. The input is a list of genomes. The output is a mapping from genome to a list of 2-tuples, where each 2-tuple give a variant code and a subsystem name. Variant codes of -1 (or *-1) amount to assertions that the genome contains no active variant. A variant code of 0 means &quot;work in progress&quot;, and presence or absence of the subsystem in the genome should be undetermined.</p>

</dd>
</dl>

<h2 id="subsystems_to_genomes">subsystems_to_genomes</h2>

<pre><code>  $return = $obj-&gt;subsystems_to_genomes($subsystems)</code></pre>

<dl>

<dt id="Parameter-and-return-types39">Parameter and return types</dt>
<dd>



<pre>
$subsystems is a subsystems
$return is a reference to a hash where the key is a subsystem and the value is a reference to a list where each element is a reference to a list containing 2 items:
	0: a variant
	1: a genome
subsystems is a reference to a list where each element is a subsystem
subsystem is a string
variant is a string
genome is a string</pre>

</dd>
<dt id="Description39">Description</dt>
<dd>

<p>The routine subsystems_to_genomes is used to determine which genomes are in specified subsystems. The input is the list of subsystem names of interest. The output is a map from the subsystem names to lists of 2-tuples, where each 2-tuple is a [variant-code,genome ID] pair.</p>

</dd>
</dl>

<h2 id="subsystems_to_fids">subsystems_to_fids</h2>

<pre><code>  $return = $obj-&gt;subsystems_to_fids($subsystems, $genomes)</code></pre>

<dl>

<dt id="Parameter-and-return-types40">Parameter and return types</dt>
<dd>



<pre>
$subsystems is a subsystems
$genomes is a genomes
$return is a reference to a hash where the key is a subsystem and the value is a reference to a hash where the key is a genome and the value is a reference to a list where each element is a reference to a list containing 2 items:
	0: a variant
	1: a fids
subsystems is a reference to a list where each element is a subsystem
subsystem is a string
genomes is a reference to a list where each element is a genome
genome is a string
variant is a string
fids is a reference to a list where each element is a fid
fid is a string</pre>

</dd>
<dt id="Description40">Description</dt>
<dd>

<p>The routine subsystems_to_fids allows the user to map subsystem names into the fids that occur in genomes in the subsystems. Specifically, the input is a list of subsystem names. What is returned is a mapping from subsystem names to a &quot;genome-mapping&quot;. The genome-mapping takes genome IDs to 2-tuples that capture the variant code of the genome and the fids from the genome that are included in the subsystem.</p>

</dd>
</dl>

<h2 id="subsystems_to_roles">subsystems_to_roles</h2>

<pre><code>  $return = $obj-&gt;subsystems_to_roles($subsystems, $aux)</code></pre>

<dl>

<dt id="Parameter-and-return-types41">Parameter and return types</dt>
<dd>



<pre>
$subsystems is a subsystems
$aux is an aux
$return is a reference to a hash where the key is a subsystem and the value is a roles
subsystems is a reference to a list where each element is a subsystem
subsystem is a string
aux is an int
roles is a reference to a list where each element is a role
role is a string</pre>

</dd>
<dt id="Description41">Description</dt>
<dd>

<p>The routine subsystem_to_roles is used to determine the role descriptions that occur in a subsystem. The input is a list of subsystem names. A map is returned connecting subsystem names to lists of roles. &#39;aux&#39; is a boolean variable. If it is 0, auxiliary roles are not returned. If it is 1, they are returned.</p>

</dd>
</dl>

<h2 id="subsystems_to_spreadsheets">subsystems_to_spreadsheets</h2>

<pre><code>  $return = $obj-&gt;subsystems_to_spreadsheets($subsystems, $genomes)</code></pre>

<dl>

<dt id="Parameter-and-return-types42">Parameter and return types</dt>
<dd>



<pre>
$subsystems is a subsystems
$genomes is a genomes
$return is a reference to a hash where the key is a subsystem and the value is a reference to a hash where the key is a genome and the value is a row
subsystems is a reference to a list where each element is a subsystem
subsystem is a string
genomes is a reference to a list where each element is a genome
genome is a string
row is a reference to a list containing 2 items:
	0: a variant
	1: a reference to a hash where the key is a role and the value is a fids
variant is a string
role is a string
fids is a reference to a list where each element is a fid
fid is a string</pre>

</dd>
<dt id="Description42">Description</dt>
<dd>

<p>The subsystem_to_spreadsheet routine allows a user to extract the subsystem spreadsheets for a specified set of subsystem names. In the returned output, each subsystem is mapped to a hash that takes as input a genome ID and maps it to the &quot;row&quot; for the genome in the subsystem. The &quot;row&quot; is itself a 2-tuple composed of the variant code, and a mapping from role descriptions to lists of fids. We suggest writing a simple test script to get, say, the subsystem named &#39;Histidine Degradation&#39;, extracting the spreadsheet, and then using something like Dumper to make sure that it all makes sense.</p>

</dd>
</dl>

<h2 id="all_roles_used_in_models">all_roles_used_in_models</h2>

<pre><code>  $return = $obj-&gt;all_roles_used_in_models()</code></pre>

<dl>

<dt id="Parameter-and-return-types43">Parameter and return types</dt>
<dd>



<pre>
$return is a roles
roles is a reference to a list where each element is a role
role is a string</pre>

</dd>
<dt id="Description43">Description</dt>
<dd>

<p>The all_roles_used_in_models allows a user to access the set of roles that are included in current models. This is important. There are far fewer roles used in models than overall. Hence, the returned set represents the minimal set we need to clean up in order to properly support modeling.</p>

</dd>
</dl>

<h2 id="complexes_to_complex_data">complexes_to_complex_data</h2>

<pre><code>  $return = $obj-&gt;complexes_to_complex_data($complexes)</code></pre>

<dl>

<dt id="Parameter-and-return-types44">Parameter and return types</dt>
<dd>



<pre>
$complexes is a complexes
$return is a reference to a hash where the key is a complex and the value is a complex_data
complexes is a reference to a list where each element is a complex
complex is a string
complex_data is a reference to a hash where the following keys are defined:
	complex_name has a value which is a name
	complex_roles has a value which is a roles_with_flags
	complex_reactions has a value which is a reactions
name is a string
roles_with_flags is a reference to a list where each element is a role_with_flag
role_with_flag is a reference to a list containing 2 items:
	0: a role
	1: an optional
role is a string
optional is a string
reactions is a reference to a list where each element is a reaction
reaction is a string</pre>

</dd>
<dt id="Description44">Description</dt>
<dd>

</dd>
</dl>

<h2 id="genomes_to_genome_data">genomes_to_genome_data</h2>

<pre><code>  $return = $obj-&gt;genomes_to_genome_data($genomes)</code></pre>

<dl>

<dt id="Parameter-and-return-types45">Parameter and return types</dt>
<dd>



<pre>
$genomes is a genomes
$return is a reference to a hash where the key is a genome and the value is a genome_data
genomes is a reference to a list where each element is a genome
genome is a string
genome_data is a reference to a hash where the following keys are defined:
	complete has a value which is an int
	contigs has a value which is an int
	dna_size has a value which is an int
	gc_content has a value which is a float
	genetic_code has a value which is an int
	pegs has a value which is an int
	rnas has a value which is an int
	scientific_name has a value which is a string
	taxonomy has a value which is a string
	genome_md5 has a value which is a string</pre>

</dd>
<dt id="Description45">Description</dt>
<dd>

</dd>
</dl>

<h2 id="fids_to_regulon_data">fids_to_regulon_data</h2>

<pre><code>  $return = $obj-&gt;fids_to_regulon_data($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types46">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a regulons_data
fids is a reference to a list where each element is a fid
fid is a string
regulons_data is a reference to a list where each element is a regulon_data
regulon_data is a reference to a hash where the following keys are defined:
	regulon_id has a value which is a regulon
	regulon_set has a value which is a fids
	tfs has a value which is a fids
regulon is a string</pre>

</dd>
<dt id="Description46">Description</dt>
<dd>

</dd>
</dl>

<h2 id="regulons_to_fids">regulons_to_fids</h2>

<pre><code>  $return = $obj-&gt;regulons_to_fids($regulons)</code></pre>

<dl>

<dt id="Parameter-and-return-types47">Parameter and return types</dt>
<dd>



<pre>
$regulons is a regulons
$return is a reference to a hash where the key is a regulon and the value is a fids
regulons is a reference to a list where each element is a regulon
regulon is a string
fids is a reference to a list where each element is a fid
fid is a string</pre>

</dd>
<dt id="Description47">Description</dt>
<dd>

</dd>
</dl>

<h2 id="fids_to_feature_data">fids_to_feature_data</h2>

<pre><code>  $return = $obj-&gt;fids_to_feature_data($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types48">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a feature_data
fids is a reference to a list where each element is a fid
fid is a string
feature_data is a reference to a hash where the following keys are defined:
	feature_id has a value which is a fid
	genome_name has a value which is a string
	feature_function has a value which is a string
	feature_length has a value which is an int
	feature_publications has a value which is a pubrefs
	feature_location has a value which is a location
pubrefs is a reference to a list where each element is a pubref
pubref is a reference to a list containing 3 items:
	0: (id) a string
	1: (link) a string
	2: (title) a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig
	1: a begin
	2: a strand
	3: a length
contig is a string
begin is an int
strand is a string
length is an int</pre>

</dd>
<dt id="Description48">Description</dt>
<dd>

</dd>
</dl>

<h2 id="equiv_sequence_assertions">equiv_sequence_assertions</h2>

<pre><code>  $return = $obj-&gt;equiv_sequence_assertions($proteins)</code></pre>

<dl>

<dt id="Parameter-and-return-types49">Parameter and return types</dt>
<dd>



<pre>
$proteins is a proteins
$return is a reference to a hash where the key is a protein and the value is a function_assertions
proteins is a reference to a list where each element is a protein
protein is a string
function_assertions is a reference to a list where each element is a function_assertion
function_assertion is a reference to a list containing 3 items:
	0: an id
	1: a function
	2: a source
id is a string
function is a string
source is a string</pre>

</dd>
<dt id="Description49">Description</dt>
<dd>

<p>Different groups have made assertions of function for numerous protein sequences. The equiv_sequence_assertions allows the user to gather function assertions from all of the sources. Each assertion includes a field indicating whether the person making the assertion viewed themself as an &quot;expert&quot;. The routine gathers assertions for all proteins having identical protein sequence.</p>

</dd>
</dl>

<h2 id="fids_to_atomic_regulons">fids_to_atomic_regulons</h2>

<pre><code>  $return = $obj-&gt;fids_to_atomic_regulons($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types50">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is an atomic_regulon_size_pairs
fids is a reference to a list where each element is a fid
fid is a string
atomic_regulon_size_pairs is a reference to a list where each element is an atomic_regulon_size_pair
atomic_regulon_size_pair is a reference to a list containing 2 items:
	0: an atomic_regulon
	1: an atomic_regulon_size
atomic_regulon is a string
atomic_regulon_size is an int</pre>

</dd>
<dt id="Description50">Description</dt>
<dd>

<p>The fids_to_atomic_regulons allows one to map fids into regulons that contain the fids. Normally a fid will be in at most one regulon, but we support multiple regulons.</p>

</dd>
</dl>

<h2 id="atomic_regulons_to_fids">atomic_regulons_to_fids</h2>

<pre><code>  $return = $obj-&gt;atomic_regulons_to_fids($atomic_regulons)</code></pre>

<dl>

<dt id="Parameter-and-return-types51">Parameter and return types</dt>
<dd>



<pre>
$atomic_regulons is an atomic_regulons
$return is a reference to a hash where the key is an atomic_regulon and the value is a fids
atomic_regulons is a reference to a list where each element is an atomic_regulon
atomic_regulon is a string
fids is a reference to a list where each element is a fid
fid is a string</pre>

</dd>
<dt id="Description51">Description</dt>
<dd>

<p>The atomic_regulons_to_fids routine allows the user to access the set of fids that make up a regulon. Regulons may arise from several sources; hence, fids can be in multiple regulons.</p>

</dd>
</dl>

<h2 id="fids_to_protein_sequences">fids_to_protein_sequences</h2>

<pre><code>  $return = $obj-&gt;fids_to_protein_sequences($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types52">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a protein_sequence
fids is a reference to a list where each element is a fid
fid is a string
protein_sequence is a string</pre>

</dd>
<dt id="Description52">Description</dt>
<dd>

<p>fids_to_protein_sequences allows the user to look up the amino acid sequences corresponding to each of a set of fids. You can also get the sequence from proteins (i.e., md5 values). This routine saves you having to look up the md5 sequence and then accessing the protein string in a separate call.</p>

</dd>
</dl>

<h2 id="fids_to_proteins">fids_to_proteins</h2>

<pre><code>  $return = $obj-&gt;fids_to_proteins($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types53">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a md5
fids is a reference to a list where each element is a fid
fid is a string
md5 is a string</pre>

</dd>
<dt id="Description53">Description</dt>
<dd>

</dd>
</dl>

<h2 id="fids_to_dna_sequences">fids_to_dna_sequences</h2>

<pre><code>  $return = $obj-&gt;fids_to_dna_sequences($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types54">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a dna_sequence
fids is a reference to a list where each element is a fid
fid is a string
dna_sequence is a string</pre>

</dd>
<dt id="Description54">Description</dt>
<dd>

<p>fids_to_dna_sequences allows the user to look up the DNA sequences corresponding to each of a set of fids.</p>

</dd>
</dl>

<h2 id="roles_to_fids">roles_to_fids</h2>

<pre><code>  $return = $obj-&gt;roles_to_fids($roles, $genomes)</code></pre>

<dl>

<dt id="Parameter-and-return-types55">Parameter and return types</dt>
<dd>



<pre>
$roles is a roles
$genomes is a genomes
$return is a reference to a hash where the key is a role and the value is a fid
roles is a reference to a list where each element is a role
role is a string
genomes is a reference to a list where each element is a genome
genome is a string
fid is a string</pre>

</dd>
<dt id="Description55">Description</dt>
<dd>

<p>A &quot;function&quot; is a set of &quot;roles&quot; (often called &quot;functional roles&quot;);</p>

<pre><code>                F1 / F2  (where F1 and F2 are roles)  is a function that implements
                          two functional roles in different domains of the protein.
                F1 @ F2 implements multiple roles through broad specificity
                F1; F2  is thought to implement F1 or f2 (uncertainty)

            You often wish to find the fids in one or more genomes that
            implement specific functional roles.  To do this, you can use
            roles_to_fids.</code></pre>

</dd>
</dl>

<h2 id="reactions_to_complexes">reactions_to_complexes</h2>

<pre><code>  $return = $obj-&gt;reactions_to_complexes($reactions)</code></pre>

<dl>

<dt id="Parameter-and-return-types56">Parameter and return types</dt>
<dd>



<pre>
$reactions is a reactions
$return is a reference to a hash where the key is a reaction and the value is a complexes_with_flags
reactions is a reference to a list where each element is a reaction
reaction is a string
complexes_with_flags is a reference to a list where each element is a complex_with_flag
complex_with_flag is a reference to a list containing 2 items:
	0: a complex
	1: an optional
complex is a string
optional is a string</pre>

</dd>
<dt id="Description56">Description</dt>
<dd>

<p>Reactions are thought of as being either spontaneous or implemented by one or more Complexes. Complexes connect to Roles. Hence, the connection of fids or roles to reactions goes through Complexes.</p>

</dd>
</dl>

<h2 id="aliases_to_fids">aliases_to_fids</h2>

<pre><code>  $return = $obj-&gt;aliases_to_fids($aliases)</code></pre>

<dl>

<dt id="Parameter-and-return-types57">Parameter and return types</dt>
<dd>



<pre>
$aliases is an aliases
$return is a reference to a hash where the key is an alias and the value is a fid
aliases is a reference to a list where each element is an alias
alias is a string
fid is a string</pre>

</dd>
<dt id="Description57">Description</dt>
<dd>

</dd>
</dl>

<h2 id="aliases_to_fids_by_source">aliases_to_fids_by_source</h2>

<pre><code>  $return = $obj-&gt;aliases_to_fids_by_source($aliases, $source)</code></pre>

<dl>

<dt id="Parameter-and-return-types58">Parameter and return types</dt>
<dd>



<pre>
$aliases is an aliases
$source is a string
$return is a reference to a hash where the key is an alias and the value is a fid
aliases is a reference to a list where each element is an alias
alias is a string
fid is a string</pre>

</dd>
<dt id="Description58">Description</dt>
<dd>

</dd>
</dl>

<h2 id="source_ids_to_fids">source_ids_to_fids</h2>

<pre><code>  $return = $obj-&gt;source_ids_to_fids($aliases)</code></pre>

<dl>

<dt id="Parameter-and-return-types59">Parameter and return types</dt>
<dd>



<pre>
$aliases is an aliases
$return is a reference to a hash where the key is a string and the value is a reference to a list where each element is a fid
aliases is a reference to a list where each element is an alias
alias is a string
fid is a string</pre>

</dd>
<dt id="Description59">Description</dt>
<dd>

</dd>
</dl>

<h2 id="external_ids_to_fids">external_ids_to_fids</h2>

<pre><code>  $return = $obj-&gt;external_ids_to_fids($external_ids, $prefix_match)</code></pre>

<dl>

<dt id="Parameter-and-return-types60">Parameter and return types</dt>
<dd>



<pre>
$external_ids is an external_ids
$prefix_match is an int
$return is a reference to a hash where the key is an external_id and the value is a fid
external_ids is a reference to a list where each element is an external_id
external_id is a string
fid is a string</pre>

</dd>
<dt id="Description60">Description</dt>
<dd>

</dd>
</dl>

<h2 id="reaction_strings">reaction_strings</h2>

<pre><code>  $return = $obj-&gt;reaction_strings($reactions, $name_parameter)</code></pre>

<dl>

<dt id="Parameter-and-return-types61">Parameter and return types</dt>
<dd>



<pre>
$reactions is a reactions
$name_parameter is a name_parameter
$return is a reference to a hash where the key is a reaction and the value is a string
reactions is a reference to a list where each element is a reaction
reaction is a string
name_parameter is a string</pre>

</dd>
<dt id="Description61">Description</dt>
<dd>

<p>Reaction_strings are text strings that represent (albeit crudely) the details of Reactions.</p>

</dd>
</dl>

<h2 id="roles_to_complexes">roles_to_complexes</h2>

<pre><code>  $return = $obj-&gt;roles_to_complexes($roles)</code></pre>

<dl>

<dt id="Parameter-and-return-types62">Parameter and return types</dt>
<dd>



<pre>
$roles is a roles
$return is a reference to a hash where the key is a role and the value is a complexes
roles is a reference to a list where each element is a role
role is a string
complexes is a reference to a list where each element is a complex
complex is a string</pre>

</dd>
<dt id="Description62">Description</dt>
<dd>

<p>roles_to_complexes allows a user to connect Roles to Complexes, from there, the connection exists to Reactions (although in the actual ER-model model, the connection from Complex to Reaction goes through ReactionComplex). Since Roles also connect to fids, the connection between fids and Reactions is induced.</p>

<p>The &quot;name_parameter&quot; can be 0, 1 or &#39;only&#39;. If 1, then the compound name will be included with the ID in the output. If only, the compound name will be included instead of the ID. If 0, only the ID will be included. The default is 0.</p>

</dd>
</dl>

<h2 id="complexes_to_roles">complexes_to_roles</h2>

<pre><code>  $return = $obj-&gt;complexes_to_roles($complexes)</code></pre>

<dl>

<dt id="Parameter-and-return-types63">Parameter and return types</dt>
<dd>



<pre>
$complexes is a complexes
$return is a reference to a hash where the key is a complexes and the value is a roles
complexes is a reference to a list where each element is a complex
complex is a string
roles is a reference to a list where each element is a role
role is a string</pre>

</dd>
<dt id="Description63">Description</dt>
<dd>

</dd>
</dl>

<h2 id="fids_to_subsystem_data">fids_to_subsystem_data</h2>

<pre><code>  $return = $obj-&gt;fids_to_subsystem_data($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types64">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a ss_var_role_tuples
fids is a reference to a list where each element is a fid
fid is a string
ss_var_role_tuples is a reference to a list where each element is a ss_var_role_tuple
ss_var_role_tuple is a reference to a list containing 3 items:
	0: a subsystem
	1: a variant
	2: a role
subsystem is a string
variant is a string
role is a string</pre>

</dd>
<dt id="Description64">Description</dt>
<dd>

</dd>
</dl>

<h2 id="representative">representative</h2>

<pre><code>  $return = $obj-&gt;representative($genomes)</code></pre>

<dl>

<dt id="Parameter-and-return-types65">Parameter and return types</dt>
<dd>



<pre>
$genomes is a genomes
$return is a reference to a hash where the key is a genome and the value is a genome
genomes is a reference to a list where each element is a genome
genome is a string</pre>

</dd>
<dt id="Description65">Description</dt>
<dd>

</dd>
</dl>

<h2 id="otu_members">otu_members</h2>

<pre><code>  $return = $obj-&gt;otu_members($genomes)</code></pre>

<dl>

<dt id="Parameter-and-return-types66">Parameter and return types</dt>
<dd>



<pre>
$genomes is a genomes
$return is a reference to a hash where the key is a genome and the value is a reference to a hash where the key is a genome and the value is a genome_name
genomes is a reference to a list where each element is a genome
genome is a string
genome_name is a string</pre>

</dd>
<dt id="Description66">Description</dt>
<dd>

</dd>
</dl>

<h2 id="otus_to_representatives">otus_to_representatives</h2>

<pre><code>  $return = $obj-&gt;otus_to_representatives($otus)</code></pre>

<dl>

<dt id="Parameter-and-return-types67">Parameter and return types</dt>
<dd>



<pre>
$otus is a reference to a list where each element is an int
$return is a reference to a hash where the key is an int and the value is a genome
genome is a string</pre>

</dd>
<dt id="Description67">Description</dt>
<dd>

</dd>
</dl>

<h2 id="fids_to_genomes">fids_to_genomes</h2>

<pre><code>  $return = $obj-&gt;fids_to_genomes($fids)</code></pre>

<dl>

<dt id="Parameter-and-return-types68">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$return is a reference to a hash where the key is a fid and the value is a genome
fids is a reference to a list where each element is a fid
fid is a string
genome is a string</pre>

</dd>
<dt id="Description68">Description</dt>
<dd>

</dd>
</dl>

<h2 id="text_search">text_search</h2>

<pre><code>  $return = $obj-&gt;text_search($input, $start, $count, $entities)</code></pre>

<dl>

<dt id="Parameter-and-return-types69">Parameter and return types</dt>
<dd>



<pre>
$input is a string
$start is an int
$count is an int
$entities is a reference to a list where each element is a string
$return is a reference to a hash where the key is an entity_name and the value is a reference to a list where each element is a search_hit
entity_name is a string
search_hit is a reference to a list containing 2 items:
	0: a weight
	1: a reference to a hash where the key is a field_name and the value is a string
weight is an int
field_name is a string</pre>

</dd>
<dt id="Description69">Description</dt>
<dd>

<p>text_search performs a search against a full-text index maintained for the CDMI. The parameter &quot;input&quot; is the text string to be searched for. The parameter &quot;entities&quot; defines the entities to be searched. If the list is empty, all indexed entities will be searched. The &quot;start&quot; and &quot;count&quot; parameters limit the results to &quot;count&quot; hits starting at &quot;start&quot;.</p>

</dd>
</dl>

<h2 id="corresponds">corresponds</h2>

<pre><code>  $return = $obj-&gt;corresponds($fids, $genome)</code></pre>

<dl>

<dt id="Parameter-and-return-types70">Parameter and return types</dt>
<dd>



<pre>
$fids is a fids
$genome is a genome
$return is a reference to a hash where the key is a fid and the value is a correspondence
fids is a reference to a list where each element is a fid
fid is a string
genome is a string
correspondence is a reference to a hash where the following keys are defined:
	to has a value which is a fid
	iden has a value which is a float
	ncontext has a value which is an int
	b1 has a value which is an int
	e1 has a value which is an int
	ln1 has a value which is an int
	b2 has a value which is an int
	e2 has a value which is an int
	ln2 has a value which is an int
	score has a value which is an int</pre>

</dd>
<dt id="Description70">Description</dt>
<dd>

</dd>
</dl>

<h2 id="corresponds_from_sequences">corresponds_from_sequences</h2>

<pre><code>  $return = $obj-&gt;corresponds_from_sequences($g1_sequences, $g1_locations, $g2_sequences, $g2_locations)</code></pre>

<dl>

<dt id="Parameter-and-return-types71">Parameter and return types</dt>
<dd>



<pre>
$g1_sequences is a reference to a list where each element is a reference to a list containing 2 items:
	0: a fid
	1: a protein_sequence
$g1_locations is a reference to a list where each element is a reference to a list containing 2 items:
	0: a fid
	1: a location
$g2_sequences is a reference to a list where each element is a reference to a list containing 2 items:
	0: a fid
	1: a protein_sequence
$g2_locations is a reference to a list where each element is a reference to a list containing 2 items:
	0: a fid
	1: a location
$return is a reference to a hash where the key is a fid and the value is a correspondence
fid is a string
protein_sequence is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig
	1: a begin
	2: a strand
	3: a length
contig is a string
begin is an int
strand is a string
length is an int
correspondence is a reference to a hash where the following keys are defined:
	to has a value which is a fid
	iden has a value which is a float
	ncontext has a value which is an int
	b1 has a value which is an int
	e1 has a value which is an int
	ln1 has a value which is an int
	b2 has a value which is an int
	e2 has a value which is an int
	ln2 has a value which is an int
	score has a value which is an int</pre>

</dd>
<dt id="Description71">Description</dt>
<dd>

</dd>
</dl>

<h2 id="close_genomes">close_genomes</h2>

<pre><code>  $return = $obj-&gt;close_genomes($seq_set, $n)</code></pre>

<dl>

<dt id="Parameter-and-return-types72">Parameter and return types</dt>
<dd>



<pre>
$seq_set is a seq_set
$n is an int
$return is a reference to a list where each element is a reference to a list containing 2 items:
	0: a genome
	1: (ident) a float
seq_set is a reference to a list where each element is a seq_triple
seq_triple is a reference to a list containing 3 items:
	0: an id
	1: a comment
	2: a sequence
id is a string
comment is a string
sequence is a string
genome is a string</pre>

</dd>
<dt id="Description72">Description</dt>
<dd>

</dd>
</dl>

<h2 id="representative_sequences">representative_sequences</h2>

<pre><code>  $return_1, $return_2 = $obj-&gt;representative_sequences($seq_set, $rep_seq_parms)</code></pre>

<dl>

<dt id="Parameter-and-return-types73">Parameter and return types</dt>
<dd>



<pre>
$seq_set is a seq_set
$rep_seq_parms is a rep_seq_parms
$return_1 is an id_set
$return_2 is a reference to a list where each element is an id_set
seq_set is a reference to a list where each element is a seq_triple
seq_triple is a reference to a list containing 3 items:
	0: an id
	1: a comment
	2: a sequence
id is a string
comment is a string
sequence is a string
rep_seq_parms is a reference to a hash where the following keys are defined:
	existing_reps has a value which is a seq_set
	order has a value which is a string
	alg has a value which is an int
	type_sim has a value which is a string
	cutoff has a value which is a float
id_set is a reference to a list where each element is an id</pre>

</dd>
<dt id="Description73">Description</dt>
<dd>

<p>we return two arguments. The first is the list of representative triples, and the second is the list of sets (the first entry always being the representative sequence)</p>

</dd>
</dl>

<h2 id="align_sequences">align_sequences</h2>

<pre><code>  $return = $obj-&gt;align_sequences($seq_set, $align_seq_parms)</code></pre>

<dl>

<dt id="Parameter-and-return-types74">Parameter and return types</dt>
<dd>



<pre>
$seq_set is a seq_set
$align_seq_parms is an align_seq_parms
$return is an alignment
seq_set is a reference to a list where each element is a seq_triple
seq_triple is a reference to a list containing 3 items:
	0: an id
	1: a comment
	2: a sequence
id is a string
comment is a string
sequence is a string
align_seq_parms is a reference to a hash where the following keys are defined:
	muscle_parms has a value which is a muscle_parms_t
	mafft_parms has a value which is a mafft_parms_t
	tool has a value which is a string
	align_ends_with_clustal has a value which is an int
muscle_parms_t is a reference to a hash where the following keys are defined:
	anchors has a value which is an int
	brenner has a value which is an int
	cluster has a value which is an int
	dimer has a value which is an int
	diags has a value which is an int
	diags1 has a value which is an int
	diags2 has a value which is an int
	le has a value which is an int
	noanchors has a value which is an int
	sp has a value which is an int
	spn has a value which is an int
	stable has a value which is an int
	sv has a value which is an int
	anchorspacing has a value which is a string
	center has a value which is a string
	cluster1 has a value which is a string
	cluster2 has a value which is a string
	diagbreak has a value which is a string
	diaglength has a value which is a string
	diagmargin has a value which is a string
	distance1 has a value which is a string
	distance2 has a value which is a string
	gapopen has a value which is a string
	log has a value which is a string
	loga has a value which is a string
	matrix has a value which is a string
	maxhours has a value which is a string
	maxiters has a value which is a string
	maxmb has a value which is a string
	maxtrees has a value which is a string
	minbestcolscore has a value which is a string
	minsmoothscore has a value which is a string
	objscore has a value which is a string
	refinewindow has a value which is a string
	root1 has a value which is a string
	root2 has a value which is a string
	scorefile has a value which is a string
	seqtype has a value which is a string
	smoothscorecell has a value which is a string
	smoothwindow has a value which is a string
	spscore has a value which is a string
	SUEFF has a value which is a string
	usetree has a value which is a string
	weight1 has a value which is a string
	weight2 has a value which is a string
mafft_parms_t is a reference to a hash where the following keys are defined:
	sixmerpair has a value which is an int
	amino has a value which is an int
	anysymbol has a value which is an int
	auto has a value which is an int
	clustalout has a value which is an int
	dpparttree has a value which is an int
	fastapair has a value which is an int
	fastaparttree has a value which is an int
	fft has a value which is an int
	fmodel has a value which is an int
	genafpair has a value which is an int
	globalpair has a value which is an int
	inputorder has a value which is an int
	localpair has a value which is an int
	memsave has a value which is an int
	nofft has a value which is an int
	noscore has a value which is an int
	parttree has a value which is an int
	reorder has a value which is an int
	treeout has a value which is an int
	alg has a value which is a string
	aamatrix has a value which is a string
	bl has a value which is a string
	ep has a value which is a string
	groupsize has a value which is a string
	jtt has a value which is a string
	lap has a value which is a string
	lep has a value which is a string
	lepx has a value which is a string
	LOP has a value which is a string
	LEXP has a value which is a string
	maxiterate has a value which is a string
	op has a value which is a string
	partsize has a value which is a string
	retree has a value which is a string
	thread has a value which is a string
	tm has a value which is a string
	weighti has a value which is a string
alignment is a seq_set</pre>

</dd>
<dt id="Description74">Description</dt>
<dd>

</dd>
</dl>

<h2 id="build_tree">build_tree</h2>

<pre><code>  $return = $obj-&gt;build_tree($alignment, $build_tree_parms)</code></pre>

<dl>

<dt id="Parameter-and-return-types75">Parameter and return types</dt>
<dd>



<pre>
$alignment is an alignment
$build_tree_parms is a build_tree_parms
$return is a newick_tree
alignment is a seq_set
seq_set is a reference to a list where each element is a seq_triple
seq_triple is a reference to a list containing 3 items:
	0: an id
	1: a comment
	2: a sequence
id is a string
comment is a string
sequence is a string
build_tree_parms is a reference to a hash where the following keys are defined:
	bootstrap has a value which is a string
	model has a value which is a string
	nclasses has a value which is a string
	nproc has a value which is a string
	rate has a value which is a string
	search has a value which is a string
	tool has a value which is a string
	tool_params has a value which is a string
newick_tree is a string</pre>

</dd>
<dt id="Description75">Description</dt>
<dd>

</dd>
</dl>

<h2 id="alignment_by_id">alignment_by_id</h2>

<pre><code>  $return = $obj-&gt;alignment_by_id($aln_id)</code></pre>

<dl>

<dt id="Parameter-and-return-types76">Parameter and return types</dt>
<dd>



<pre>
$aln_id is an aln_id
$return is an alignment
aln_id is a string
alignment is a seq_set
seq_set is a reference to a list where each element is a seq_triple
seq_triple is a reference to a list containing 3 items:
	0: an id
	1: a comment
	2: a sequence
id is a string
comment is a string
sequence is a string</pre>

</dd>
<dt id="Description76">Description</dt>
<dd>

</dd>
</dl>

<h2 id="tree_by_id">tree_by_id</h2>

<pre><code>  $return = $obj-&gt;tree_by_id($tree_id)</code></pre>

<dl>

<dt id="Parameter-and-return-types77">Parameter and return types</dt>
<dd>



<pre>
$tree_id is a tree_id
$return is a newick_tree
tree_id is a string
newick_tree is a string</pre>

</dd>
<dt id="Description77">Description</dt>
<dd>

</dd>
</dl>

<h2 id="all_entities">all_entities</h2>

<pre><code>  $return = $obj-&gt;all_entities()</code></pre>

<dl>

<dt id="Parameter-and-return-types78">Parameter and return types</dt>
<dd>



<pre>
$return is an entity_names
entity_names is a reference to a list where each element is an entity_name
entity_name is a string</pre>

</dd>
<dt id="Description78">Description</dt>
<dd>

<p>Returns a list of all entities in the database.</p>

</dd>
</dl>

<h2 id="all_relationships">all_relationships</h2>

<pre><code>  $return = $obj-&gt;all_relationships()</code></pre>

<dl>

<dt id="Parameter-and-return-types79">Parameter and return types</dt>
<dd>



<pre>
$return is a relationship_names
relationship_names is a reference to a list where each element is a relationship_name
relationship_name is a string</pre>

</dd>
<dt id="Description79">Description</dt>
<dd>

<p>Returns a list of all relationships in the database.</p>

</dd>
</dl>

<h2 id="get_entity">get_entity</h2>

<pre><code>  $return = $obj-&gt;get_entity($entity_names)</code></pre>

<dl>

<dt id="Parameter-and-return-types80">Parameter and return types</dt>
<dd>



<pre>
$entity_names is an entity_names
$return is a reference to a hash where the key is a string and the value is an entity_info
entity_names is a reference to a list where each element is an entity_name
entity_name is a string
entity_info is a reference to a hash where the following keys are defined:
	name has a value which is a string
	relationships has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (rel_name) a string
	1: (entity_name) a string	fields has a value which is a reference to a hash where the key is a field_name and the value is a field_info
field_name is a string
field_info is a reference to a hash where the following keys are defined:
	name has a value which is a string
	notes has a value which is a string
	type has a value which is a string
</pre>

</dd>
<dt id="Description80">Description</dt>
<dd>

<p>Returns information about a set of entities in the database. Invalid entity names are ignored.</p>

</dd>
</dl>

<h2 id="get_relationship">get_relationship</h2>

<pre><code>  $return = $obj-&gt;get_relationship($relationship_names)</code></pre>

<dl>

<dt id="Parameter-and-return-types81">Parameter and return types</dt>
<dd>



<pre>
$relationship_names is a relationship_names
$return is a reference to a hash where the key is a string and the value is a relationship_info
relationship_names is a reference to a list where each element is a relationship_name
relationship_name is a string
relationship_info is a reference to a hash where the following keys are defined:
	name has a value which is a string
	from_entity has a value which is a string
	to_entity has a value which is a string
	real_table has a value which is a boolean
	converse has a value which is a string
	fields has a value which is a reference to a hash where the key is a field_name and the value is a field_info
boolean is an int
field_name is a string
field_info is a reference to a hash where the following keys are defined:
	name has a value which is a string
	notes has a value which is a string
	type has a value which is a string</pre>

</dd>
<dt id="Description81">Description</dt>
<dd>

<p>Returns information about a set of relationships in the database. Invalid relationship names are ignored.</p>

</dd>
</dl>

<h2 id="version">version</h2>

<pre><code>  $return = $obj-&gt;version()</code></pre>

<dl>

<dt id="Parameter-and-return-types82">Parameter and return types</dt>
<dd>



<pre>
$return is a string
</pre>

</dd>
<dt id="Description82">Description</dt>
<dd>

<p>Return the module version. This is a Semantic Versioning number.</p>

</dd>
</dl>

<h1 id="TYPES">TYPES</h1>

<h2 id="annotator">annotator</h2>

<dl>

<dt id="Definition">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="annotation_time">annotation_time</h2>

<dl>

<dt id="Definition1">Definition</dt>
<dd>



<pre>
an int
</pre>

</dd>
</dl>

<h2 id="comment">comment</h2>

<dl>

<dt id="Definition2">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="fid">fid</h2>

<dl>

<dt id="Description83">Description</dt>
<dd>

<p>A fid is a &quot;feature id&quot;. A feature represents an ordered list of regions from the contigs of a genome. Features all have types. This allows you to speak of not only protein-encoding genes (PEGs) and RNAs, but also binding sites, large regions, etc. The location of a fid is defined as a list of &quot;location of a contiguous DNA string&quot; pieces (see the description of the type &quot;location&quot;)</p>

</dd>
<dt id="Definition3">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="protein_family">protein_family</h2>

<dl>

<dt id="Description84">Description</dt>
<dd>

<p>A protein_family is thought of as a set of isofunctional, homologous protein sequences. This is not exactly what other groups have meant by &quot;protein families&quot;. There is no hierarchy of super-family, family, sub-family. We plan on loading different collections of protein families, but in many cases there will need to be a transformation into the concept used by Kbase.</p>

</dd>
<dt id="Definition4">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="role">role</h2>

<dl>

<dt id="Description85">Description</dt>
<dd>

<p>The concept of &quot;role&quot; or &quot;functional role&quot; is basically an atomic functional unit. The &quot;function of a protein&quot; is made up of one or more roles. That is, a bifunctional protein with an assigned function of</p>

<pre><code>   5-Enolpyruvylshikimate-3-phosphate synthase (EC 2.5.1.19) / Cytidylate kinase (EC 2.7.4.14)</code></pre>

<p>would implement two distinct roles (the &quot;function1 / function2&quot; notation is intended to assert that the initial part of the protein implements function1, and the terminal part of the protein implements function2). It is worth noting that a protein often implements multiple roles due to broad specificity. In this case, we suggest describing the protein function as</p>

<pre><code>     function1 @ function2</code></pre>

<p>That is the &#39; / &#39; separator is used to represent multiple roles implemented by distinct domains of the protein, while &#39; @ &#39; is used to represent multiple roles implemented by distinct domains.</p>

</dd>
<dt id="Definition5">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="subsystem">subsystem</h2>

<dl>

<dt id="Description86">Description</dt>
<dd>

<p>A substem is composed of two components: a set of roles that are gathered to be annotated simultaneously and a spreadsheet depicting the proteins within each genome that implement the roles. The set of roles may correspond to a pathway, a complex, an inventory (say, &quot;transporters&quot;) or whatever other principle an annotator used to formulate the subsystem.</p>

<p>The subsystem spreadsheet is a list of &quot;rows&quot;, each representing the subsytem in a specific genome. Each row includes a variant code (indicating what version of the molecular machine exists in the genome) and cells. Each cell is a 2-tuple:</p>

<pre><code>     [role,protein-encoding genes that implement the role in the genome]</code></pre>

<p>Annotators construct subsystems, and in the process impose a controlled vocabulary for roles and functions.</p>

</dd>
<dt id="Definition6">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="variant">variant</h2>

<dl>

<dt id="Definition7">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="variant_of_subsystem">variant_of_subsystem</h2>

<dl>

<dt id="Definition8">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: a subsystem
1: a variant</pre>

</dd>
</dl>

<h2 id="variant_subsystem_pairs">variant_subsystem_pairs</h2>

<dl>

<dt id="Definition9">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a variant_of_subsystem
</pre>

</dd>
</dl>

<h2 id="type_of_fid">type_of_fid</h2>

<dl>

<dt id="Definition10">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="types_of_fids">types_of_fids</h2>

<dl>

<dt id="Definition11">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a type_of_fid
</pre>

</dd>
</dl>

<h2 id="length">length</h2>

<dl>

<dt id="Definition12">Definition</dt>
<dd>



<pre>
an int
</pre>

</dd>
</dl>

<h2 id="begin">begin</h2>

<dl>

<dt id="Definition13">Definition</dt>
<dd>



<pre>
an int
</pre>

</dd>
</dl>

<h2 id="strand">strand</h2>

<dl>

<dt id="Description87">Description</dt>
<dd>

<p>In encodings of locations, we often specify strands. We specify the strand as &#39;+&#39; or &#39;-&#39;</p>

</dd>
<dt id="Definition14">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="contig">contig</h2>

<dl>

<dt id="Definition15">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="region_of_dna">region_of_dna</h2>

<dl>

<dt id="Description88">Description</dt>
<dd>

<p>A region of DNA is maintained as a tuple of four components:</p>

<pre><code>                the contig
                the beginning position (from 1)
                the strand
                the length

           We often speak of &quot;a region&quot;.  By &quot;location&quot;, we mean a sequence
           of regions from the same genome (perhaps from distinct contigs).</code></pre>

</dd>
<dt id="Definition16">Definition</dt>
<dd>



<pre>
a reference to a list containing 4 items:
0: a contig
1: a begin
2: a strand
3: a length</pre>

</dd>
</dl>

<h2 id="location">location</h2>

<dl>

<dt id="Description89">Description</dt>
<dd>

<p>a &quot;location&quot; refers to a sequence of regions</p>

</dd>
<dt id="Definition17">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a region_of_dna
</pre>

</dd>
</dl>

<h2 id="locations">locations</h2>

<dl>

<dt id="Definition18">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a location
</pre>

</dd>
</dl>

<h2 id="region_of_dna_string">region_of_dna_string</h2>

<dl>

<dt id="Description90">Description</dt>
<dd>

<p>we often need to represent regions or locations as strings. We would use something like</p>

<pre><code>     contigA_200+100,contigA_402+188</code></pre>

<p>to represent a location composed of two regions</p>

</dd>
<dt id="Definition19">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="region_of_dna_strings">region_of_dna_strings</h2>

<dl>

<dt id="Definition20">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a region_of_dna_string
</pre>

</dd>
</dl>

<h2 id="location_string">location_string</h2>

<dl>

<dt id="Definition21">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="dna">dna</h2>

<dl>

<dt id="Definition22">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="function">function</h2>

<dl>

<dt id="Definition23">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="protein">protein</h2>

<dl>

<dt id="Definition24">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="md5">md5</h2>

<dl>

<dt id="Definition25">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="genome">genome</h2>

<dl>

<dt id="Definition26">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="taxonomic_group">taxonomic_group</h2>

<dl>

<dt id="Definition27">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="annotation">annotation</h2>

<dl>

<dt id="Description91">Description</dt>
<dd>

<p>The Kbase stores annotations relating to features. Each annotation is a 3-tuple:</p>

<pre><code>     the text of the annotation (often a record of assertion of function)

     the annotator attaching the annotation to the feature

     the time (in seconds from the epoch) at which the annotation was attached</code></pre>

</dd>
<dt id="Definition28">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: a comment
1: an annotator
2: an annotation_time</pre>

</dd>
</dl>

<h2 id="pubref">pubref</h2>

<dl>

<dt id="Description92">Description</dt>
<dd>

<p>The Kbase will include a growing body of literature supporting protein functions, asserted phenotypes, etc. References are encoded as 3-tuples:</p>

<pre><code>     an id (often a PubMed ID)

     a URL to the paper

     a title of the paper</code></pre>

<p>The URL and title are often missing (but, can usually be inferred from the pubmed ID).</p>

</dd>
<dt id="Definition29">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: (id) a string
1: (link) a string
2: (title) a string</pre>

</dd>
</dl>

<h2 id="scored_fid">scored_fid</h2>

<dl>

<dt id="Definition30">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: a fid
1: (score) a float</pre>

</dd>
</dl>

<h2 id="annotations">annotations</h2>

<dl>

<dt id="Definition31">Definition</dt>
<dd>



<pre>
a reference to a list where each element is an annotation
</pre>

</dd>
</dl>

<h2 id="pubrefs">pubrefs</h2>

<dl>

<dt id="Definition32">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a pubref
</pre>

</dd>
</dl>

<h2 id="roles">roles</h2>

<dl>

<dt id="Definition33">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a role
</pre>

</dd>
</dl>

<h2 id="optional">optional</h2>

<dl>

<dt id="Definition34">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="role_with_flag">role_with_flag</h2>

<dl>

<dt id="Definition35">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: a role
1: an optional</pre>

</dd>
</dl>

<h2 id="roles_with_flags">roles_with_flags</h2>

<dl>

<dt id="Definition36">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a role_with_flag
</pre>

</dd>
</dl>

<h2 id="scored_fids">scored_fids</h2>

<dl>

<dt id="Definition37">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a scored_fid
</pre>

</dd>
</dl>

<h2 id="proteins">proteins</h2>

<dl>

<dt id="Definition38">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a protein
</pre>

</dd>
</dl>

<h2 id="functions">functions</h2>

<dl>

<dt id="Definition39">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a function
</pre>

</dd>
</dl>

<h2 id="taxonomic_groups">taxonomic_groups</h2>

<dl>

<dt id="Definition40">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a taxonomic_group
</pre>

</dd>
</dl>

<h2 id="subsystems">subsystems</h2>

<dl>

<dt id="Definition41">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a subsystem
</pre>

</dd>
</dl>

<h2 id="contigs">contigs</h2>

<dl>

<dt id="Definition42">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a contig
</pre>

</dd>
</dl>

<h2 id="md5s">md5s</h2>

<dl>

<dt id="Definition43">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a md5
</pre>

</dd>
</dl>

<h2 id="genomes">genomes</h2>

<dl>

<dt id="Definition44">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a genome
</pre>

</dd>
</dl>

<h2 id="pair_of_fids">pair_of_fids</h2>

<dl>

<dt id="Definition45">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: a fid
1: a fid</pre>

</dd>
</dl>

<h2 id="pairs_of_fids">pairs_of_fids</h2>

<dl>

<dt id="Definition46">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a pair_of_fids
</pre>

</dd>
</dl>

<h2 id="protein_families">protein_families</h2>

<dl>

<dt id="Definition47">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a protein_family
</pre>

</dd>
</dl>

<h2 id="score">score</h2>

<dl>

<dt id="Definition48">Definition</dt>
<dd>



<pre>
a float
</pre>

</dd>
</dl>

<h2 id="evidence">evidence</h2>

<dl>

<dt id="Definition49">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a pair_of_fids
</pre>

</dd>
</dl>

<h2 id="fids">fids</h2>

<dl>

<dt id="Definition50">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a fid
</pre>

</dd>
</dl>

<h2 id="row">row</h2>

<dl>

<dt id="Definition51">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: a variant
1: a reference to a hash where the key is a role and the value is a fids</pre>

</dd>
</dl>

<h2 id="fid_function_pair">fid_function_pair</h2>

<dl>

<dt id="Definition52">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: a fid
1: a function</pre>

</dd>
</dl>

<h2 id="fid_function_pairs">fid_function_pairs</h2>

<dl>

<dt id="Definition53">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a fid_function_pair
</pre>

</dd>
</dl>

<h2 id="fc_protein_family">fc_protein_family</h2>

<dl>

<dt id="Description93">Description</dt>
<dd>

<p>A functionally coupled protein family identifies a family, a score, and a function (of the related family)</p>

</dd>
<dt id="Definition54">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: a protein_family
1: a score
2: a function</pre>

</dd>
</dl>

<h2 id="fc_protein_families">fc_protein_families</h2>

<dl>

<dt id="Definition55">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a fc_protein_family
</pre>

</dd>
</dl>

<h2 id="allele">allele</h2>

<dl>

<dt id="Description94">Description</dt>
<dd>

<p>We now have a number of types and functions relating to ObservationalUnits (ous), alleles and traits. We think of a reference genome and a set of ous that have measured differences (SNPs) when compared to the reference genome. Each allele is associated with a position on a contig of the reference genome. Prior analysis has associated traits with the alleles that impact them. We are interested in supporting operations that locate genes in the region of an allele (i.e., genes of the reference genome that are in a region containining an allele). Similarly, we wish to locate the alleles that impact a trait, map the alleles to regions, loacte the possibly impacted genes, relate these to subsystems, etc.</p>

</dd>
<dt id="Definition56">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="alleles">alleles</h2>

<dl>

<dt id="Definition57">Definition</dt>
<dd>



<pre>
a reference to a list where each element is an allele
</pre>

</dd>
</dl>

<h2 id="trait">trait</h2>

<dl>

<dt id="Definition58">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="traits">traits</h2>

<dl>

<dt id="Definition59">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a trait
</pre>

</dd>
</dl>

<h2 id="ou">ou</h2>

<dl>

<dt id="Definition60">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="ous">ous</h2>

<dl>

<dt id="Definition61">Definition</dt>
<dd>



<pre>
a reference to a list where each element is an ou
</pre>

</dd>
</dl>

<h2 id="bp_loc">bp_loc</h2>

<dl>

<dt id="Definition62">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: a contig
1: (position) an int</pre>

</dd>
</dl>

<h2 id="measurement_type">measurement_type</h2>

<dl>

<dt id="Definition63">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="measurement_value">measurement_value</h2>

<dl>

<dt id="Definition64">Definition</dt>
<dd>



<pre>
a float
</pre>

</dd>
</dl>

<h2 id="aux">aux</h2>

<dl>

<dt id="Definition65">Definition</dt>
<dd>



<pre>
an int
</pre>

</dd>
</dl>

<h2 id="fields">fields</h2>

<dl>

<dt id="Definition66">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a string
</pre>

</dd>
</dl>

<h2 id="complex">complex</h2>

<dl>

<dt id="Definition67">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="complex_with_flag">complex_with_flag</h2>

<dl>

<dt id="Definition68">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: a complex
1: an optional</pre>

</dd>
</dl>

<h2 id="complexes_with_flags">complexes_with_flags</h2>

<dl>

<dt id="Definition69">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a complex_with_flag
</pre>

</dd>
</dl>

<h2 id="complexes">complexes</h2>

<dl>

<dt id="Definition70">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a complex
</pre>

</dd>
</dl>

<h2 id="name">name</h2>

<dl>

<dt id="Definition71">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="reaction">reaction</h2>

<dl>

<dt id="Definition72">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="reactions">reactions</h2>

<dl>

<dt id="Definition73">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a reaction
</pre>

</dd>
</dl>

<h2 id="complex_data">complex_data</h2>

<dl>

<dt id="Description95">Description</dt>
<dd>

<p>Reactions do not connect directly to roles. Rather, the conceptual model is that one or more roles together form a complex. A complex implements one or more reactions. The actual data relating to a complex is spread over two entities: Complex and ReactionComplex. It is convenient to be able to offer access to the complex name, the reactions it implements, and the roles that make it up in a single invocation.</p>

</dd>
<dt id="Definition74">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
complex_name has a value which is a name
complex_roles has a value which is a roles_with_flags
complex_reactions has a value which is a reactions</pre>

</dd>
</dl>

<h2 id="genome_data">genome_data</h2>

<dl>

<dt id="Definition75">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
complete has a value which is an int
contigs has a value which is an int
dna_size has a value which is an int
gc_content has a value which is a float
genetic_code has a value which is an int
pegs has a value which is an int
rnas has a value which is an int
scientific_name has a value which is a string
taxonomy has a value which is a string
genome_md5 has a value which is a string</pre>

</dd>
</dl>

<h2 id="regulon">regulon</h2>

<dl>

<dt id="Definition76">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="regulons">regulons</h2>

<dl>

<dt id="Definition77">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a regulon
</pre>

</dd>
</dl>

<h2 id="regulon_data">regulon_data</h2>

<dl>

<dt id="Definition78">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
regulon_id has a value which is a regulon
regulon_set has a value which is a fids
tfs has a value which is a fids</pre>

</dd>
</dl>

<h2 id="regulons_data">regulons_data</h2>

<dl>

<dt id="Definition79">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a regulon_data
</pre>

</dd>
</dl>

<h2 id="feature_data">feature_data</h2>

<dl>

<dt id="Definition80">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
feature_id has a value which is a fid
genome_name has a value which is a string
feature_function has a value which is a string
feature_length has a value which is an int
feature_publications has a value which is a pubrefs
feature_location has a value which is a location</pre>

</dd>
</dl>

<h2 id="expert">expert</h2>

<dl>

<dt id="Definition81">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="source">source</h2>

<dl>

<dt id="Definition82">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="id">id</h2>

<dl>

<dt id="Definition83">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="function_assertion">function_assertion</h2>

<dl>

<dt id="Definition84">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: an id
1: a function
2: a source</pre>

</dd>
</dl>

<h2 id="function_assertions">function_assertions</h2>

<dl>

<dt id="Definition85">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a function_assertion
</pre>

</dd>
</dl>

<h2 id="atomic_regulon">atomic_regulon</h2>

<dl>

<dt id="Definition86">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="atomic_regulon_size">atomic_regulon_size</h2>

<dl>

<dt id="Definition87">Definition</dt>
<dd>



<pre>
an int
</pre>

</dd>
</dl>

<h2 id="atomic_regulon_size_pair">atomic_regulon_size_pair</h2>

<dl>

<dt id="Definition88">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: an atomic_regulon
1: an atomic_regulon_size</pre>

</dd>
</dl>

<h2 id="atomic_regulon_size_pairs">atomic_regulon_size_pairs</h2>

<dl>

<dt id="Definition89">Definition</dt>
<dd>



<pre>
a reference to a list where each element is an atomic_regulon_size_pair
</pre>

</dd>
</dl>

<h2 id="atomic_regulons">atomic_regulons</h2>

<dl>

<dt id="Definition90">Definition</dt>
<dd>



<pre>
a reference to a list where each element is an atomic_regulon
</pre>

</dd>
</dl>

<h2 id="protein_sequence">protein_sequence</h2>

<dl>

<dt id="Definition91">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="dna_sequence">dna_sequence</h2>

<dl>

<dt id="Definition92">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="alias">alias</h2>

<dl>

<dt id="Definition93">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="aliases">aliases</h2>

<dl>

<dt id="Definition94">Definition</dt>
<dd>



<pre>
a reference to a list where each element is an alias
</pre>

</dd>
</dl>

<h2 id="external_id">external_id</h2>

<dl>

<dt id="Definition95">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="external_ids">external_ids</h2>

<dl>

<dt id="Definition96">Definition</dt>
<dd>



<pre>
a reference to a list where each element is an external_id
</pre>

</dd>
</dl>

<h2 id="name_parameter">name_parameter</h2>

<dl>

<dt id="Definition97">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="ss_var_role_tuple">ss_var_role_tuple</h2>

<dl>

<dt id="Definition98">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: a subsystem
1: a variant
2: a role</pre>

</dd>
</dl>

<h2 id="ss_var_role_tuples">ss_var_role_tuples</h2>

<dl>

<dt id="Definition99">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a ss_var_role_tuple
</pre>

</dd>
</dl>

<h2 id="genome_name">genome_name</h2>

<dl>

<dt id="Definition100">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="entity_name">entity_name</h2>

<dl>

<dt id="Definition101">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="weight">weight</h2>

<dl>

<dt id="Definition102">Definition</dt>
<dd>



<pre>
an int
</pre>

</dd>
</dl>

<h2 id="field_name">field_name</h2>

<dl>

<dt id="Definition103">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="search_hit">search_hit</h2>

<dl>

<dt id="Definition104">Definition</dt>
<dd>



<pre>
a reference to a list containing 2 items:
0: a weight
1: a reference to a hash where the key is a field_name and the value is a string</pre>

</dd>
</dl>

<h2 id="correspondence">correspondence</h2>

<dl>

<dt id="Description96">Description</dt>
<dd>

<p>A correspondence is generated as a mapping of fids to fids. The mapping attempts to map a fid to another that performs the same function. The correspondence describes the regions that are similar, the strength of the similarity, the number of genes in the chromosomal context that appear to &quot;correspond&quot; and a score from 0 to 1 that loosely corresponds to confidence in the correspondence.</p>

</dd>
<dt id="Definition105">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
to has a value which is a fid
iden has a value which is a float
ncontext has a value which is an int
b1 has a value which is an int
e1 has a value which is an int
ln1 has a value which is an int
b2 has a value which is an int
e2 has a value which is an int
ln2 has a value which is an int
score has a value which is an int</pre>

</dd>
</dl>

<h2 id="sequence">sequence</h2>

<dl>

<dt id="Description97">Description</dt>
<dd>

<p>A close_genomes is used to get a set of relatively close genomes (for each input genome, a set of close genomes is calculated, but the result should be viewed as quite approximate. It is quite slow, using similarities for a universal protein as the basis for the assessments. It produces estimates of degree of similarity for the universal proteins it samples.</p>

<p>Up to n genomes will be returned for each input genome.</p>

</dd>
<dt id="Definition106">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="seq_triple">seq_triple</h2>

<dl>

<dt id="Definition107">Definition</dt>
<dd>



<pre>
a reference to a list containing 3 items:
0: an id
1: a comment
2: a sequence</pre>

</dd>
</dl>

<h2 id="seq_set">seq_set</h2>

<dl>

<dt id="Definition108">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a seq_triple
</pre>

</dd>
</dl>

<h2 id="id_set">id_set</h2>

<dl>

<dt id="Definition109">Definition</dt>
<dd>



<pre>
a reference to a list where each element is an id
</pre>

</dd>
</dl>

<h2 id="alignment">alignment</h2>

<dl>

<dt id="Definition110">Definition</dt>
<dd>



<pre>
a seq_set
</pre>

</dd>
</dl>

<h2 id="rep_seq_parms">rep_seq_parms</h2>

<dl>

<dt id="Description98">Description</dt>
<dd>

<p>fractions or bits</p>

</dd>
<dt id="Definition111">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
existing_reps has a value which is a seq_set
order has a value which is a string
alg has a value which is an int
type_sim has a value which is a string
cutoff has a value which is a float</pre>

</dd>
</dl>

<h2 id="muscle_parms_t">muscle_parms_t</h2>

<dl>

<dt id="Definition112">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
anchors has a value which is an int
brenner has a value which is an int
cluster has a value which is an int
dimer has a value which is an int
diags has a value which is an int
diags1 has a value which is an int
diags2 has a value which is an int
le has a value which is an int
noanchors has a value which is an int
sp has a value which is an int
spn has a value which is an int
stable has a value which is an int
sv has a value which is an int
anchorspacing has a value which is a string
center has a value which is a string
cluster1 has a value which is a string
cluster2 has a value which is a string
diagbreak has a value which is a string
diaglength has a value which is a string
diagmargin has a value which is a string
distance1 has a value which is a string
distance2 has a value which is a string
gapopen has a value which is a string
log has a value which is a string
loga has a value which is a string
matrix has a value which is a string
maxhours has a value which is a string
maxiters has a value which is a string
maxmb has a value which is a string
maxtrees has a value which is a string
minbestcolscore has a value which is a string
minsmoothscore has a value which is a string
objscore has a value which is a string
refinewindow has a value which is a string
root1 has a value which is a string
root2 has a value which is a string
scorefile has a value which is a string
seqtype has a value which is a string
smoothscorecell has a value which is a string
smoothwindow has a value which is a string
spscore has a value which is a string
SUEFF has a value which is a string
usetree has a value which is a string
weight1 has a value which is a string
weight2 has a value which is a string</pre>

</dd>
</dl>

<h2 id="mafft_parms_t">mafft_parms_t</h2>

<dl>

<dt id="Description99">Description</dt>
<dd>

<p>linsi | einsi | ginsi | nwnsi | nwns | fftnsi | fftns (D)</p>

</dd>
<dt id="Definition113">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
sixmerpair has a value which is an int
amino has a value which is an int
anysymbol has a value which is an int
auto has a value which is an int
clustalout has a value which is an int
dpparttree has a value which is an int
fastapair has a value which is an int
fastaparttree has a value which is an int
fft has a value which is an int
fmodel has a value which is an int
genafpair has a value which is an int
globalpair has a value which is an int
inputorder has a value which is an int
localpair has a value which is an int
memsave has a value which is an int
nofft has a value which is an int
noscore has a value which is an int
parttree has a value which is an int
reorder has a value which is an int
treeout has a value which is an int
alg has a value which is a string
aamatrix has a value which is a string
bl has a value which is a string
ep has a value which is a string
groupsize has a value which is a string
jtt has a value which is a string
lap has a value which is a string
lep has a value which is a string
lepx has a value which is a string
LOP has a value which is a string
LEXP has a value which is a string
maxiterate has a value which is a string
op has a value which is a string
partsize has a value which is a string
retree has a value which is a string
thread has a value which is a string
tm has a value which is a string
weighti has a value which is a string</pre>

</dd>
</dl>

<h2 id="align_seq_parms">align_seq_parms</h2>

<dl>

<dt id="Definition114">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
muscle_parms has a value which is a muscle_parms_t
mafft_parms has a value which is a mafft_parms_t
tool has a value which is a string
align_ends_with_clustal has a value which is an int</pre>

</dd>
</dl>

<h2 id="newick_tree">newick_tree</h2>

<dl>

<dt id="Definition115">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="build_tree_parms">build_tree_parms</h2>

<dl>

<dt id="Definition116">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
bootstrap has a value which is a string
model has a value which is a string
nclasses has a value which is a string
nproc has a value which is a string
rate has a value which is a string
search has a value which is a string
tool has a value which is a string
tool_params has a value which is a string</pre>

</dd>
</dl>

<h2 id="aln_id">aln_id</h2>

<dl>

<dt id="Definition117">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="tree_id">tree_id</h2>

<dl>

<dt id="Definition118">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="entity_name1">entity_name</h2>

<dl>

<dt id="Definition119">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="entity_names">entity_names</h2>

<dl>

<dt id="Definition120">Definition</dt>
<dd>



<pre>
a reference to a list where each element is an entity_name
</pre>

</dd>
</dl>

<h2 id="relationship_name">relationship_name</h2>

<dl>

<dt id="Definition121">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="relationship_names">relationship_names</h2>

<dl>

<dt id="Definition122">Definition</dt>
<dd>



<pre>
a reference to a list where each element is a relationship_name
</pre>

</dd>
</dl>

<h2 id="field_name1">field_name</h2>

<dl>

<dt id="Definition123">Definition</dt>
<dd>



<pre>
a string
</pre>

</dd>
</dl>

<h2 id="boolean">boolean</h2>

<dl>

<dt id="Definition124">Definition</dt>
<dd>



<pre>
an int
</pre>

</dd>
</dl>

<h2 id="field_info">field_info</h2>

<dl>

<dt id="Description100">Description</dt>
<dd>

<p>Information about a field in the database. Includes the name of the field, any associated formatted notes, and the type.</p>

</dd>
<dt id="Definition125">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
name has a value which is a string
notes has a value which is a string
type has a value which is a string</pre>

</dd>
</dl>

<h2 id="entity_info">entity_info</h2>

<dl>

<dt id="Description101">Description</dt>
<dd>

<p>Information about an entity in the database, including the entity name and its relationships and fields.</p>

</dd>
<dt id="Definition126">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
name has a value which is a string
relationships has a value which is a reference to a list where each element is a reference to a list containing 2 items:
0: (rel_name) a string
1: (entity_name) a stringfields has a value which is a reference to a hash where the key is a field_name and the value is a field_info</pre>

</dd>
</dl>

<h2 id="relationship_info">relationship_info</h2>

<dl>

<dt id="Description102">Description</dt>
<dd>

<p>Information about a relationship in the database, including the entities it relates, its name and converse name, and its fields. The real_table boolean designates that the relationship is a real table in the database rather than the converse relationship to that table.</p>

</dd>
<dt id="Definition127">Definition</dt>
<dd>



<pre>
a reference to a hash where the following keys are defined:
name has a value which is a string
from_entity has a value which is a string
to_entity has a value which is a string
real_table has a value which is a boolean
converse has a value which is a string
fields has a value which is a reference to a hash where the key is a field_name and the value is a field_info</pre>

</dd>
</dl>


</body>

</html>


